{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9f28f7b9-82e0-4370-b6af-186249f2594b",
   "metadata": {},
   "source": [
    "# Analyze correlation of amino-acid fitnesses with positive selection values\n",
    "This notebook looks at how dN/dS values from FEL correlation with the amino-acid fitness values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e0b14084-028a-4b12-bdaf-8fcf16406d60",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-26T23:46:51.583449Z",
     "iopub.status.busy": "2023-05-26T23:46:51.582486Z",
     "iopub.status.idle": "2023-05-26T23:46:51.896626Z",
     "shell.execute_reply": "2023-05-26T23:46:51.895747Z",
     "shell.execute_reply.started": "2023-05-26T23:46:51.583383Z"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "######## snakemake preamble start (automatically inserted, do not edit) ########\n",
    "import sys; sys.path.extend(['/fh/fast/bloom_j/software/miniconda3/envs/SARS2-mut-fitness/lib/python3.9/site-packages', '/home/jbloom/.cache/snakemake/snakemake/source-cache/runtime-cache/tmpr43t1mvr/file/fh/fast/bloom_j/computational_notebooks/jbloom/2023/SARS2-mut-fitness/notebooks', '/fh/fast/bloom_j/computational_notebooks/jbloom/2023/SARS2-mut-fitness/notebooks']); import pickle; snakemake = pickle.loads(b'\\x80\\x04\\x95\\x9b\\x16\\x00\\x00\\x00\\x00\\x00\\x00\\x8c\\x10snakemake.script\\x94\\x8c\\tSnakemake\\x94\\x93\\x94)\\x81\\x94}\\x94(\\x8c\\x05input\\x94\\x8c\\x0csnakemake.io\\x94\\x8c\\nInputFiles\\x94\\x93\\x94)\\x81\\x94(\\x8cHresults_public_2023-05-11/positive_selection/positive_selection_data.csv\\x94\\x8c3results_public_2023-05-11/aa_fitness/aa_fitness.csv\\x94e}\\x94(\\x8c\\x06_names\\x94}\\x94(\\x8c\\x12positive_selection\\x94K\\x00N\\x86\\x94\\x8c\\naa_fitness\\x94K\\x01N\\x86\\x94u\\x8c\\x12_allowed_overrides\\x94]\\x94(\\x8c\\x05index\\x94\\x8c\\x04sort\\x94eh\\x15\\x8c\\tfunctools\\x94\\x8c\\x07partial\\x94\\x93\\x94h\\x06\\x8c\\x19Namedlist._used_attribute\\x94\\x93\\x94\\x85\\x94R\\x94(h\\x1b)}\\x94\\x8c\\x05_name\\x94h\\x15sNt\\x94bh\\x16h\\x19h\\x1b\\x85\\x94R\\x94(h\\x1b)}\\x94h\\x1fh\\x16sNt\\x94bh\\x0fh\\nh\\x11h\\x0bub\\x8c\\x06output\\x94h\\x06\\x8c\\x0bOutputFiles\\x94\\x93\\x94)\\x81\\x94\\x8cIresults_public_2023-05-11/positive_selection/positive_selection_corr.html\\x94a}\\x94(h\\r}\\x94\\x8c\\tcorr_html\\x94K\\x00N\\x86\\x94sh\\x13]\\x94(h\\x15h\\x16eh\\x15h\\x19h\\x1b\\x85\\x94R\\x94(h\\x1b)}\\x94h\\x1fh\\x15sNt\\x94bh\\x16h\\x19h\\x1b\\x85\\x94R\\x94(h\\x1b)}\\x94h\\x1fh\\x16sNt\\x94bh,h)ub\\x8c\\x06params\\x94h\\x06\\x8c\\x06Params\\x94\\x93\\x94)\\x81\\x94}\\x94(h\\r}\\x94h\\x13]\\x94(h\\x15h\\x16eh\\x15h\\x19h\\x1b\\x85\\x94R\\x94(h\\x1b)}\\x94h\\x1fh\\x15sNt\\x94bh\\x16h\\x19h\\x1b\\x85\\x94R\\x94(h\\x1b)}\\x94h\\x1fh\\x16sNt\\x94bub\\x8c\\twildcards\\x94h\\x06\\x8c\\tWildcards\\x94\\x93\\x94)\\x81\\x94\\x8c\\x11public_2023-05-11\\x94a}\\x94(h\\r}\\x94\\x8c\\x03mat\\x94K\\x00N\\x86\\x94sh\\x13]\\x94(h\\x15h\\x16eh\\x15h\\x19h\\x1b\\x85\\x94R\\x94(h\\x1b)}\\x94h\\x1fh\\x15sNt\\x94bh\\x16h\\x19h\\x1b\\x85\\x94R\\x94(h\\x1b)}\\x94h\\x1fh\\x16sNt\\x94b\\x8c\\x03mat\\x94hJub\\x8c\\x07threads\\x94K\\x01\\x8c\\tresources\\x94h\\x06\\x8c\\tResources\\x94\\x93\\x94)\\x81\\x94(K\\x01K\\x01\\x8c\\x04/tmp\\x94e}\\x94(h\\r}\\x94(\\x8c\\x06_cores\\x94K\\x00N\\x86\\x94\\x8c\\x06_nodes\\x94K\\x01N\\x86\\x94\\x8c\\x06tmpdir\\x94K\\x02N\\x86\\x94uh\\x13]\\x94(h\\x15h\\x16eh\\x15h\\x19h\\x1b\\x85\\x94R\\x94(h\\x1b)}\\x94h\\x1fh\\x15sNt\\x94bh\\x16h\\x19h\\x1b\\x85\\x94R\\x94(h\\x1b)}\\x94h\\x1fh\\x16sNt\\x94bhaK\\x01hcK\\x01heh^ub\\x8c\\x03log\\x94h\\x06\\x8c\\x03Log\\x94\\x93\\x94)\\x81\\x94\\x8cMresults_public_2023-05-11/positive_selection/analyze_positive_selection.ipynb\\x94a}\\x94(h\\r}\\x94\\x8c\\x08notebook\\x94K\\x00N\\x86\\x94sh\\x13]\\x94(h\\x15h\\x16eh\\x15h\\x19h\\x1b\\x85\\x94R\\x94(h\\x1b)}\\x94h\\x1fh\\x15sNt\\x94bh\\x16h\\x19h\\x1b\\x85\\x94R\\x94(h\\x1b)}\\x94h\\x1fh\\x16sNt\\x94bhwhtub\\x8c\\x06config\\x94}\\x94(\\x8c\\tmat_trees\\x94}\\x94(\\x8c\\x11public_2023-05-11\\x94\\x8c\\x83http://hgdownload.soe.ucsc.edu/goldenPath/wuhCor1/UShER_SARS-CoV-2/2023/05/11/public-2023-05-11.all.masked.nextclade.pangolin.pb.gz\\x94\\x8c\\x11public_2023-03-27\\x94\\x8c\\x83http://hgdownload.soe.ucsc.edu/goldenPath/wuhCor1/UShER_SARS-CoV-2/2023/03/27/public-2023-03-27.all.masked.nextclade.pangolin.pb.gz\\x94\\x8c\\x11public_2022-12-18\\x94\\x8c\\x83http://hgdownload.soe.ucsc.edu/goldenPath/wuhCor1/UShER_SARS-CoV-2/2022/12/18/public-2022-12-18.all.masked.nextclade.pangolin.pb.gz\\x94\\x8c\\x11gisaid_2023-03-29\\x94\\x8cQhttps://hgwdev.gi.ucsc.edu/~angie/2847b92/gisaidAndPublic.2023-03-29.masked.pb.gz\\x94u\\x8c\\x0bcurrent_mat\\x94\\x8c\\x11public_2023-05-11\\x94\\x8c\\tref_fasta\\x94\\x8cGhttp://hgdownload.soe.ucsc.edu/goldenPath/wuhCor1/bigZips/wuhCor1.fa.gz\\x94\\x8c\\x07ref_gtf\\x94\\x8cPhttp://hgdownload.soe.ucsc.edu/goldenPath/wuhCor1/bigZips/genes/ncbiGenes.gtf.gz\\x94\\x8c\\x0eadd_to_ref_gtf\\x94}\\x94\\x8c\\x05ORF9b\\x94]\\x94(M|nM\\xa1oes\\x8c\\x11min_clade_samples\\x94M\\x10\\'\\x8c\\x0esample_subsets\\x94}\\x94(\\x8c\\x03all\\x94\\x8c\\x01.\\x94\\x8c\\x03USA\\x94\\x8c\\x03USA\\x94\\x8c\\x07England\\x94\\x8c\\x07England\\x94u\\x8c\\x12clade_founder_json\\x94\\x8cyhttps://raw.githubusercontent.com/neherlab/SC2_variant_rates/7e738194a8c6592082f1caa9a6ca70cb68289790/data/clade_gts.json\\x94\\x8c\\x11clades_to_exclude\\x94]\\x94\\x8c\\x0321A\\x94a\\x8c\\x10max_nt_mutations\\x94K\\x04\\x8c\\x15max_reversions_to_ref\\x94K\\x01\\x8c\\x1fmax_reversions_to_clade_founder\\x94K\\x01\\x8c\\x1bexclude_ref_to_founder_muts\\x94\\x88\\x8c\\x12site_include_range\\x94}\\x94(\\x8c\\x05start\\x94M\\n\\x01\\x8c\\x03end\\x94M\\xeasu\\x8c\\x10sites_to_exclude\\x94]\\x94(K\\x99M}\\x04M\\x96\\x08MI\\x0cM\\xec\\rM\\xc2\\x0eM\\xd2\\x0fMo\\x18MV\\x1fMV\"MZ4M{6M\\x12YME_MF_MeaM\\xfd\\x15M\\xc3\\x1aM\\xa0\\x1cM\\xbfmM\\xb2re\\x8c\\rsite_mask_vcf\\x94\\x8cvhttps://raw.githubusercontent.com/W-L/ProblematicSites_SARS-CoV2/master/subset_vcf/problematic_sites_sarsCov2.mask.vcf\\x94\\x8c\\x12usher_masked_sites\\x94\\x8c\\x1cdata/usher_masked_sites.yaml\\x94\\x8c\\x1dsynonymous_spectra_min_counts\\x94M\\x88\\x13\\x8c\\x0eorf1ab_to_nsps\\x94}\\x94(\\x8c\\x04nsp1\\x94]\\x94(K\\x01K\\xb4e\\x8c\\x04nsp2\\x94]\\x94(K\\xb5M2\\x03e\\x8c\\x04nsp3\\x94]\\x94(M3\\x03M\\xcb\\ne\\x8c\\x04nsp4\\x94]\\x94(M\\xcc\\nM\\xbf\\x0ce\\x8c\\x0bnsp5 (Mpro)\\x94]\\x94(M\\xc0\\x0cM\\xf1\\re\\x8c\\x04nsp6\\x94]\\x94(M\\xf2\\rM\\x13\\x0fe\\x8c\\x04nsp7\\x94]\\x94(M\\x14\\x0fMf\\x0fe\\x8c\\x04nsp8\\x94]\\x94(Mg\\x0fM,\\x10e\\x8c\\x04nsp9\\x94]\\x94(M-\\x10M\\x9d\\x10e\\x8c\\x05nsp10\\x94]\\x94(M\\x9e\\x10M(\\x11e\\x8c\\x0cnsp12 (RdRp)\\x94]\\x94(M)\\x11M\\xcc\\x14e\\x8c\\x05nsp13\\x94]\\x94(M\\xcd\\x14M%\\x17e\\x8c\\x05nsp14\\x94]\\x94(M&\\x17M4\\x19e\\x8c\\x05nsp15\\x94]\\x94(M5\\x19M\\x8e\\x1ae\\x8c\\x05nsp16\\x94]\\x94(M\\x8f\\x1aM\\xb8\\x1beu\\x8c\\rgene_overlaps\\x94}\\x94(\\x8c\\x07exclude\\x94]\\x94(]\\x94(\\x8c\\x05ORF1a\\x94\\x8c\\x06ORF1ab\\x94e]\\x94(\\x8c\\x05ORF7a\\x94\\x8c\\x05ORF7b\\x94ee\\x8c\\x06retain\\x94]\\x94]\\x94(\\x8c\\x01N\\x94\\x8c\\x05ORF9b\\x94eau\\x8c\\x13fitness_pseudocount\\x94G?\\xe0\\x00\\x00\\x00\\x00\\x00\\x00\\x8c\\x12min_expected_count\\x94K\\n\\x8c\\x14clade_corr_min_count\\x94J \\xa1\\x07\\x00\\x8c\\x19aa_fitness_init_ref_clade\\x94\\x8c\\x0319A\\x94\\x8c!aa_fitness_heatmap_minimal_domain\\x94]\\x94(J\\xfa\\xff\\xff\\xffK\\x02e\\x8c\\x0eclade_synonyms\\x94}\\x94(\\x8c\\x0319A\\x94\\x8c\\x01B\\x94\\x8c\\x0320A\\x94\\x8c\\x03B.1\\x94\\x8c\\x0320B\\x94\\x8c\\x05B.1.1\\x94\\x8c\\x0320C\\x94\\x8c\\x07B.1.367\\x94\\x8c\\x0320E\\x94\\x8c\\x07B.1.177\\x94\\x8c\\x0320F\\x94\\x8c\\x03D.2\\x94\\x8c\\x0320G\\x94\\x8c\\x05B.1.2\\x94\\x8c\\x0320I\\x94\\x8c\\x05Alpha\\x94\\x8c\\x0320J\\x94\\x8c\\x05Gamma\\x94\\x8c\\x0321C\\x94\\x8c\\x07Epsilon\\x94\\x8c\\x0321F\\x94\\x8c\\x04Iota\\x94\\x8c\\x0321I\\x94\\x8c\\x05Delta\\x94\\x8c\\x0321J\\x94\\x8c\\x05Delta\\x94\\x8c\\x0321K\\x94\\x8c\\x0cOmicron BA.1\\x94\\x8c\\x0321L\\x94\\x8c\\x0cOmicron BA.2\\x94\\x8c\\x0322A\\x94\\x8c\\x0cOmicron BA.4\\x94\\x8c\\x0322B\\x94\\x8c\\x0cOmicron BA.5\\x94\\x8c\\x0322C\\x94\\x8c\\x11Omicron BA.2.12.1\\x94\\x8c\\x0322D\\x94\\x8c\\x0fOmicron BA.2.75\\x94\\x8c\\x0322E\\x94\\x8c\\x0cOmicron BQ.1\\x94u\\x8c\\x0cdms_datasets\\x94}\\x94(\\x8c\\x13dadonaite_ba1_spike\\x94}\\x94(\\x8c\\x05study\\x94\\x8c;https://www.biorxiv.org/content/10.1101/2022.10.13.512056v1\\x94\\x8c\\x0bdescription\\x94\\x8c\\x1dspike (Dadonaite et al, 2022)\\x94\\x8c\\x04gene\\x94\\x8c\\x01S\\x94\\x8c\\x03url\\x94\\x8c\\x8bhttps://raw.githubusercontent.com/dms-vep/SARS-CoV-2_Omicron_BA.1_spike_DMS_mAbs/main/results/muteffects_functional/muteffects_observed.csv\\x94\\x8c\\x0bfilter_cols\\x94}\\x94\\x8c\\ntimes_seen\\x94K\\x03su\\x8c\\tstarr_rbd\\x94}\\x94(\\x8c\\x05study\\x94\\x8cOhttps://journals.plos.org/plospathogens/article?id=10.1371/journal.ppat.1010951\\x94\\x8c\\x0bdescription\\x94\\x8c\\x17RBD (Starr et al, 2022)\\x94\\x8c\\x04gene\\x94j\\x1e\\x01\\x00\\x00\\x8c\\x03url\\x94\\x8c\\xadhttps://media.githubusercontent.com/media/jbloomlab/SARS-CoV-2-RBD_DMS_Omicron/f76ba2b2bec18ede9fa9da18c9ccc52389b1ba3a/results/final_variant_scores/final_variant_scores.csv\\x94u\\x8c\\x0ciketani_mpro\\x94}\\x94(\\x8c\\x05study\\x94\\x8cChttps://www.sciencedirect.com/science/article/pii/S1931312822004024\\x94\\x8c\\x0bdescription\\x94\\x8c\\x1aMpro (Iketani et al, 2022)\\x94\\x8c\\x04gene\\x94\\x8c\\x0bnsp5 (Mpro)\\x94\\x8c\\x03url\\x94\\x8c\\xachttps://raw.githubusercontent.com/alejandrochavezlab/3CL_protease_DMS/dc802ee3ea9d43005ca97a28af078767fd66777c/outputs/results/normalized_to_wt_and_stop/activity_scores.csv\\x94\\x8c\\x06wt_seq\\x94\\x8c\\x0cdata/Mpro.fa\\x94u\\x8c\\x0fflynn_mpro_2022\\x94}\\x94(\\x8c\\x05study\\x94\\x8c(https://elifesciences.org/articles/77433\\x94\\x8c\\x0bdescription\\x94\\x8c\\x18Mpro (Flynn et al, 2022)\\x94\\x8c\\x04gene\\x94\\x8c\\x0bnsp5 (Mpro)\\x94\\x8c\\x03url\\x94\\x8cKhttps://cdn.elifesciences.org/articles/77433/elife-77433-fig2-data1-v2.xlsx\\x94\\x8c\\x06wt_seq\\x94\\x8c\\x0cdata/Mpro.fa\\x94u\\x8c\\x0fflynn_mpro_2023\\x94}\\x94(\\x8c\\x05study\\x94\\x8c)https://doi.org/10.1101/2023.03.02.530652\\x94\\x8c\\x0bdescription\\x94\\x8c\\x18Mpro (Flynn et al, 2023)\\x94\\x8c\\x04gene\\x94\\x8c\\x0bnsp5 (Mpro)\\x94\\x8c\\x03url\\x94\\x8cahttps://www.biorxiv.org/content/biorxiv/early/2023/03/02/2023.03.02.530652/DC1/embed/media-1.xlsx\\x94uu\\x8c\\rneher_fitness\\x94\\x8c\\x19data/Neher_aa_fitness.csv\\x94\\x8c\\x14clade_fixed_muts_ref\\x94\\x8c\\x0319A\\x94\\x8c\\x14terminal_pseudocount\\x94G?\\xe0\\x00\\x00\\x00\\x00\\x00\\x00\\x8c\\x19terminal_min_actual_count\\x94K\\x05\\x8c\\x12positive_selection\\x94\\x8c^https://raw.githubusercontent.com/spond/SARS-CoV-2-variation/master/windowed-sites-fel-all.csv\\x94\\x8c\\x08docs_url\\x94\\x8c-https://jbloomlab.github.io/SARS2-mut-fitness\\x94\\x8c\\x15docs_plot_annotations\\x94\\x8c\\x1fdata/docs_plot_annotations.yaml\\x94u\\x8c\\x04rule\\x94\\x8c\\x1aanalyze_positive_selection\\x94\\x8c\\x0fbench_iteration\\x94N\\x8c\\tscriptdir\\x94\\x8cP/fh/fast/bloom_j/computational_notebooks/jbloom/2023/SARS2-mut-fitness/notebooks\\x94ub.'); from snakemake.logging import logger; logger.printshellcmds = False; import os; os.chdir(r'/fh/fast/bloom_j/computational_notebooks/jbloom/2023/SARS2-mut-fitness');\n",
    "######## snakemake preamble end #########\n",
    "snakemake.params.min_expected_count = 10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "cf8d9563-72f3-47da-a1c1-2934714ce0bd",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-26T23:46:51.901050Z",
     "iopub.status.busy": "2023-05-26T23:46:51.900702Z",
     "iopub.status.idle": "2023-05-26T23:46:51.905453Z",
     "shell.execute_reply": "2023-05-26T23:46:51.904508Z",
     "shell.execute_reply.started": "2023-05-26T23:46:51.901022Z"
    }
   },
   "outputs": [],
   "source": [
    "# get variables from `snakemake`\n",
    "positive_selection_csv = snakemake.input.positive_selection\n",
    "aa_fitness_csv = snakemake.input.aa_fitness\n",
    "corr_html = snakemake.output.corr_html\n",
    "min_expected_count = snakemake.params.min_expected_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "321f4ab5-e53a-4c7e-be22-3dc552bfa495",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-26T23:46:51.909390Z",
     "iopub.status.busy": "2023-05-26T23:46:51.908928Z",
     "iopub.status.idle": "2023-05-26T23:46:52.970981Z",
     "shell.execute_reply": "2023-05-26T23:46:52.970073Z",
     "shell.execute_reply.started": "2023-05-26T23:46:51.909364Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "import altair as alt\n",
    "\n",
    "import numpy\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "import scipy.stats.mstats"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02e207bc-caeb-4cb9-9cf5-f09b67729129",
   "metadata": {},
   "source": [
    "Get dN/dS values averaging the dN (`beta`) and dS (`alpha`) values over timeframes first, and clipping the dN/dS values at 0.05 and 20:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "812a4e92-7a53-4611-b2aa-c6c14cd4ca87",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-27T00:00:47.686479Z",
     "iopub.status.busy": "2023-05-27T00:00:47.685814Z",
     "iopub.status.idle": "2023-05-27T00:00:47.960996Z",
     "shell.execute_reply": "2023-05-27T00:00:47.960351Z",
     "shell.execute_reply.started": "2023-05-27T00:00:47.686434Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>site</th>\n",
       "      <th>dn</th>\n",
       "      <th>log_dnds</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>nsp5 (Mpro)</td>\n",
       "      <td>1</td>\n",
       "      <td>1.108253</td>\n",
       "      <td>1.177224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>nsp5 (Mpro)</td>\n",
       "      <td>2</td>\n",
       "      <td>1.369948</td>\n",
       "      <td>0.887329</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>nsp5 (Mpro)</td>\n",
       "      <td>3</td>\n",
       "      <td>0.340120</td>\n",
       "      <td>-0.978962</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>nsp5 (Mpro)</td>\n",
       "      <td>4</td>\n",
       "      <td>1.140580</td>\n",
       "      <td>-0.488341</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>nsp5 (Mpro)</td>\n",
       "      <td>5</td>\n",
       "      <td>2.353438</td>\n",
       "      <td>1.495519</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9408</th>\n",
       "      <td>nsp9</td>\n",
       "      <td>109</td>\n",
       "      <td>3.373902</td>\n",
       "      <td>2.995732</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9409</th>\n",
       "      <td>nsp9</td>\n",
       "      <td>110</td>\n",
       "      <td>0.339979</td>\n",
       "      <td>-1.458756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9410</th>\n",
       "      <td>nsp9</td>\n",
       "      <td>111</td>\n",
       "      <td>0.684415</td>\n",
       "      <td>0.711573</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9411</th>\n",
       "      <td>nsp9</td>\n",
       "      <td>112</td>\n",
       "      <td>0.609013</td>\n",
       "      <td>-2.394275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9412</th>\n",
       "      <td>nsp9</td>\n",
       "      <td>113</td>\n",
       "      <td>2.185392</td>\n",
       "      <td>-0.023591</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9413 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene  site        dn  log_dnds\n",
       "0     nsp5 (Mpro)     1  1.108253  1.177224\n",
       "1     nsp5 (Mpro)     2  1.369948  0.887329\n",
       "2     nsp5 (Mpro)     3  0.340120 -0.978962\n",
       "3     nsp5 (Mpro)     4  1.140580 -0.488341\n",
       "4     nsp5 (Mpro)     5  2.353438  1.495519\n",
       "...           ...   ...       ...       ...\n",
       "9408         nsp9   109  3.373902  2.995732\n",
       "9409         nsp9   110  0.339979 -1.458756\n",
       "9410         nsp9   111  0.684415  0.711573\n",
       "9411         nsp9   112  0.609013 -2.394275\n",
       "9412         nsp9   113  2.185392 -0.023591\n",
       "\n",
       "[9413 rows x 4 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_map = {\n",
    "    \"3C\": \"nsp5 (Mpro)\",\n",
    "    \"RdRp\": \"nsp12 (RdRp)\",\n",
    "    \"endornase\": \"nsp15\",\n",
    "    \"exonuclease\": \"nsp14\",\n",
    "    \"helicase\": \"nsp13\",\n",
    "    \"leader\": \"nsp1\",\n",
    "    \"methyltransferase\": \"nsp16\",\n",
    "}\n",
    "\n",
    "positive_selection = (\n",
    "    pd.read_csv(positive_selection_csv)\n",
    "    .query(\"(alpha != 0) or (beta != 0)\")\n",
    "    .groupby([\"gene\", \"site\"], as_index=False)\n",
    "    .aggregate({\"alpha\": \"mean\", \"beta\": \"mean\"})\n",
    "    .assign(\n",
    "        dnds=lambda x: numpy.clip(x[\"beta\"] / x[\"alpha\"], a_min=0.05, a_max=20),\n",
    "        log_dnds=lambda x: numpy.log(x[\"dnds\"]),\n",
    "        gene=lambda x: x[\"gene\"].map(lambda g: gene_map[g] if g in gene_map else g),\n",
    "    )\n",
    "    .rename(columns={\"beta\": \"dn\"})\n",
    "    .drop(columns=[\"alpha\", \"dnds\"])\n",
    ")\n",
    "\n",
    "positive_selection"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57f59b47-80e9-469f-9bae-32d0af6bd264",
   "metadata": {},
   "source": [
    "Read amino-acid fitnesses, and for each site:\n",
    " - the mean fitness of mutations\n",
    " - the number of effective amino-acids when amino acids are assigned a probability weight of $e^{fitness}$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "8bd92aad-fb86-4291-ae86-ac2aef086462",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-26T23:59:16.576599Z",
     "iopub.status.busy": "2023-05-26T23:59:16.575996Z",
     "iopub.status.idle": "2023-05-26T23:59:18.441875Z",
     "shell.execute_reply": "2023-05-26T23:59:18.441209Z",
     "shell.execute_reply.started": "2023-05-26T23:59:16.576542Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>site</th>\n",
       "      <th>mean_fitness</th>\n",
       "      <th>entropy</th>\n",
       "      <th>n_effective</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>E</td>\n",
       "      <td>1</td>\n",
       "      <td>-4.044917</td>\n",
       "      <td>0.309668</td>\n",
       "      <td>1.362972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>E</td>\n",
       "      <td>2</td>\n",
       "      <td>-2.786233</td>\n",
       "      <td>0.774795</td>\n",
       "      <td>2.170148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>E</td>\n",
       "      <td>3</td>\n",
       "      <td>-1.758716</td>\n",
       "      <td>1.171519</td>\n",
       "      <td>3.226890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>E</td>\n",
       "      <td>4</td>\n",
       "      <td>-1.636612</td>\n",
       "      <td>1.097318</td>\n",
       "      <td>2.996120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>E</td>\n",
       "      <td>5</td>\n",
       "      <td>-1.096513</td>\n",
       "      <td>1.234616</td>\n",
       "      <td>3.437060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9736</th>\n",
       "      <td>nsp9</td>\n",
       "      <td>109</td>\n",
       "      <td>-2.159867</td>\n",
       "      <td>0.743339</td>\n",
       "      <td>2.102945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9737</th>\n",
       "      <td>nsp9</td>\n",
       "      <td>110</td>\n",
       "      <td>-2.573740</td>\n",
       "      <td>0.744499</td>\n",
       "      <td>2.105387</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9738</th>\n",
       "      <td>nsp9</td>\n",
       "      <td>111</td>\n",
       "      <td>-3.732329</td>\n",
       "      <td>0.571722</td>\n",
       "      <td>1.771315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9739</th>\n",
       "      <td>nsp9</td>\n",
       "      <td>112</td>\n",
       "      <td>-2.238080</td>\n",
       "      <td>0.949155</td>\n",
       "      <td>2.583527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9740</th>\n",
       "      <td>nsp9</td>\n",
       "      <td>113</td>\n",
       "      <td>-3.126759</td>\n",
       "      <td>1.052402</td>\n",
       "      <td>2.864523</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9741 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene  site  mean_fitness   entropy  n_effective\n",
       "0        E     1     -4.044917  0.309668     1.362972\n",
       "1        E     2     -2.786233  0.774795     2.170148\n",
       "2        E     3     -1.758716  1.171519     3.226890\n",
       "3        E     4     -1.636612  1.097318     2.996120\n",
       "4        E     5     -1.096513  1.234616     3.437060\n",
       "...    ...   ...           ...       ...          ...\n",
       "9736  nsp9   109     -2.159867  0.743339     2.102945\n",
       "9737  nsp9   110     -2.573740  0.744499     2.105387\n",
       "9738  nsp9   111     -3.732329  0.571722     1.771315\n",
       "9739  nsp9   112     -2.238080  0.949155     2.583527\n",
       "9740  nsp9   113     -3.126759  1.052402     2.864523\n",
       "\n",
       "[9741 rows x 5 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aa_fitness = (\n",
    "    pd.read_csv(aa_fitness_csv)\n",
    "    .query(\"gene not in ['ORF1a', 'ORF1ab']\")\n",
    "    .query(\"expected_count >= @min_expected_count\")\n",
    "    .rename(columns={\"aa_site\": \"site\"})\n",
    "    .assign(\n",
    "        p_aa=lambda x: numpy.exp(x[\"fitness\"]),\n",
    "        p=lambda x: x[\"p_aa\"] / x.groupby([\"gene\", \"site\"])[\"p_aa\"].transform(\"sum\"),\n",
    "    )\n",
    "    .groupby([\"gene\", \"site\"], as_index=False)\n",
    "    .aggregate(\n",
    "        mean_fitness=pd.NamedAgg(\"fitness\", \"mean\"),\n",
    "        entropy=pd.NamedAgg(\"p\", lambda p: -(p * numpy.log(p)).sum()),\n",
    "    )\n",
    "    .assign(n_effective=lambda x: numpy.exp(x[\"entropy\"]))\n",
    ")\n",
    "\n",
    "assert not (set(positive_selection[\"gene\"].unique()) - set(aa_fitness[\"gene\"].unique()))\n",
    "\n",
    "aa_fitness"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c55ad63-8ea2-4200-b1ab-5377b02d1355",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-26T23:29:30.097371Z",
     "iopub.status.busy": "2023-05-26T23:29:30.096723Z",
     "iopub.status.idle": "2023-05-26T23:29:30.112190Z",
     "shell.execute_reply": "2023-05-26T23:29:30.111483Z",
     "shell.execute_reply.started": "2023-05-26T23:29:30.097327Z"
    },
    "tags": []
   },
   "source": [
    "Merge amino-acid fitness estimates aggregated at site level with dN/dS values:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "dedc5481-5693-47d5-9073-64ada95049e4",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-27T00:01:21.295416Z",
     "iopub.status.busy": "2023-05-27T00:01:21.294868Z",
     "iopub.status.idle": "2023-05-27T00:01:21.344786Z",
     "shell.execute_reply": "2023-05-27T00:01:21.343439Z",
     "shell.execute_reply.started": "2023-05-27T00:01:21.295373Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site</th>\n",
       "      <th>mean_fitness</th>\n",
       "      <th>entropy</th>\n",
       "      <th>n_effective</th>\n",
       "      <th>dn</th>\n",
       "      <th>log_dnds</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>site</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.086640</td>\n",
       "      <td>-0.060732</td>\n",
       "      <td>-0.070155</td>\n",
       "      <td>-0.058933</td>\n",
       "      <td>-0.117274</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean_fitness</th>\n",
       "      <td>-0.086640</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.830976</td>\n",
       "      <td>0.807706</td>\n",
       "      <td>0.342034</td>\n",
       "      <td>0.431270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>entropy</th>\n",
       "      <td>-0.060732</td>\n",
       "      <td>0.830976</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.979358</td>\n",
       "      <td>0.257510</td>\n",
       "      <td>0.389792</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>n_effective</th>\n",
       "      <td>-0.070155</td>\n",
       "      <td>0.807706</td>\n",
       "      <td>0.979358</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.272298</td>\n",
       "      <td>0.390753</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dn</th>\n",
       "      <td>-0.058933</td>\n",
       "      <td>0.342034</td>\n",
       "      <td>0.257510</td>\n",
       "      <td>0.272298</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.552796</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>log_dnds</th>\n",
       "      <td>-0.117274</td>\n",
       "      <td>0.431270</td>\n",
       "      <td>0.389792</td>\n",
       "      <td>0.390753</td>\n",
       "      <td>0.552796</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  site  mean_fitness   entropy  n_effective        dn  \\\n",
       "site          1.000000     -0.086640 -0.060732    -0.070155 -0.058933   \n",
       "mean_fitness -0.086640      1.000000  0.830976     0.807706  0.342034   \n",
       "entropy      -0.060732      0.830976  1.000000     0.979358  0.257510   \n",
       "n_effective  -0.070155      0.807706  0.979358     1.000000  0.272298   \n",
       "dn           -0.058933      0.342034  0.257510     0.272298  1.000000   \n",
       "log_dnds     -0.117274      0.431270  0.389792     0.390753  0.552796   \n",
       "\n",
       "              log_dnds  \n",
       "site         -0.117274  \n",
       "mean_fitness  0.431270  \n",
       "entropy       0.389792  \n",
       "n_effective   0.390753  \n",
       "dn            0.552796  \n",
       "log_dnds      1.000000  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>dn</th>\n",
       "      <th>log_dnds</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ORF6</td>\n",
       "      <td>0.299415</td>\n",
       "      <td>-0.041830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ORF7a</td>\n",
       "      <td>0.295115</td>\n",
       "      <td>0.028967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ORF8</td>\n",
       "      <td>0.330476</td>\n",
       "      <td>0.090213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>nsp1</td>\n",
       "      <td>0.155482</td>\n",
       "      <td>0.120638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>E</td>\n",
       "      <td>0.436386</td>\n",
       "      <td>0.227921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ORF3a</td>\n",
       "      <td>0.356401</td>\n",
       "      <td>0.243789</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>nsp8</td>\n",
       "      <td>0.100931</td>\n",
       "      <td>0.283252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>nsp7</td>\n",
       "      <td>0.271862</td>\n",
       "      <td>0.306922</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>nsp9</td>\n",
       "      <td>0.189855</td>\n",
       "      <td>0.311464</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>M</td>\n",
       "      <td>0.492178</td>\n",
       "      <td>0.318985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>nsp15</td>\n",
       "      <td>0.388229</td>\n",
       "      <td>0.343119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>nsp6</td>\n",
       "      <td>0.361858</td>\n",
       "      <td>0.391154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>nsp10</td>\n",
       "      <td>0.144116</td>\n",
       "      <td>0.403038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>nsp14</td>\n",
       "      <td>0.333687</td>\n",
       "      <td>0.411092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>nsp2</td>\n",
       "      <td>0.286734</td>\n",
       "      <td>0.421929</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>nsp3</td>\n",
       "      <td>0.331995</td>\n",
       "      <td>0.429013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>nsp13</td>\n",
       "      <td>0.346298</td>\n",
       "      <td>0.438740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>nsp16</td>\n",
       "      <td>0.421093</td>\n",
       "      <td>0.442902</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>nsp4</td>\n",
       "      <td>0.379908</td>\n",
       "      <td>0.443482</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>nsp5 (Mpro)</td>\n",
       "      <td>0.357783</td>\n",
       "      <td>0.453265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>nsp12 (RdRp)</td>\n",
       "      <td>0.414892</td>\n",
       "      <td>0.479159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>S</td>\n",
       "      <td>0.404380</td>\n",
       "      <td>0.505507</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>N</td>\n",
       "      <td>0.553152</td>\n",
       "      <td>0.578905</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            gene        dn  log_dnds\n",
       "12          ORF6  0.299415 -0.041830\n",
       "15         ORF7a  0.295115  0.028967\n",
       "18          ORF8  0.330476  0.090213\n",
       "24          nsp1  0.155482  0.120638\n",
       "0              E  0.436386  0.227921\n",
       "9          ORF3a  0.356401  0.243789\n",
       "63          nsp8  0.100931  0.283252\n",
       "60          nsp7  0.271862  0.306922\n",
       "66          nsp9  0.189855  0.311464\n",
       "3              M  0.492178  0.318985\n",
       "39         nsp15  0.388229  0.343119\n",
       "57          nsp6  0.361858  0.391154\n",
       "27         nsp10  0.144116  0.403038\n",
       "36         nsp14  0.333687  0.411092\n",
       "45          nsp2  0.286734  0.421929\n",
       "48          nsp3  0.331995  0.429013\n",
       "33         nsp13  0.346298  0.438740\n",
       "42         nsp16  0.421093  0.442902\n",
       "51          nsp4  0.379908  0.443482\n",
       "54   nsp5 (Mpro)  0.357783  0.453265\n",
       "30  nsp12 (RdRp)  0.414892  0.479159\n",
       "21             S  0.404380  0.505507\n",
       "6              N  0.553152  0.578905"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = (\n",
    "    aa_fitness\n",
    "    .merge(positive_selection, on=[\"gene\", \"site\"], validate=\"one_to_one\")\n",
    ")\n",
    "\n",
    "display(df.corr())\n",
    "\n",
    "(\n",
    "    df\n",
    "    .groupby(\"gene\")\n",
    "    [[\"mean_fitness\", \"dn\", \"log_dnds\"]]\n",
    "    .corr()\n",
    "    .reset_index()\n",
    "    .query(\"level_1 == 'mean_fitness'\")\n",
    "    [[\"gene\", \"dn\", \"log_dnds\"]]\n",
    "    .sort_values(\"log_dnds\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "11b31851-3370-4a7f-ae64-0823a7e7e49f",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-26T23:46:53.420113Z",
     "iopub.status.busy": "2023-05-26T23:46:53.419919Z",
     "iopub.status.idle": "2023-05-26T23:46:53.426977Z",
     "shell.execute_reply": "2023-05-26T23:46:53.426326Z",
     "shell.execute_reply.started": "2023-05-26T23:46:53.420093Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.001   -2.995732\n",
       "0.010   -2.995732\n",
       "0.100   -2.995732\n",
       "0.500   -0.780895\n",
       "0.900    2.755768\n",
       "0.990    2.995732\n",
       "Name: log_dnds, dtype: float64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"log_dnds\"].quantile([0.001, 0.01, 0.1, 0.5, 0.9, 0.99])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "163bb53d-d6b3-47ba-ab91-38961819dd50",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
