{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f6f4d8e5-186a-406e-bc5d-b2fc30bf979b",
   "metadata": {},
   "source": [
    "# Analyze and plot the amino-acid fitness values\n",
    "\n",
    "Get variables from `snakemake`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37e2e1c4-fc82-4bdc-aa93-3cc6d10e1e84",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "if \"snakemake\" in globals() or \"snakemake\" in locals():\n",
    "    # get variables from snakemake\n",
    "    min_expected_count = snakemake.params.min_expected_count\n",
    "    clade_corr_min_count = snakemake.params.clade_corr_min_count\n",
    "    init_ref_clade = snakemake.params.init_ref_clade\n",
    "    clade_synonyms = snakemake.params.clade_synonyms\n",
    "    heatmap_minimal_domain = snakemake.params.heatmap_minimal_domain\n",
    "    orf1ab_to_nsps = snakemake.params.orf1ab_to_nsps\n",
    "    aamut_all_csv = snakemake.input.aamut_all\n",
    "    aamut_by_subset_csv = snakemake.input.aamut_by_subset\n",
    "    aamut_by_clade_csv = snakemake.input.aamut_by_clade\n",
    "    ref_coding_sites_csv = snakemake.input.ref_coding_sites\n",
    "    aafitness_csv = snakemake.input.aafitness\n",
    "    clade_founder_nts_csv = snakemake.input.clade_founder_nts\n",
    "    outdir = snakemake.output.outdir\n",
    "else:\n",
    "    # manually defined if running interactively for debugging\n",
    "    import yaml\n",
    "\n",
    "    with open(\"../config.yaml\") as f:\n",
    "        config = yaml.safe_load(f)\n",
    "\n",
    "    min_expected_count = config[\"min_expected_count\"]\n",
    "    clade_corr_min_count = config[\"clade_corr_min_count\"]\n",
    "    init_ref_clade = config[\"aa_fitness_init_ref_clade\"]\n",
    "    clade_synonyms = config[\"clade_synonyms\"]\n",
    "    heatmap_minimal_domain = config[\"aa_fitness_heatmap_minimal_domain\"]\n",
    "    orf1ab_to_nsps = config[\"orf1ab_to_nsps\"]\n",
    "    aamut_all_csv = \"../results/aa_fitness/aamut_fitness_all.csv\"\n",
    "    aamut_by_subset_csv = \"../results/aa_fitness/aamut_fitness_by_subset.csv\"\n",
    "    aamut_by_clade_csv = \"../results/aa_fitness/aamut_fitness_by_clade.csv\"\n",
    "    ref_coding_sites_csv = \"../results/ref/coding_sites.csv\"\n",
    "    aafitness_csv = \"../results/aa_fitness/aa_fitness.csv\"\n",
    "    clade_founder_nts_csv = \"../results/clade_founder_nts/clade_founder_nts.csv\"\n",
    "    outdir = \"../results/aa_fitness/plots\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "690c0cfe-d382-4531-af5f-3a6b9145d045",
   "metadata": {},
   "source": [
    "Import Python modules:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c7e5dc6-ac49-427d-b48b-1687e252b911",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import itertools\n",
    "import os\n",
    "\n",
    "import altair as alt\n",
    "\n",
    "import Bio.Seq\n",
    "\n",
    "import numpy\n",
    "\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ba22a6b-00ee-4675-b13c-365c858f5edc",
   "metadata": {},
   "source": [
    "Some settings:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ca2271f-6ef8-401a-868a-728a433ee9b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "_ = alt.data_transformers.disable_max_rows()\n",
    "\n",
    "os.makedirs(outdir, exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "488374b7-2eb2-4905-b0d9-62e36b683066",
   "metadata": {},
   "source": [
    "Read input data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1021825-2777-4563-842b-1d8f78cb293c",
   "metadata": {},
   "outputs": [],
   "source": [
    "aamut_all = pd.read_csv(aamut_all_csv)\n",
    "aamut_by_subset = pd.read_csv(aamut_by_subset_csv)\n",
    "aamut_by_clade = pd.read_csv(aamut_by_clade_csv)\n",
    "aafitness = pd.read_csv(aafitness_csv)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "590b3de8-abee-43f1-9f15-c3f07df04755",
   "metadata": {},
   "source": [
    "Define function that gives clade labels:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "116a81d5-01a3-4ae2-bf27-86322c604b84",
   "metadata": {},
   "outputs": [],
   "source": [
    "def clade_label(clade):\n",
    "    if clade in clade_synonyms:\n",
    "        return f\"{clade} ({clade_synonyms[clade]})\"\n",
    "    else:\n",
    "        return clade"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "89803bf7-aa82-4ddf-8e2b-0d9c02239f1e",
   "metadata": {},
   "source": [
    "## Correlation in fitness among subsets and clades\n",
    "Plot correlation in mutation fitness effect ($\\Delta f_{xy}$ values) among subsets (geographic locations) and clades with large numbers of counts.\n",
    "We just plot ORF1ab and not its constituent nsps to avoid double counting.\n",
    "We only plot amino-acid mutations (not synonymous ones).\n",
    "We do not include the \"all\" subset as that contains all the other subsets:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b50de36d-0d26-44c9-9518-5e285a05d091",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def plot_corr_scatters(corr_df_tidy):\n",
    "    \"Plot set of correlation scatters.\"\"\"\n",
    "    \n",
    "    subsets = corr_df_tidy[\"subset\"].unique()\n",
    "    genes = corr_df_tidy[\"gene\"].unique()\n",
    "    \n",
    "    corr_df_wide = pd.merge(\n",
    "        *[\n",
    "            corr_df_tidy\n",
    "            .pivot_table(\n",
    "                index=[\"gene\", \"aa_mutation\"],\n",
    "                values=prop,\n",
    "                columns=\"subset\",\n",
    "            )\n",
    "            .reset_index()\n",
    "            .rename(columns={subset: f\"{prop} {subset}\" for subset in subsets})\n",
    "            for prop in [\"delta_fitness\", \"expected_count\"]\n",
    "        ]\n",
    "    )\n",
    "    \n",
    "    delta_fitness_min = corr_df_tidy[\"delta_fitness\"].min()\n",
    "    delta_fitness_max = corr_df_tidy[\"delta_fitness\"].max()\n",
    "\n",
    "    gene_selection = alt.selection_multi(\n",
    "        fields=[\"gene\"], bind=\"legend\",\n",
    "    )\n",
    "\n",
    "    expected_count_selection = alt.selection_single(\n",
    "        bind=alt.binding_range(\n",
    "            min=1,\n",
    "            max=min(5 * min_expected_count, corr_df_tidy[\"expected_count\"].quantile(0.9)),\n",
    "            step=1,\n",
    "            name=\"minimum expected count\",\n",
    "        ),\n",
    "        fields=[\"cutoff\"],\n",
    "        init={\"cutoff\": min_expected_count},\n",
    "    )\n",
    "    \n",
    "    highlight = alt.selection_single(\n",
    "        on=\"mouseover\",\n",
    "        fields=[\"gene\", \"aa_mutation\"],\n",
    "        empty=\"none\",\n",
    "    )\n",
    "    \n",
    "    corr_charts = []\n",
    "    base_chart = alt.Chart(corr_df_wide)\n",
    "    for subset1, subset2 in itertools.combinations(subsets, 2):\n",
    "        base = (\n",
    "            base_chart\n",
    "            .encode(\n",
    "                x=alt.X(\n",
    "                    f\"delta_fitness {subset1}\",\n",
    "                    title=f\"{subset1} fitness effect\",\n",
    "                    scale=alt.Scale(domain=(delta_fitness_min, delta_fitness_max), nice=False),\n",
    "                ),\n",
    "                y=alt.Y(\n",
    "                    f\"delta_fitness {subset2}\",\n",
    "                    title=f\"{subset2} fitness effect\",\n",
    "                    scale=alt.Scale(domain=(delta_fitness_min, delta_fitness_max), nice=False),\n",
    "                ),\n",
    "                tooltip=[\n",
    "                    \"gene\",\n",
    "                    \"aa_mutation\",\n",
    "                    alt.Tooltip(\n",
    "                        f\"delta_fitness {subset1}\", title=f\"{subset1} fitness effect\",\n",
    "                    ),\n",
    "                    alt.Tooltip(\n",
    "                        f\"delta_fitness {subset2}\", title=f\"{subset2} fitness effect\",\n",
    "                    ),\n",
    "                    f\"expected_count {subset1}\",\n",
    "                    f\"expected_count {subset2}\",\n",
    "                ],\n",
    "            )\n",
    "            .mark_circle(opacity=0.3)\n",
    "            .properties(width=200, height=200)\n",
    "            .transform_filter(gene_selection)\n",
    "            .transform_filter(\n",
    "                (alt.datum[f\"expected_count {subset1}\"] >= expected_count_selection[\"cutoff\"] - 1e-6)\n",
    "                & (alt.datum[f\"expected_count {subset2}\"] >= expected_count_selection[\"cutoff\"] - 1e-6)\n",
    "            )\n",
    "        )\n",
    "    \n",
    "        scatter = (\n",
    "            base\n",
    "            .encode(\n",
    "                color=alt.Color(\n",
    "                    \"gene\",\n",
    "                    scale=alt.Scale(\n",
    "                        domain=genes,\n",
    "                        range=[\"#5778a4\"] * len(genes),\n",
    "                    ),\n",
    "                    legend=alt.Legend(\n",
    "                        symbolOpacity=1,\n",
    "                        orient=\"bottom\",\n",
    "                        title=\"click / shift-click to select specific genes to show\",\n",
    "                        titleLimit=500,\n",
    "                        columns=6,\n",
    "                    ),\n",
    "                ),\n",
    "                size=alt.condition(highlight, alt.value(85), alt.value(30)),\n",
    "                opacity=alt.condition(highlight, alt.value(1), alt.value(0.3)),\n",
    "                strokeWidth=alt.condition(highlight, alt.value(1.5), alt.value(0)),\n",
    "            )\n",
    "            .mark_circle(stroke=\"black\")\n",
    "        )\n",
    "    \n",
    "        # regression line and correlation coefficient: https://stackoverflow.com/a/60239699\n",
    "        line = (\n",
    "            base\n",
    "            .transform_regression(\n",
    "                f\"delta_fitness {subset1}\",\n",
    "                f\"delta_fitness {subset2}\",\n",
    "                extent=(delta_fitness_min, delta_fitness_max),\n",
    "            )\n",
    "            .mark_line(color=\"orange\", clip=True)\n",
    "        )\n",
    "    \n",
    "        params_r = (\n",
    "            base\n",
    "            .transform_regression(\n",
    "                f\"delta_fitness {subset1}\",\n",
    "                f\"delta_fitness {subset2}\",\n",
    "                params=True,\n",
    "            )\n",
    "            .transform_calculate(\n",
    "                r=alt.expr.sqrt(alt.datum[\"rSquared\"]),\n",
    "                label='\"r = \" + format(datum.r, \".3f\")',\n",
    "            )\n",
    "            .mark_text(align=\"left\", color=\"orange\", fontWeight=\"bold\")\n",
    "            .encode(\n",
    "                x=alt.value(5),\n",
    "                y=alt.value(8),\n",
    "                text=alt.Text(\"label:N\"),\n",
    "            )\n",
    "        )\n",
    "        \n",
    "        # show number of points\n",
    "        params_n = (\n",
    "            base\n",
    "            .transform_filter(\n",
    "                (~alt.expr.isNaN(alt.datum[f\"delta_fitness {subset1}\"]))\n",
    "                & (~alt.expr.isNaN(alt.datum[f\"delta_fitness {subset2}\"]))\n",
    "            )\n",
    "            .transform_calculate(dummy=alt.datum[f\"delta_fitness {subset1}\"])\n",
    "            .transform_aggregate(n=\"valid(dummy)\")\n",
    "            .transform_calculate(label='\"n = \" + datum.n')\n",
    "            .mark_text(align=\"left\", color=\"orange\", fontWeight=\"bold\")\n",
    "            .encode(\n",
    "                x=alt.value(5),\n",
    "                y=alt.value(20),\n",
    "                text=alt.Text(\"label:N\"),\n",
    "            )\n",
    "        )\n",
    "    \n",
    "        chart = (\n",
    "            (scatter + line + params_r + params_n)\n",
    "            .add_selection(gene_selection)\n",
    "            .add_selection(expected_count_selection)\n",
    "            .add_selection(highlight)\n",
    "        )\n",
    "    \n",
    "        corr_charts.append(chart)\n",
    "    \n",
    "    ncols = 4\n",
    "    rows = []\n",
    "    for i in range(0, len(corr_charts), ncols):\n",
    "        rows.append(alt.hconcat(*corr_charts[i: i + ncols]))\n",
    "    corr_chart = alt.vconcat(*rows).configure_axis(grid=False)\n",
    "    return corr_chart\n",
    "    \n",
    "# correlations for subsets\n",
    "subset_corr_df = (\n",
    "    aamut_by_subset\n",
    "    .query(\"not subset_of_ORF1ab\")\n",
    "    .query(\"aa_mutation.str[0] != aa_mutation.str[-1]\")\n",
    "    [[\"subset\", \"gene\", \"aa_mutation\", \"expected_count\", \"delta_fitness\"]]\n",
    "    .query(\"subset != 'all'\")\n",
    ")\n",
    "subset_corr_chart = plot_corr_scatters(subset_corr_df)\n",
    "subset_corr_chart_file = os.path.join(outdir, \"subset_corr_chart.html\")\n",
    "print(f\"Saving to {subset_corr_chart_file}\\n\\n\")\n",
    "subset_corr_chart.save(subset_corr_chart_file)\n",
    "display(subset_corr_chart)\n",
    "\n",
    "# correlations for clades\n",
    "clade_corr_df = (\n",
    "    aamut_by_clade\n",
    "    .query(\"not subset_of_ORF1ab\")\n",
    "    .query(\"aa_mutation.str[0] != aa_mutation.str[-1]\")\n",
    "    [[\"clade\", \"gene\", \"aa_mutation\", \"expected_count\", \"delta_fitness\"]]\n",
    "    .assign(\n",
    "        clade_counts=lambda x: x.groupby(\"clade\")[\"expected_count\"].transform(\"sum\"),\n",
    "        clade=lambda x: x[\"clade\"].map(clade_label).str.replace(\".\", \"_\", regex=False),\n",
    "    )\n",
    "    .query(\"clade_counts >= @clade_corr_min_count\")\n",
    "    .drop(columns=\"clade_counts\")\n",
    "    .rename(columns={\"clade\": \"subset\"})\n",
    ")\n",
    "clade_corr_chart = plot_corr_scatters(clade_corr_df)\n",
    "clade_corr_chart_file = os.path.join(outdir, \"clade_corr_chart.html\")\n",
    "print(f\"Saving to {clade_corr_chart_file}\")\n",
    "clade_corr_chart.save(clade_corr_chart_file)\n",
    "display(clade_corr_chart)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "232e9057-f4a9-44cf-a093-b4cacc7fe3f1",
   "metadata": {},
   "source": [
    "## Correlations among different gene subsets\n",
    "Correlations of mutation effects among clades among different gene subsets using a few different cutoffs for expected counts.\n",
    "We compute the average correlation across all pairs of distinct clades with adequate counts:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dfd3efa2-75f8-455b-a9c8-40d158782f53",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "count_thresholds = [\n",
    "    min_expected_count // 2,\n",
    "    min_expected_count,\n",
    "    min_expected_count * 2,\n",
    "    min_expected_count * 4,\n",
    "]\n",
    "\n",
    "def get_subset_corrs(df):\n",
    "    \"\"\"Average correlations of pairs of distinct subsets for spike/non-spike mutations.\"\"\"\n",
    "    return (\n",
    "        pd.concat(\n",
    "            [\n",
    "            df.query(\"expected_count >= @n\").assign(count_threshold=n)\n",
    "            for n in count_thresholds\n",
    "            ]\n",
    "        )\n",
    "        .pivot_table(\n",
    "            index=[\"count_threshold\", \"gene\", \"aa_mutation\"],\n",
    "            columns=\"subset\",\n",
    "            values=\"delta_fitness\",\n",
    "        )\n",
    "        .reset_index()\n",
    "        .assign(\n",
    "            protein=lambda x: numpy.where(\n",
    "                x[\"gene\"] == \"S\", \"spike\", \"non-spike\",\n",
    "            ),\n",
    "        )\n",
    "        .groupby([\"count_threshold\", \"protein\"])\n",
    "        .corr()\n",
    "        .reset_index()\n",
    "        .melt(\n",
    "            id_vars=[\"count_threshold\", \"protein\", \"subset\"],\n",
    "            value_name=\"correlation\",\n",
    "            var_name=\"subset_2\",\n",
    "        )\n",
    "        .query(\"subset != subset_2\")\n",
    "    )\n",
    "\n",
    "subset_corrs = (\n",
    "    pd.concat(\n",
    "        [\n",
    "            get_subset_corrs(clade_corr_df).assign(subset_type=\"subset by clade\"),\n",
    "            get_subset_corrs(subset_corr_df).assign(subset_type=\"subset by geography\"),\n",
    "        ]\n",
    "    )\n",
    "    .groupby([\"subset_type\", \"count_threshold\", \"protein\"], as_index=False)\n",
    "    .aggregate({\"correlation\": \"mean\"})\n",
    ")\n",
    "\n",
    "subset_corrs_chart = (\n",
    "    alt.Chart(subset_corrs)\n",
    "    .encode(\n",
    "        x=alt.X(\"count_threshold:N\", title=\"min expected counts\"),\n",
    "        y=alt.Y(\n",
    "            \"correlation\",\n",
    "            title=\"average correlation among subsets\",\n",
    "            scale=alt.Scale(domain=(0, 1)),\n",
    "        ),\n",
    "        color=alt.Color(\"protein\", scale=alt.Scale(range=[\"#999999\", \"#D55E00\"])),\n",
    "        shape=alt.Shape(\"protein\"),\n",
    "        fill=alt.Fill(\"protein\", scale=alt.Scale(range=[\"#999999\", \"white\"])),\n",
    "        fillOpacity=alt.Fill(\"protein\", scale=alt.Scale(range=[0.6, 0])),\n",
    "        column=alt.Column(\n",
    "            \"subset_type\",\n",
    "            title=None,\n",
    "            spacing=20,\n",
    "            header=alt.Header(labelFontSize=11, labelFontWeight=\"bold\"),\n",
    "        ),\n",
    "        tooltip=[\n",
    "            alt.Tooltip(c, format=\".3f\") if subset_corrs[c].dtype == float else c\n",
    "            for c in subset_corrs.columns\n",
    "        ],\n",
    "    )\n",
    "    .mark_point(size=50)\n",
    "    .properties(height=190, width=alt.Step(25))\n",
    ")\n",
    "\n",
    "subset_corrs_chart_file = os.path.join(outdir, \"spike_nonspike_corrs.html\")\n",
    "print(f\"Saving to {subset_corrs_chart_file}\")\n",
    "subset_corrs_chart.save(subset_corrs_chart_file)\n",
    "\n",
    "subset_corrs_chart"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c46ac53-8142-47fb-bc30-0f51fb27ec70",
   "metadata": {},
   "source": [
    "Correlations among clades by amino-acid divergence.\n",
    "First get clade founder amino acids:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "701e16fc-8312-4e0e-af94-eca34a4ca7e5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# codon translation table\n",
    "codon_table = {\n",
    "    f\"{nt1}{nt2}{nt3}\": str(Bio.Seq.Seq(f\"{nt1}{nt2}{nt3}\").translate())\n",
    "    for nt1 in \"ACGT\" for nt2 in \"ACGT\" for nt3 in \"ACGT\"\n",
    "}\n",
    "\n",
    "# get clade founder amino-acids\n",
    "clade_founder_aas = (\n",
    "    pd.read_csv(clade_founder_nts_csv)\n",
    "    .query(\"gene != 'noncoding'\")\n",
    "    [[\"clade\", \"gene\", \"codon\", \"codon_site\"]]\n",
    "    .drop_duplicates()\n",
    "    .assign(\n",
    "        gene=lambda x: x[\"gene\"].str.split(\";\"),\n",
    "        codon=lambda x: x[\"codon\"].str.split(\";\"),\n",
    "        codon_site=lambda x: x[\"codon_site\"].str.split(\";\"),\n",
    "    )\n",
    "    .explode([\"gene\", \"codon\", \"codon_site\"])\n",
    "    .assign(\n",
    "        aa=lambda x: x[\"codon\"].map(codon_table),\n",
    "        codon_site=lambda x: x[\"codon_site\"].astype(int),\n",
    "        clade=lambda x: x[\"clade\"].map(clade_label),\n",
    "    )\n",
    "    .rename(columns={\"codon_site\": \"site\", \"aa\": \"amino acid\"})\n",
    "    .drop(columns=\"codon\")\n",
    "    .query(\"gene != 'ORF1a'\")  # this is just subset of ORF1ab\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a8310a3-f25f-4628-9a0d-60a1235a6f58",
   "metadata": {},
   "source": [
    "Now compute amino-acid divergence among spike and non-spike proteins:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "433efa9c-8f08-4423-a957-9323f17a0a92",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "aa_divergences = []\n",
    "for c1, c2 in itertools.combinations(clade_corr_df[\"subset\"].unique(), 2):\n",
    "    aa_divergences.append(\n",
    "        clade_founder_aas\n",
    "        .assign(\n",
    "            clade=lambda x: x[\"clade\"].str.replace(\".\", \"_\", regex=False),\n",
    "            protein=lambda x: numpy.where(\n",
    "                x[\"gene\"] == \"S\", \"spike\", \"non-spike\",\n",
    "            ),\n",
    "        )\n",
    "        .query(\"(clade == @c1) or (clade == @c2)\")\n",
    "        .pivot_table(\n",
    "            index=[\"protein\", \"site\"], columns=\"clade\", aggfunc=\"first\", values=\"amino acid\",\n",
    "        )\n",
    "        .reset_index()\n",
    "        .assign(ident=lambda x: x[c1] == x[c2])\n",
    "        .groupby(\"protein\", as_index=False)\n",
    "        .aggregate(\n",
    "            n=pd.NamedAgg(\"site\", \"count\"),\n",
    "            n_ident=pd.NamedAgg(\"ident\", \"sum\"),\n",
    "        )\n",
    "        .assign(\n",
    "            percent_divergence=lambda x: 100 * (1 - x[\"n_ident\"] / x[\"n\"]),\n",
    "            clade_1=c1,\n",
    "            clade_2=c2\n",
    "        )\n",
    "    )\n",
    "\n",
    "corrs_by_cladediv_chart_file = os.path.join(outdir, \"spike_nonspike_corrs_by_divergence.html\")\n",
    "\n",
    "if len(aa_divergences):\n",
    "    aa_divergences = pd.concat(\n",
    "        [\n",
    "            pd.concat(aa_divergences, ignore_index=True),\n",
    "            pd.concat(aa_divergences, ignore_index=True).rename(\n",
    "                columns={\"clade_1\": \"clade_2\", \"clade_2\": \"clade_1\"}\n",
    "            ),\n",
    "        ]\n",
    "    )\n",
    "\n",
    "    # Combine the amino-acid divergences with the clade-to-clade correlations:\n",
    "    corrs_by_cladediv = (\n",
    "        get_subset_corrs(clade_corr_df)\n",
    "        .rename(columns={\"subset\": \"clade_1\", \"subset_2\": \"clade_2\"})\n",
    "        .merge(\n",
    "            aa_divergences[[\"protein\", \"clade_1\", \"clade_2\", \"percent_divergence\"]],\n",
    "            on=[\"clade_1\", \"clade_2\", \"protein\"],\n",
    "            how=\"left\",\n",
    "            validate=\"many_to_one\",\n",
    "        )\n",
    "    )\n",
    "\n",
    "    assert corrs_by_cladediv.notnull().all().all()\n",
    "\n",
    "    # Now plot correlation in mutation effects by percent protein divergence:\n",
    "    count_threshold_selection = alt.selection_single(\n",
    "        fields=[\"count_threshold\"],\n",
    "        bind=alt.binding_radio(\n",
    "            options=corrs_by_cladediv[\"count_threshold\"].unique(),\n",
    "            name=\"minimum expected counts\",\n",
    "        ),\n",
    "        init={\"count_threshold\": min_expected_count},\n",
    "    )\n",
    "\n",
    "    corrs_by_cladediv_chart = (\n",
    "        alt.Chart(corrs_by_cladediv)\n",
    "        .encode(\n",
    "            x=alt.X(\n",
    "                \"percent_divergence\",\n",
    "                axis=alt.Axis(grid=False),\n",
    "                title=\"% protein divergence between clades\",\n",
    "            ),\n",
    "            y=alt.Y(\n",
    "                \"correlation\",\n",
    "                title=\"correlation between clade pairs\",\n",
    "                scale=alt.Scale(domain=(0, 1)),\n",
    "            ),\n",
    "            color=alt.Color(\"protein\", scale=alt.Scale(range=[\"#999999\", \"#D55E00\"])),\n",
    "            shape=alt.Shape(\"protein\"),\n",
    "            fill=alt.Fill(\"protein\", scale=alt.Scale(range=[\"#999999\", \"white\"])),\n",
    "            fillOpacity=alt.Fill(\"protein\", scale=alt.Scale(range=[0.6, 0])),\n",
    "            tooltip=[\n",
    "                alt.Tooltip(c, format=\".3f\") if corrs_by_cladediv[c].dtype == float else c\n",
    "                for c in corrs_by_cladediv.columns\n",
    "            ],\n",
    "        )\n",
    "        .mark_point(size=50)\n",
    "        .add_selection(count_threshold_selection)\n",
    "        .transform_filter(count_threshold_selection)\n",
    "        .properties(width=190, height=190)\n",
    "    )\n",
    "\n",
    "    print(f\"Saving to {corrs_by_cladediv_chart_file}\")\n",
    "    corrs_by_cladediv_chart.save(corrs_by_cladediv_chart_file)\n",
    "\n",
    "    display(corrs_by_cladediv_chart)\n",
    "\n",
    "else:\n",
    "    print(f\"Creating empty file {corrs_by_cladediv_chart_file}\")\n",
    "    (\n",
    "        alt.Chart(pd.DataFrame({\"x\": [0.5], \"y\": [0.5], \"text\": \"empty chart\"}))\n",
    "        .encode(x=\"x\", y=\"y\", text=\"text\")\n",
    "        .mark_text()\n",
    "        .save(corrs_by_cladediv_chart_file)\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f77aadb-a223-45e2-a957-b4eab6511088",
   "metadata": {},
   "source": [
    "## Histograms of mutation effects\n",
    "Histograms of mutation effects.\n",
    "We make a version both with ORF1ab labeled genes and nsp labeled genes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9caf39f6-0145-42b9-a98c-e86c11fa4733",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "mut_type_order = [\"synonymous\", \"nonsynonymous\", \"stop\"]\n",
    "mut_type_colors = [\"#009E73\", \"#56B4E9\", \"#E69F00\"]\n",
    "\n",
    "gene_selection = alt.selection_multi(\n",
    "    fields=[\"gene\"], bind=\"legend\",\n",
    ")\n",
    "\n",
    "expected_count_selection = alt.selection_single(\n",
    "    bind=alt.binding_range(\n",
    "        min=1,\n",
    "        max=min(5 * min_expected_count, aamut_all[\"expected_count\"].quantile(0.9)),\n",
    "        step=1,\n",
    "        name=\"minimum expected count\",\n",
    "    ),\n",
    "    fields=[\"cutoff\"],\n",
    "    init={\"cutoff\": min_expected_count},\n",
    ")\n",
    "\n",
    "namings = [\"nsp\"]  # can also be [\"nsp\", \"ORF1ab\"] to plot ORF1ab naming\n",
    "\n",
    "for orf1ab_nsp in namings:\n",
    "    query_str = (\n",
    "        \"not subset_of_ORF1ab\" if orf1ab_nsp == \"ORF1ab\"\n",
    "        else \"gene != 'ORF1ab'\"\n",
    "    )\n",
    "    \n",
    "    hist_df = (\n",
    "        aamut_all\n",
    "        .query(query_str)\n",
    "        .assign(\n",
    "            mut_type=lambda x: x[\"aa_mutation\"].map(\n",
    "                lambda m: (\n",
    "                    \"synonymous\" if m[0] == m[-1]\n",
    "                    else \"stop\" if m[-1] == \"*\" else \"nonsynonymous\"\n",
    "                )\n",
    "            )\n",
    "        )\n",
    "        [[\"gene\", \"expected_count\", \"delta_fitness\", \"mut_type\"]]\n",
    "    )\n",
    "    \n",
    "    delta_fitness_max = hist_df[\"delta_fitness\"].max()\n",
    "    delta_fitness_min = hist_df[\"delta_fitness\"].min()\n",
    "\n",
    "    hist_chart = (\n",
    "        alt.Chart(hist_df)\n",
    "        .encode(\n",
    "            x=alt.X(\n",
    "                \"delta_fitness\",\n",
    "                bin=alt.Bin(step=0.5),\n",
    "                scale=alt.Scale(domain=(delta_fitness_min, delta_fitness_max)),\n",
    "                axis=alt.Axis(\n",
    "                    values=[\n",
    "                        2 * t\n",
    "                        for t in range(\n",
    "                            int(delta_fitness_min // 2), int(delta_fitness_max // 2)\n",
    "                        )\n",
    "                    ]\n",
    "                ),\n",
    "                title=\"fitness effect of mutation\",\n",
    "            ),\n",
    "            y=alt.Y(\"count()\", title=\"number of mutations\"),\n",
    "            color=alt.Color(\n",
    "                \"mut_type\", scale=alt.Scale(range=mut_type_colors), legend=None, sort=mut_type_order,\n",
    "            ),\n",
    "            stroke=alt.Stroke(\n",
    "                \"mut_type\", scale=alt.Scale(range=mut_type_colors), legend=None, sort=mut_type_order,\n",
    "            ),\n",
    "            opacity=alt.Opacity(\n",
    "                \"gene\",\n",
    "                scale=alt.Scale(\n",
    "                    domain=hist_df[\"gene\"].unique(),\n",
    "                    range=[1] * hist_df[\"gene\"].nunique(),\n",
    "                ),\n",
    "                legend=alt.Legend(\n",
    "                    symbolFillColor=\"darkgray\",\n",
    "                    orient=\"bottom\",\n",
    "                    title=\"click / shift-click to select specific genes to show\",\n",
    "                    titleLimit=500,\n",
    "                    columns=13,\n",
    "                    padding=5,\n",
    "                ),\n",
    "            ),    \n",
    "            facet=alt.Facet(\n",
    "                \"mut_type\",\n",
    "                title=None,\n",
    "                columns=len(mut_type_order),\n",
    "                header=alt.Header(labelFontSize=12, labelFontWeight=\"bold\", labelPadding=2),\n",
    "                spacing=8,\n",
    "                sort=mut_type_order,\n",
    "            ),\n",
    "        )\n",
    "        .mark_bar(clip=True)\n",
    "        .transform_filter(gene_selection)\n",
    "        .transform_filter(\n",
    "            alt.datum[f\"expected_count\"] >= expected_count_selection[\"cutoff\"] - 1e-6\n",
    "        )\n",
    "        .add_selection(gene_selection)\n",
    "        .add_selection(expected_count_selection)\n",
    "        .properties(width=235, height=125)\n",
    "        .resolve_scale(y=\"independent\")\n",
    "        .configure_axis(grid=False)\n",
    "    )\n",
    "    \n",
    "    if len(namings) > 1:\n",
    "        chartfile = os.path.join(outdir, f\"effects_histogram_{orf1ab_nsp}_naming.html\")\n",
    "    else:\n",
    "        chartfile = os.path.join(outdir, \"effects_histogram.html\")\n",
    "    print(f\"Saving to {chartfile}\")\n",
    "    hist_chart.save(chartfile)\n",
    "\n",
    "    display(hist_chart)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ef4f3fe-a65c-4b0e-856e-a0837fdaead2",
   "metadata": {},
   "source": [
    "## Distributions of mutation effects across genes\n",
    "This plot summarizes the above histograms for all genes in a single plot, showing the individual nsps (not ORF1ab).\n",
    "\n",
    "For one version of the plot, we order genes by position in the genome, so first get that order:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e08ee4aa-fdec-4823-b2ca-349ffc662d87",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get order of all genes after ORF1ab\n",
    "gene_position_order = (\n",
    "    pd.read_csv(ref_coding_sites_csv)\n",
    "    [[\"gene\"]]\n",
    "    .drop_duplicates()\n",
    "    .assign(gene=lambda x: x[\"gene\"].str.split(\";\"))\n",
    "    .explode(\"gene\")\n",
    "    .query(\"gene not in ['noncoding', 'ORF1a', 'ORF1ab']\")\n",
    "    .drop_duplicates()\n",
    "    [\"gene\"]\n",
    "    .tolist()\n",
    ")\n",
    "\n",
    "# add nsp order\n",
    "gene_position_order = list(orf1ab_to_nsps) + gene_position_order"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4070e30-0f07-4d38-a76a-10d747e24a2f",
   "metadata": {},
   "source": [
    "Now make the plots:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15ee20d8-0bb7-4559-a5e6-243baf2a9bdc",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "dist_df = (\n",
    "    aamut_all\n",
    "    .query(\"gene != 'ORF1ab'\")\n",
    "    .assign(\n",
    "        mut_type=lambda x: x[\"aa_mutation\"].map(\n",
    "            lambda m: (\n",
    "                \"synonymous\" if m[0] == m[-1]\n",
    "                else \"stop\" if m[-1] == \"*\" else \"nonsynonymous\"\n",
    "            )\n",
    "        )\n",
    "    )\n",
    "    [[\"gene\", \"expected_count\", \"delta_fitness\", \"mut_type\"]]\n",
    ")\n",
    "\n",
    "# order by median effect of nonsynonymous mutations\n",
    "gene_order = (\n",
    "    dist_df\n",
    "    .query(\"mut_type == 'nonsynonymous'\")\n",
    "    .query(\"expected_count >= @min_expected_count\")\n",
    "    .groupby(\"gene\", as_index=False)\n",
    "    .aggregate({\"delta_fitness\": \"median\"})\n",
    "    .sort_values(\"delta_fitness\")\n",
    "    [\"gene\"]\n",
    "    .tolist()\n",
    ")\n",
    "\n",
    "mut_type_selection = alt.selection_multi(\n",
    "    fields=[\"mut_type\"], bind=\"legend\", toggle=\"true\",\n",
    ")\n",
    "\n",
    "base_dist_chart = (\n",
    "    alt.Chart(dist_df)\n",
    "    .transform_filter(\n",
    "        alt.datum[f\"expected_count\"] >= expected_count_selection[\"cutoff\"] - 1e-6\n",
    "    )\n",
    "    .transform_aggregate(\n",
    "        median=\"median(delta_fitness)\",\n",
    "        lower_quartile=\"q1(delta_fitness)\",\n",
    "        upper_quartile=\"q3(delta_fitness)\",\n",
    "        n=\"valid(delta_fitness)\",\n",
    "        groupby=[\"gene\", \"mut_type\"],\n",
    "    )\n",
    "    .transform_calculate(\n",
    "        medianplus=alt.datum[\"median\"] + 0.1,\n",
    "        medianminus=alt.datum[\"median\"] - 0.1,\n",
    "    )\n",
    "    .encode(\n",
    "        x=alt.X(\n",
    "            \"mut_type\",\n",
    "            title=None,\n",
    "            axis=alt.Axis(labels=False, ticks=True, values=[mut_type_order[1]]),\n",
    "            scale=alt.Scale(\n",
    "                # dummy added for space between samples\n",
    "                domain=[\"dummy\", *mut_type_order, \"dummy2\"]),\n",
    "        ),\n",
    "        color=alt.Color(\n",
    "            \"mut_type\",\n",
    "            title=\"mutation type\",\n",
    "            legend=alt.Legend(orient=\"bottom\", titleOrient=\"left\"),\n",
    "            scale=alt.Scale(range=mut_type_colors),\n",
    "            sort=mut_type_order,\n",
    "        ),\n",
    "        opacity=alt.condition(mut_type_selection, alt.value(1), alt.value(0.2)),\n",
    "        tooltip=[\n",
    "            \"gene\",\n",
    "            alt.Tooltip(\"mut_type\", title=\"mutation type\"),\n",
    "            alt.Tooltip(\"median:Q\", format=\".2f\"),\n",
    "            alt.Tooltip(\"lower_quartile:Q\", format=\".2f\"),\n",
    "            alt.Tooltip(\"upper_quartile:Q\", format=\".2f\"),\n",
    "            alt.Tooltip(\"n:Q\", title=\"number of mutations\"),\n",
    "        ],\n",
    "    )\n",
    ")\n",
    "\n",
    "barwidth = 6\n",
    "\n",
    "median_dist_chart = (\n",
    "    base_dist_chart\n",
    "    .encode(\n",
    "        y=alt.Y(\"medianplus:Q\", title=\"fitness effect of mutation\"),\n",
    "        y2=alt.Y2(\"medianminus:Q\", title=\"fitness effect of mutation\"),\n",
    "    )\n",
    "    .mark_bar(size=barwidth, filled=True)\n",
    ")\n",
    "\n",
    "bars_dist_chart = (\n",
    "    base_dist_chart\n",
    "    .encode(\n",
    "        y=\"lower_quartile:Q\",\n",
    "        y2=\"upper_quartile:Q\",\n",
    "    )\n",
    "    .mark_bar(size=barwidth, filled=True, fillOpacity=0.5, strokeWidth=0)\n",
    ")\n",
    "\n",
    "zero_line = (\n",
    "   alt.Chart(pd.DataFrame({\"zero\": [0]}))\n",
    "    .encode(y=\"zero:Q\")\n",
    "    .mark_rule(color=\"gray\", strokeDash=[3, 3])\n",
    ")\n",
    "\n",
    "height = 200\n",
    "\n",
    "for genesort, chartfile in [\n",
    "    (gene_order, os.path.join(outdir, \"effects_dist.html\")),\n",
    "    (gene_position_order, os.path.join(outdir, \"effects_dist_position_order.html\")),\n",
    "]:\n",
    "    dist_chart = (\n",
    "        (bars_dist_chart + median_dist_chart + zero_line)\n",
    "        .properties(height=height, width=alt.Step(barwidth))\n",
    "        .facet(\n",
    "            column=alt.Column(\n",
    "                \"gene\",\n",
    "                sort=genesort,\n",
    "                title=None,\n",
    "                header=alt.Header(\n",
    "                    labelOrient=\"bottom\",\n",
    "                    labelAngle=-90,\n",
    "                    labelAlign=\"right\",\n",
    "                    labelBaseline=\"middle\",\n",
    "                    labelPadding=height + 10,\n",
    "                ),\n",
    "            ),\n",
    "            spacing=0,\n",
    "        )\n",
    "        .add_selection(expected_count_selection)\n",
    "        .add_selection(mut_type_selection)\n",
    "        .configure_axis(grid=False)\n",
    "        .configure_view(stroke=None)\n",
    "    )\n",
    "\n",
    "    print(f\"\\nSaving to {chartfile}\")\n",
    "    dist_chart.save(chartfile)\n",
    "\n",
    "    display(dist_chart)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7bb7d81c-7e76-4fab-aee1-8473262368ef",
   "metadata": {},
   "source": [
    "## Plot results for individual genes\n",
    "Now we plot heatmaps of the amino-acid fitness estimates for each protein.\n",
    "\n",
    "First, get the \"wildtype\" amino acid at each site in our reference sequence:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "010ef466-5fa0-437c-9663-8589fe14f6ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "# now convert ORF1ab numbers to nsp numbers\n",
    "orf1ab_to_nsps_df = pd.concat(\n",
    "    [\n",
    "        pd.DataFrame(\n",
    "            [(i, i - start + 1) for i in range(start, end + 1)],\n",
    "            columns=[\"ORF1ab_site\", \"nsp_site\"],\n",
    "        ).assign(nsp=nsp).drop_duplicates()\n",
    "        for nsp, (start, end) in orf1ab_to_nsps.items()\n",
    "    ],\n",
    "    ignore_index=True,\n",
    ")\n",
    "\n",
    "clade_founder_aas = pd.concat(\n",
    "    [\n",
    "        clade_founder_aas,\n",
    "        (\n",
    "            clade_founder_aas\n",
    "            .query(\"gene == 'ORF1ab'\")\n",
    "            .merge(\n",
    "                orf1ab_to_nsps_df,\n",
    "                left_on=\"site\",\n",
    "                right_on=\"ORF1ab_site\",\n",
    "                validate=\"many_to_one\",\n",
    "            )\n",
    "            .drop(columns=[\"gene\", \"ORF1ab_site\", \"site\"])\n",
    "            .rename(columns={\"nsp_site\": \"site\", \"nsp\": \"gene\"})\n",
    "        ),\n",
    "    ],\n",
    "    ignore_index=False,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13449038-b096-4c26-827f-6ce0be0d7a93",
   "metadata": {},
   "source": [
    "Now plot the heatmaps:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b34fbcd2-5a1d-4868-9d06-daa6f03d6265",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def plot_aa_fitness(gene, fitness_df, clade_founder_df):\n",
    "    \"\"\"Plot of amino-acid fitness values.\"\"\"\n",
    "    \n",
    "    # biochemically ordered alphabet\n",
    "    aas = tuple(\"RKHDEQNSTYWFAILMVGPC*\")\n",
    "    assert set(fitness_df[\"amino acid\"]).issubset(aas)\n",
    "    \n",
    "    sites = fitness_df[\"site\"].unique().tolist()\n",
    "    \n",
    "    expected_count_selection = alt.selection_single(\n",
    "        bind=alt.binding_range(\n",
    "            min=1,\n",
    "            max=min(5 * min_expected_count, fitness_df[\"expected_count\"].quantile(0.9)),\n",
    "            step=1,\n",
    "            name=\"minimum expected count\",\n",
    "        ),\n",
    "        fields=[\"cutoff\"],\n",
    "        init={\"cutoff\": min_expected_count},\n",
    "    )\n",
    "   \n",
    "    site_zoom_brush = alt.selection_interval(\n",
    "        encodings=[\"x\"],\n",
    "        mark=alt.BrushConfig(\n",
    "            stroke=\"gold\", strokeWidth=1.5, fill=\"yellow\", fillOpacity=0.3,\n",
    "        ),\n",
    "    )\n",
    "        \n",
    "    base = (\n",
    "        alt.Chart(fitness_df)\n",
    "        .encode(x=alt.X(\"site:O\", axis=alt.Axis(labelOverlap=\"parity\")))\n",
    "        .transform_filter(\n",
    "            alt.datum[f\"expected_count\"] >= expected_count_selection[\"cutoff\"] - 1e-6\n",
    "        )\n",
    "    )\n",
    "    \n",
    "    heatmap_y = alt.Y(\"amino acid\", sort=aas, scale=alt.Scale(domain=aas))\n",
    "    heatmap_base = (\n",
    "        base\n",
    "        .encode(y=heatmap_y)\n",
    "        .properties(width=alt.Step(12), height=alt.Step(12))\n",
    "    )\n",
    "    \n",
    "    # background fill for missing values in heatmap, imputing dummy stat\n",
    "    # to get all cells\n",
    "    heatmap_bg = (\n",
    "        heatmap_base\n",
    "        .transform_impute(\n",
    "            impute=\"_stat_dummy\",\n",
    "            key=\"amino acid\",\n",
    "            keyvals=aas,\n",
    "            groupby=[\"site\"],\n",
    "            value=None,\n",
    "        )\n",
    "        .mark_rect(color=\"gray\", opacity=0.25)\n",
    "    )\n",
    "    \n",
    "    # heatmap showing non-filtered amino acids\n",
    "    heatmap_aas = (\n",
    "        heatmap_base\n",
    "        .encode(\n",
    "            color=alt.Color(\n",
    "                \"fitness:Q\",\n",
    "                legend=alt.Legend(\n",
    "                    orient=\"bottom\",\n",
    "                    titleOrient=\"left\",\n",
    "                    gradientLength=150,\n",
    "                    gradientStrokeColor=\"black\",\n",
    "                    gradientStrokeWidth=0.5,\n",
    "                ),\n",
    "                scale=alt.Scale(\n",
    "                    zero=True,\n",
    "                    nice=False,\n",
    "                    type=\"linear\",\n",
    "                    domainMid=0,\n",
    "                    domain=alt.DomainUnionWith(heatmap_minimal_domain),\n",
    "                ),\n",
    "            ),\n",
    "            stroke=alt.value(\"black\"),\n",
    "            tooltip=[\n",
    "                alt.Tooltip(c, format=\".3g\")\n",
    "                if fitness_df[c].dtype == float\n",
    "                else c\n",
    "                for c in fitness_df.columns\n",
    "            ],\n",
    "        )\n",
    "        .mark_rect()\n",
    "    )\n",
    "    \n",
    "    # place X values at \"wildtype\"\n",
    "    wildtype_clade_selection = alt.selection_single(\n",
    "        fields=[\"clade\"],\n",
    "        bind=alt.binding_select(\n",
    "            options=clade_founder_df[\"clade\"].unique(),\n",
    "            name=\"X denotes wildtype in\",\n",
    "        ),\n",
    "        init={\"clade\": clade_label(init_ref_clade)},\n",
    "    )\n",
    "    heatmap_wildtype = (\n",
    "        alt.Chart(clade_founder_df.query(\"site in @sites\"))\n",
    "        .encode(\n",
    "            x=alt.X(\"site:O\"),\n",
    "            y=heatmap_y,\n",
    "        )\n",
    "        .mark_text(text=\"x\", color=\"black\")\n",
    "        .add_selection(wildtype_clade_selection)\n",
    "        .transform_filter(wildtype_clade_selection)\n",
    "        .transform_filter(site_zoom_brush)\n",
    "    )\n",
    "\n",
    "    heatmap = (\n",
    "        (heatmap_bg + heatmap_aas + heatmap_wildtype)\n",
    "        .add_selection(expected_count_selection)\n",
    "        .transform_filter(site_zoom_brush)\n",
    "    )\n",
    "    \n",
    "    # make lineplot\n",
    "    site_statistics = [\"mean\", \"max\", \"min\"]\n",
    "    site_stat = alt.selection_single(\n",
    "        bind=alt.binding_radio(\n",
    "            options=site_statistics,\n",
    "            name=\"site fitness statistic\",\n",
    "        ),\n",
    "        fields=[\"site fitness statistic\"],\n",
    "        init={\"site fitness statistic\": site_statistics[0]},\n",
    "    )\n",
    "    \n",
    "    lineplot = (\n",
    "        base\n",
    "        .transform_filter(alt.datum[\"amino acid\"] != \"*\")\n",
    "        .transform_aggregate(\n",
    "            **{stat: f\"{stat}(fitness)\" for stat in site_statistics},\n",
    "            groupby=[\"site\"],\n",
    "        )\n",
    "        .transform_fold(\n",
    "            site_statistics,\n",
    "            [\"site fitness statistic\", \"site fitness\"],\n",
    "        )\n",
    "        .add_selection(site_stat)\n",
    "        .add_selection(site_zoom_brush)\n",
    "        .transform_filter(site_stat)\n",
    "        .encode(\n",
    "            y=alt.Y(\"site fitness:Q\", axis=alt.Axis(grid=False)),\n",
    "            tooltip=[\n",
    "                \"site\",\n",
    "                alt.Tooltip(\"site fitness:Q\", format=\".3g\"),\n",
    "                \"site fitness statistic:N\",\n",
    "            ],\n",
    "        )\n",
    "        .mark_area(color=\"black\", opacity=0.7)\n",
    "        .properties(\n",
    "            height=75,\n",
    "            width=min(750, 12 * fitness_df[\"site\"].nunique()),\n",
    "            title=alt.TitleParams(\n",
    "                \"use this site plot to zoom into regions on the heat map\",\n",
    "                anchor=\"start\",\n",
    "                fontWeight=\"normal\",\n",
    "                fontSize=11,\n",
    "            ),\n",
    "        )\n",
    "    )\n",
    "    \n",
    "    show_stop = alt.selection_single(\n",
    "        fields=[\"_dummy\"],\n",
    "        bind=alt.binding_radio(\n",
    "            options=[\"yes\", \"no\"],\n",
    "            name=\"show stop in magenta on top site plot\",\n",
    "        ),\n",
    "        init={\"_dummy\": \"no\"},\n",
    "    )\n",
    "    \n",
    "    stopplot = (\n",
    "        base\n",
    "        .add_selection(show_stop)\n",
    "        .transform_filter(alt.datum[\"amino acid\"] == \"*\")\n",
    "        .transform_calculate(_dummy=\"'yes'\")\n",
    "        .transform_filter(show_stop)\n",
    "        .encode(\n",
    "            y=alt.Y(\"fitness\", title=\"site fitness\"),\n",
    "            color=alt.value(\"#CC79A7\"),\n",
    "            tooltip=[\"site\", alt.Tooltip(\"fitness\", format=\".3g\", title=\"stop fitness\")],\n",
    "        )\n",
    "        .mark_line(point=True, strokeWidth=0.5, strokeDash=[2, 2])\n",
    "    )\n",
    "    \n",
    "    return (\n",
    "        (alt.layer(lineplot, stopplot) & heatmap)\n",
    "        .properties(\n",
    "            title=alt.TitleParams(\n",
    "                f\"estimated fitness of amino acids for SARS-CoV-2 {gene} protein\",\n",
    "                fontSize=15,\n",
    "            ),\n",
    "        )\n",
    "        .resolve_scale(color=\"independent\")\n",
    "    )\n",
    "\n",
    "\n",
    "for gene, fitness_df in (\n",
    "    aafitness\n",
    "    [[\"gene\", \"aa_site\", \"aa\", \"fitness\", \"expected_count\"]]\n",
    "    .rename(columns={\"aa_site\": \"site\", \"aa\": \"amino acid\"})\n",
    "    .groupby(\"gene\")\n",
    "):\n",
    "    if gene == \"ORF1ab\":\n",
    "        continue  # do not make a plot for ORF1ab, too big\n",
    "    chart = plot_aa_fitness(gene, fitness_df, clade_founder_aas.query(\"gene == @gene\"))\n",
    "    gene_name_for_file = gene.replace(\" \", \"_\").replace(\"(\", \"\").replace(\")\", \"\")\n",
    "    chartfile = os.path.join(outdir, f\"{gene.split()[0]}.html\")\n",
    "    print(f\"\\nSaving chart for {gene} to {chartfile}\")\n",
    "    chart.save(chartfile)\n",
    "    display(chart)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5341ddd4-b2ce-4567-a9c9-427c2459935a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
