{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9f28f7b9-82e0-4370-b6af-186249f2594b",
   "metadata": {},
   "source": [
    "# Analyze correlation of amino-acid fitnesses with dN/dS values\n",
    "This notebook looks at how dN/dS values from FEL correlation with the amino-acid fitness values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "df514113-d794-43fe-987a-3ebb0b13fa2a",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-28T13:10:54.384317Z",
     "iopub.status.busy": "2023-05-28T13:10:54.383844Z",
     "iopub.status.idle": "2023-05-28T13:10:54.757528Z",
     "shell.execute_reply": "2023-05-28T13:10:54.756517Z",
     "shell.execute_reply.started": "2023-05-28T13:10:54.384284Z"
    }
   },
   "outputs": [],
   "source": [
    "######## snakemake preamble start (automatically inserted, do not edit) ########\n",
    "import sys; sys.path.extend(['/fh/fast/bloom_j/software/miniconda3/envs/SARS2-mut-fitness/lib/python3.9/site-packages', '/home/jbloom/.cache/snakemake/snakemake/source-cache/runtime-cache/tmpdaq7171u/file/fh/fast/bloom_j/computational_notebooks/jbloom/2023/SARS2-mut-fitness/notebooks', '/fh/fast/bloom_j/computational_notebooks/jbloom/2023/SARS2-mut-fitness/notebooks']); import pickle; snakemake = pickle.loads(b'\\x80\\x04\\x95\\xc0\\x17\\x00\\x00\\x00\\x00\\x00\\x00\\x8c\\x10snakemake.script\\x94\\x8c\\tSnakemake\\x94\\x93\\x94)\\x81\\x94}\\x94(\\x8c\\x05input\\x94\\x8c\\x0csnakemake.io\\x94\\x8c\\nInputFiles\\x94\\x93\\x94)\\x81\\x94(\\x8c,results_public_2023-05-11/dnds/dnds_data.csv\\x94\\x8c3results_public_2023-05-11/aa_fitness/aa_fitness.csv\\x94\\x8c?results_public_2023-05-11/dms/dadonaite_ba1_spike/processed.csv\\x94\\x8c5results_public_2023-05-11/dms/starr_rbd/processed.csv\\x94\\x8c8results_public_2023-05-11/dms/iketani_mpro/processed.csv\\x94\\x8c;results_public_2023-05-11/dms/flynn_mpro_2022/processed.csv\\x94\\x8c;results_public_2023-05-11/dms/flynn_mpro_2023/processed.csv\\x94e}\\x94(\\x8c\\x06_names\\x94}\\x94(\\x8c\\x04dnds\\x94K\\x00N\\x86\\x94\\x8c\\naa_fitness\\x94K\\x01N\\x86\\x94\\x8c\\x13dadonaite_ba1_spike\\x94K\\x02N\\x86\\x94\\x8c\\tstarr_rbd\\x94K\\x03N\\x86\\x94\\x8c\\x0ciketani_mpro\\x94K\\x04N\\x86\\x94\\x8c\\x0fflynn_mpro_2022\\x94K\\x05N\\x86\\x94\\x8c\\x0fflynn_mpro_2023\\x94K\\x06N\\x86\\x94u\\x8c\\x12_allowed_overrides\\x94]\\x94(\\x8c\\x05index\\x94\\x8c\\x04sort\\x94eh$\\x8c\\tfunctools\\x94\\x8c\\x07partial\\x94\\x93\\x94h\\x06\\x8c\\x19Namedlist._used_attribute\\x94\\x93\\x94\\x85\\x94R\\x94(h*)}\\x94\\x8c\\x05_name\\x94h$sNt\\x94bh%h(h*\\x85\\x94R\\x94(h*)}\\x94h.h%sNt\\x94bh\\x14h\\nh\\x16h\\x0bh\\x18h\\x0ch\\x1ah\\rh\\x1ch\\x0eh\\x1eh\\x0fh h\\x10ub\\x8c\\x06output\\x94h\\x06\\x8c\\x0bOutputFiles\\x94\\x93\\x94)\\x81\\x94\\x8c-results_public_2023-05-11/dnds/dnds_corr.html\\x94a}\\x94(h\\x12}\\x94\\x8c\\tcorr_html\\x94K\\x00N\\x86\\x94sh\"]\\x94(h$h%eh$h(h*\\x85\\x94R\\x94(h*)}\\x94h.h$sNt\\x94bh%h(h*\\x85\\x94R\\x94(h*)}\\x94h.h%sNt\\x94bh;h8ub\\x8c\\x06params\\x94h\\x06\\x8c\\x06Params\\x94\\x93\\x94)\\x81\\x94(K\\n}\\x94(h\\x18}\\x94(\\x8c\\x05study\\x94\\x8c;https://www.biorxiv.org/content/10.1101/2022.10.13.512056v1\\x94\\x8c\\x0bdescription\\x94\\x8c\\x1dspike (Dadonaite et al, 2022)\\x94\\x8c\\x04gene\\x94\\x8c\\x01S\\x94\\x8c\\x03url\\x94\\x8c\\x8bhttps://raw.githubusercontent.com/dms-vep/SARS-CoV-2_Omicron_BA.1_spike_DMS_mAbs/main/results/muteffects_functional/muteffects_observed.csv\\x94\\x8c\\x0bfilter_cols\\x94}\\x94\\x8c\\ntimes_seen\\x94K\\x03suh\\x1a}\\x94(\\x8c\\x05study\\x94\\x8cOhttps://journals.plos.org/plospathogens/article?id=10.1371/journal.ppat.1010951\\x94\\x8c\\x0bdescription\\x94\\x8c\\x17RBD (Starr et al, 2022)\\x94\\x8c\\x04gene\\x94hQ\\x8c\\x03url\\x94\\x8c\\xadhttps://media.githubusercontent.com/media/jbloomlab/SARS-CoV-2-RBD_DMS_Omicron/f76ba2b2bec18ede9fa9da18c9ccc52389b1ba3a/results/final_variant_scores/final_variant_scores.csv\\x94uh\\x1c}\\x94(\\x8c\\x05study\\x94\\x8cChttps://www.sciencedirect.com/science/article/pii/S1931312822004024\\x94\\x8c\\x0bdescription\\x94\\x8c\\x1aMpro (Iketani et al, 2022)\\x94\\x8c\\x04gene\\x94\\x8c\\x0bnsp5 (Mpro)\\x94\\x8c\\x03url\\x94\\x8c\\xachttps://raw.githubusercontent.com/alejandrochavezlab/3CL_protease_DMS/dc802ee3ea9d43005ca97a28af078767fd66777c/outputs/results/normalized_to_wt_and_stop/activity_scores.csv\\x94\\x8c\\x06wt_seq\\x94\\x8c\\x0cdata/Mpro.fa\\x94uh\\x1e}\\x94(\\x8c\\x05study\\x94\\x8c(https://elifesciences.org/articles/77433\\x94\\x8c\\x0bdescription\\x94\\x8c\\x18Mpro (Flynn et al, 2022)\\x94\\x8c\\x04gene\\x94\\x8c\\x0bnsp5 (Mpro)\\x94\\x8c\\x03url\\x94\\x8cKhttps://cdn.elifesciences.org/articles/77433/elife-77433-fig2-data1-v2.xlsx\\x94\\x8c\\x06wt_seq\\x94\\x8c\\x0cdata/Mpro.fa\\x94uh }\\x94(\\x8c\\x05study\\x94\\x8c)https://doi.org/10.1101/2023.03.02.530652\\x94\\x8c\\x0bdescription\\x94\\x8c\\x18Mpro (Flynn et al, 2023)\\x94\\x8c\\x04gene\\x94\\x8c\\x0bnsp5 (Mpro)\\x94\\x8c\\x03url\\x94\\x8cahttps://www.biorxiv.org/content/biorxiv/early/2023/03/02/2023.03.02.530652/DC1/embed/media-1.xlsx\\x94uue}\\x94(h\\x12}\\x94(\\x8c\\x12min_expected_count\\x94K\\x00N\\x86\\x94\\x8c\\x0cdms_datasets\\x94K\\x01N\\x86\\x94uh\"]\\x94(h$h%eh$h(h*\\x85\\x94R\\x94(h*)}\\x94h.h$sNt\\x94bh%h(h*\\x85\\x94R\\x94(h*)}\\x94h.h%sNt\\x94bh\\x80K\\nh\\x82hJub\\x8c\\twildcards\\x94h\\x06\\x8c\\tWildcards\\x94\\x93\\x94)\\x81\\x94\\x8c\\x11public_2023-05-11\\x94a}\\x94(h\\x12}\\x94\\x8c\\x03mat\\x94K\\x00N\\x86\\x94sh\"]\\x94(h$h%eh$h(h*\\x85\\x94R\\x94(h*)}\\x94h.h$sNt\\x94bh%h(h*\\x85\\x94R\\x94(h*)}\\x94h.h%sNt\\x94b\\x8c\\x03mat\\x94h\\x91ub\\x8c\\x07threads\\x94K\\x01\\x8c\\tresources\\x94h\\x06\\x8c\\tResources\\x94\\x93\\x94)\\x81\\x94(K\\x01K\\x01\\x8c\\x04/tmp\\x94e}\\x94(h\\x12}\\x94(\\x8c\\x06_cores\\x94K\\x00N\\x86\\x94\\x8c\\x06_nodes\\x94K\\x01N\\x86\\x94\\x8c\\x06tmpdir\\x94K\\x02N\\x86\\x94uh\"]\\x94(h$h%eh$h(h*\\x85\\x94R\\x94(h*)}\\x94h.h$sNt\\x94bh%h(h*\\x85\\x94R\\x94(h*)}\\x94h.h%sNt\\x94bh\\xa8K\\x01h\\xaaK\\x01h\\xach\\xa5ub\\x8c\\x03log\\x94h\\x06\\x8c\\x03Log\\x94\\x93\\x94)\\x81\\x94\\x8c1results_public_2023-05-11/dnds/analyze_dnds.ipynb\\x94a}\\x94(h\\x12}\\x94\\x8c\\x08notebook\\x94K\\x00N\\x86\\x94sh\"]\\x94(h$h%eh$h(h*\\x85\\x94R\\x94(h*)}\\x94h.h$sNt\\x94bh%h(h*\\x85\\x94R\\x94(h*)}\\x94h.h%sNt\\x94bh\\xbeh\\xbbub\\x8c\\x06config\\x94}\\x94(\\x8c\\tmat_trees\\x94}\\x94(\\x8c\\x11public_2023-05-11\\x94\\x8c\\x83http://hgdownload.soe.ucsc.edu/goldenPath/wuhCor1/UShER_SARS-CoV-2/2023/05/11/public-2023-05-11.all.masked.nextclade.pangolin.pb.gz\\x94\\x8c\\x11public_2023-03-27\\x94\\x8c\\x83http://hgdownload.soe.ucsc.edu/goldenPath/wuhCor1/UShER_SARS-CoV-2/2023/03/27/public-2023-03-27.all.masked.nextclade.pangolin.pb.gz\\x94\\x8c\\x11public_2022-12-18\\x94\\x8c\\x83http://hgdownload.soe.ucsc.edu/goldenPath/wuhCor1/UShER_SARS-CoV-2/2022/12/18/public-2022-12-18.all.masked.nextclade.pangolin.pb.gz\\x94\\x8c\\x11gisaid_2023-03-29\\x94\\x8cQhttps://hgwdev.gi.ucsc.edu/~angie/2847b92/gisaidAndPublic.2023-03-29.masked.pb.gz\\x94u\\x8c\\x0bcurrent_mat\\x94\\x8c\\x11public_2023-05-11\\x94\\x8c\\tref_fasta\\x94\\x8cGhttp://hgdownload.soe.ucsc.edu/goldenPath/wuhCor1/bigZips/wuhCor1.fa.gz\\x94\\x8c\\x07ref_gtf\\x94\\x8cPhttp://hgdownload.soe.ucsc.edu/goldenPath/wuhCor1/bigZips/genes/ncbiGenes.gtf.gz\\x94\\x8c\\x0eadd_to_ref_gtf\\x94}\\x94\\x8c\\x05ORF9b\\x94]\\x94(M|nM\\xa1oes\\x8c\\x11min_clade_samples\\x94M\\x10\\'\\x8c\\x0esample_subsets\\x94}\\x94(\\x8c\\x03all\\x94\\x8c\\x01.\\x94\\x8c\\x03USA\\x94\\x8c\\x03USA\\x94\\x8c\\x07England\\x94\\x8c\\x07England\\x94u\\x8c\\x12clade_founder_json\\x94\\x8cyhttps://raw.githubusercontent.com/neherlab/SC2_variant_rates/7e738194a8c6592082f1caa9a6ca70cb68289790/data/clade_gts.json\\x94\\x8c\\x11clades_to_exclude\\x94]\\x94\\x8c\\x0321A\\x94a\\x8c\\x10max_nt_mutations\\x94K\\x04\\x8c\\x15max_reversions_to_ref\\x94K\\x01\\x8c\\x1fmax_reversions_to_clade_founder\\x94K\\x01\\x8c\\x1bexclude_ref_to_founder_muts\\x94\\x88\\x8c\\x12site_include_range\\x94}\\x94(\\x8c\\x05start\\x94M\\n\\x01\\x8c\\x03end\\x94M\\xeasu\\x8c\\x10sites_to_exclude\\x94]\\x94(K\\x99M}\\x04M\\x96\\x08MI\\x0cM\\xec\\rM\\xc2\\x0eM\\xd2\\x0fMo\\x18MV\\x1fMV\"MZ4M{6M\\x12YME_MF_MeaM\\xfd\\x15M\\xc3\\x1aM\\xa0\\x1cM\\xbfmM\\xb2re\\x8c\\rsite_mask_vcf\\x94\\x8cvhttps://raw.githubusercontent.com/W-L/ProblematicSites_SARS-CoV2/master/subset_vcf/problematic_sites_sarsCov2.mask.vcf\\x94\\x8c\\x12usher_masked_sites\\x94\\x8c\\x1cdata/usher_masked_sites.yaml\\x94\\x8c\\x1dsynonymous_spectra_min_counts\\x94M\\x88\\x13\\x8c\\x0eorf1ab_to_nsps\\x94}\\x94(\\x8c\\x04nsp1\\x94]\\x94(K\\x01K\\xb4e\\x8c\\x04nsp2\\x94]\\x94(K\\xb5M2\\x03e\\x8c\\x04nsp3\\x94]\\x94(M3\\x03M\\xcb\\ne\\x8c\\x04nsp4\\x94]\\x94(M\\xcc\\nM\\xbf\\x0ce\\x8c\\x0bnsp5 (Mpro)\\x94]\\x94(M\\xc0\\x0cM\\xf1\\re\\x8c\\x04nsp6\\x94]\\x94(M\\xf2\\rM\\x13\\x0fe\\x8c\\x04nsp7\\x94]\\x94(M\\x14\\x0fMf\\x0fe\\x8c\\x04nsp8\\x94]\\x94(Mg\\x0fM,\\x10e\\x8c\\x04nsp9\\x94]\\x94(M-\\x10M\\x9d\\x10e\\x8c\\x05nsp10\\x94]\\x94(M\\x9e\\x10M(\\x11e\\x8c\\x0cnsp12 (RdRp)\\x94]\\x94(M)\\x11M\\xcc\\x14e\\x8c\\x05nsp13\\x94]\\x94(M\\xcd\\x14M%\\x17e\\x8c\\x05nsp14\\x94]\\x94(M&\\x17M4\\x19e\\x8c\\x05nsp15\\x94]\\x94(M5\\x19M\\x8e\\x1ae\\x8c\\x05nsp16\\x94]\\x94(M\\x8f\\x1aM\\xb8\\x1beu\\x8c\\rgene_overlaps\\x94}\\x94(\\x8c\\x07exclude\\x94]\\x94(]\\x94(\\x8c\\x05ORF1a\\x94\\x8c\\x06ORF1ab\\x94e]\\x94(\\x8c\\x05ORF7a\\x94\\x8c\\x05ORF7b\\x94ee\\x8c\\x06retain\\x94]\\x94]\\x94(\\x8c\\x01N\\x94\\x8c\\x05ORF9b\\x94eau\\x8c\\x13fitness_pseudocount\\x94G?\\xe0\\x00\\x00\\x00\\x00\\x00\\x00\\x8c\\x12min_expected_count\\x94K\\n\\x8c\\x14clade_corr_min_count\\x94J \\xa1\\x07\\x00\\x8c\\x19aa_fitness_init_ref_clade\\x94\\x8c\\x0319A\\x94\\x8c!aa_fitness_heatmap_minimal_domain\\x94]\\x94(J\\xfa\\xff\\xff\\xffK\\x02e\\x8c\\x0eclade_synonyms\\x94}\\x94(\\x8c\\x0319A\\x94\\x8c\\x01B\\x94\\x8c\\x0320A\\x94\\x8c\\x03B.1\\x94\\x8c\\x0320B\\x94\\x8c\\x05B.1.1\\x94\\x8c\\x0320C\\x94\\x8c\\x07B.1.367\\x94\\x8c\\x0320E\\x94\\x8c\\x07B.1.177\\x94\\x8c\\x0320F\\x94\\x8c\\x03D.2\\x94\\x8c\\x0320G\\x94\\x8c\\x05B.1.2\\x94\\x8c\\x0320I\\x94\\x8c\\x05Alpha\\x94\\x8c\\x0320J\\x94\\x8c\\x05Gamma\\x94\\x8c\\x0321C\\x94\\x8c\\x07Epsilon\\x94\\x8c\\x0321F\\x94\\x8c\\x04Iota\\x94\\x8c\\x0321I\\x94\\x8c\\x05Delta\\x94\\x8c\\x0321J\\x94\\x8c\\x05Delta\\x94\\x8c\\x0321K\\x94\\x8c\\x0cOmicron BA.1\\x94\\x8c\\x0321L\\x94\\x8c\\x0cOmicron BA.2\\x94\\x8c\\x0322A\\x94\\x8c\\x0cOmicron BA.4\\x94\\x8c\\x0322B\\x94\\x8c\\x0cOmicron BA.5\\x94\\x8c\\x0322C\\x94\\x8c\\x11Omicron BA.2.12.1\\x94\\x8c\\x0322D\\x94\\x8c\\x0fOmicron BA.2.75\\x94\\x8c\\x0322E\\x94\\x8c\\x0cOmicron BQ.1\\x94u\\x8c\\x0cdms_datasets\\x94hJ\\x8c\\rneher_fitness\\x94\\x8c\\x19data/Neher_aa_fitness.csv\\x94\\x8c\\x14clade_fixed_muts_ref\\x94\\x8c\\x0319A\\x94\\x8c\\x14terminal_pseudocount\\x94G?\\xe0\\x00\\x00\\x00\\x00\\x00\\x00\\x8c\\x19terminal_min_actual_count\\x94K\\x05\\x8c\\x04dnds\\x94\\x8c^https://raw.githubusercontent.com/spond/SARS-CoV-2-variation/master/windowed-sites-fel-all.csv\\x94\\x8c\\x08docs_url\\x94\\x8c-https://jbloomlab.github.io/SARS2-mut-fitness\\x94\\x8c\\x15docs_plot_annotations\\x94\\x8c\\x1fdata/docs_plot_annotations.yaml\\x94u\\x8c\\x04rule\\x94\\x8c\\x0canalyze_dnds\\x94\\x8c\\x0fbench_iteration\\x94N\\x8c\\tscriptdir\\x94\\x8cP/fh/fast/bloom_j/computational_notebooks/jbloom/2023/SARS2-mut-fitness/notebooks\\x94ub.'); from snakemake.logging import logger; logger.printshellcmds = False; import os; os.chdir(r'/fh/fast/bloom_j/computational_notebooks/jbloom/2023/SARS2-mut-fitness');\n",
    "######## snakemake preamble end #########\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "cf8d9563-72f3-47da-a1c1-2934714ce0bd",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-28T13:10:54.761666Z",
     "iopub.status.busy": "2023-05-28T13:10:54.761347Z",
     "iopub.status.idle": "2023-05-28T13:10:54.765394Z",
     "shell.execute_reply": "2023-05-28T13:10:54.764714Z",
     "shell.execute_reply.started": "2023-05-28T13:10:54.761638Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# get variables from `snakemake`\n",
    "dnds_csv = snakemake.input.dnds\n",
    "aa_fitness_csv = snakemake.input.aa_fitness\n",
    "corr_html = snakemake.output.corr_html\n",
    "min_expected_count = snakemake.params.min_expected_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "321f4ab5-e53a-4c7e-be22-3dc552bfa495",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-28T13:10:54.768972Z",
     "iopub.status.busy": "2023-05-28T13:10:54.768672Z",
     "iopub.status.idle": "2023-05-28T13:10:59.243251Z",
     "shell.execute_reply": "2023-05-28T13:10:59.241743Z",
     "shell.execute_reply.started": "2023-05-28T13:10:54.768945Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "import altair as alt\n",
    "\n",
    "import numpy\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "import scipy.stats.mstats"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02e207bc-caeb-4cb9-9cf5-f09b67729129",
   "metadata": {},
   "source": [
    "Get dN/dS values averaging the dN (`beta`) and dS (`alpha`) values over timeframes first, and clipping the dN/dS values at 0.05 and 20:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "812a4e92-7a53-4611-b2aa-c6c14cd4ca87",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-28T13:10:59.251636Z",
     "iopub.status.busy": "2023-05-28T13:10:59.251275Z",
     "iopub.status.idle": "2023-05-28T13:10:59.850880Z",
     "shell.execute_reply": "2023-05-28T13:10:59.850075Z",
     "shell.execute_reply.started": "2023-05-28T13:10:59.251604Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>site</th>\n",
       "      <th>dN</th>\n",
       "      <th>log dN/dS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>nsp5 (Mpro)</td>\n",
       "      <td>1</td>\n",
       "      <td>1.108253</td>\n",
       "      <td>1.177224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>nsp5 (Mpro)</td>\n",
       "      <td>2</td>\n",
       "      <td>1.369948</td>\n",
       "      <td>0.887329</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>nsp5 (Mpro)</td>\n",
       "      <td>3</td>\n",
       "      <td>0.340120</td>\n",
       "      <td>-0.978962</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>nsp5 (Mpro)</td>\n",
       "      <td>4</td>\n",
       "      <td>1.140580</td>\n",
       "      <td>-0.488341</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>nsp5 (Mpro)</td>\n",
       "      <td>5</td>\n",
       "      <td>2.353438</td>\n",
       "      <td>1.495519</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9408</th>\n",
       "      <td>nsp9</td>\n",
       "      <td>109</td>\n",
       "      <td>3.373902</td>\n",
       "      <td>2.995732</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9409</th>\n",
       "      <td>nsp9</td>\n",
       "      <td>110</td>\n",
       "      <td>0.339979</td>\n",
       "      <td>-1.458756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9410</th>\n",
       "      <td>nsp9</td>\n",
       "      <td>111</td>\n",
       "      <td>0.684415</td>\n",
       "      <td>0.711573</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9411</th>\n",
       "      <td>nsp9</td>\n",
       "      <td>112</td>\n",
       "      <td>0.609013</td>\n",
       "      <td>-2.394275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9412</th>\n",
       "      <td>nsp9</td>\n",
       "      <td>113</td>\n",
       "      <td>2.185392</td>\n",
       "      <td>-0.023591</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9413 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene  site        dN  log dN/dS\n",
       "0     nsp5 (Mpro)     1  1.108253   1.177224\n",
       "1     nsp5 (Mpro)     2  1.369948   0.887329\n",
       "2     nsp5 (Mpro)     3  0.340120  -0.978962\n",
       "3     nsp5 (Mpro)     4  1.140580  -0.488341\n",
       "4     nsp5 (Mpro)     5  2.353438   1.495519\n",
       "...           ...   ...       ...        ...\n",
       "9408         nsp9   109  3.373902   2.995732\n",
       "9409         nsp9   110  0.339979  -1.458756\n",
       "9410         nsp9   111  0.684415   0.711573\n",
       "9411         nsp9   112  0.609013  -2.394275\n",
       "9412         nsp9   113  2.185392  -0.023591\n",
       "\n",
       "[9413 rows x 4 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_map = {\n",
    "    \"3C\": \"nsp5 (Mpro)\",\n",
    "    \"RdRp\": \"nsp12 (RdRp)\",\n",
    "    \"endornase\": \"nsp15\",\n",
    "    \"exonuclease\": \"nsp14\",\n",
    "    \"helicase\": \"nsp13\",\n",
    "    \"leader\": \"nsp1\",\n",
    "    \"methyltransferase\": \"nsp16\",\n",
    "}\n",
    "\n",
    "dnds = (\n",
    "    pd.read_csv(dnds_csv)\n",
    "    .query(\"(alpha != 0) or (beta != 0)\")\n",
    "    .groupby([\"gene\", \"site\"], as_index=False)\n",
    "    .aggregate({\"alpha\": \"mean\", \"beta\": \"mean\"})\n",
    "    .assign(\n",
    "        dnds=lambda x: numpy.clip(x[\"beta\"] / x[\"alpha\"], a_min=0.05, a_max=20),\n",
    "        log_dnds=lambda x: numpy.log(x[\"dnds\"]),\n",
    "        gene=lambda x: x[\"gene\"].map(lambda g: gene_map[g] if g in gene_map else g),\n",
    "    )\n",
    "    .rename(columns={\"beta\": \"dN\", \"log_dnds\": \"log dN/dS\"})\n",
    "    .drop(columns=[\"alpha\", \"dnds\"])\n",
    ")\n",
    "\n",
    "dnds"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57f59b47-80e9-469f-9bae-32d0af6bd264",
   "metadata": {},
   "source": [
    "Read amino-acid fitnesses, and for each site:\n",
    " - the mean fitness of mutations\n",
    " - the number of effective amino-acids when amino acids are assigned a probability weight of $e^{fitness}$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8bd92aad-fb86-4291-ae86-ac2aef086462",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-28T13:10:59.854697Z",
     "iopub.status.busy": "2023-05-28T13:10:59.854385Z",
     "iopub.status.idle": "2023-05-28T13:11:01.613758Z",
     "shell.execute_reply": "2023-05-28T13:11:01.613029Z",
     "shell.execute_reply.started": "2023-05-28T13:10:59.854671Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>site</th>\n",
       "      <th>mean fitness</th>\n",
       "      <th>fitness entropy</th>\n",
       "      <th>fitness_n_effective</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>E</td>\n",
       "      <td>1</td>\n",
       "      <td>-4.044917</td>\n",
       "      <td>0.309668</td>\n",
       "      <td>1.362972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>E</td>\n",
       "      <td>2</td>\n",
       "      <td>-2.786233</td>\n",
       "      <td>0.774795</td>\n",
       "      <td>2.170148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>E</td>\n",
       "      <td>3</td>\n",
       "      <td>-1.758716</td>\n",
       "      <td>1.171519</td>\n",
       "      <td>3.226890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>E</td>\n",
       "      <td>4</td>\n",
       "      <td>-1.636612</td>\n",
       "      <td>1.097318</td>\n",
       "      <td>2.996120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>E</td>\n",
       "      <td>5</td>\n",
       "      <td>-1.096513</td>\n",
       "      <td>1.234616</td>\n",
       "      <td>3.437060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9736</th>\n",
       "      <td>nsp9</td>\n",
       "      <td>109</td>\n",
       "      <td>-2.159867</td>\n",
       "      <td>0.743339</td>\n",
       "      <td>2.102945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9737</th>\n",
       "      <td>nsp9</td>\n",
       "      <td>110</td>\n",
       "      <td>-2.573740</td>\n",
       "      <td>0.744499</td>\n",
       "      <td>2.105387</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9738</th>\n",
       "      <td>nsp9</td>\n",
       "      <td>111</td>\n",
       "      <td>-3.732329</td>\n",
       "      <td>0.571722</td>\n",
       "      <td>1.771315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9739</th>\n",
       "      <td>nsp9</td>\n",
       "      <td>112</td>\n",
       "      <td>-2.238080</td>\n",
       "      <td>0.949155</td>\n",
       "      <td>2.583527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9740</th>\n",
       "      <td>nsp9</td>\n",
       "      <td>113</td>\n",
       "      <td>-3.126759</td>\n",
       "      <td>1.052402</td>\n",
       "      <td>2.864523</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9741 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene  site  mean fitness  fitness entropy  fitness_n_effective\n",
       "0        E     1     -4.044917         0.309668             1.362972\n",
       "1        E     2     -2.786233         0.774795             2.170148\n",
       "2        E     3     -1.758716         1.171519             3.226890\n",
       "3        E     4     -1.636612         1.097318             2.996120\n",
       "4        E     5     -1.096513         1.234616             3.437060\n",
       "...    ...   ...           ...              ...                  ...\n",
       "9736  nsp9   109     -2.159867         0.743339             2.102945\n",
       "9737  nsp9   110     -2.573740         0.744499             2.105387\n",
       "9738  nsp9   111     -3.732329         0.571722             1.771315\n",
       "9739  nsp9   112     -2.238080         0.949155             2.583527\n",
       "9740  nsp9   113     -3.126759         1.052402             2.864523\n",
       "\n",
       "[9741 rows x 5 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aa_fitness = (\n",
    "    pd.read_csv(aa_fitness_csv)\n",
    "    .query(\"gene not in ['ORF1a', 'ORF1ab']\")\n",
    "    .query(\"expected_count >= @min_expected_count\")\n",
    "    .rename(columns={\"aa_site\": \"site\"})\n",
    "    .assign(\n",
    "        p_aa=lambda x: numpy.exp(x[\"fitness\"]),\n",
    "        p=lambda x: x[\"p_aa\"] / x.groupby([\"gene\", \"site\"])[\"p_aa\"].transform(\"sum\"),\n",
    "    )\n",
    "    .groupby([\"gene\", \"site\"], as_index=False)\n",
    "    .aggregate(\n",
    "        mean_fitness=pd.NamedAgg(\"fitness\", \"mean\"),\n",
    "        fitness_entropy=pd.NamedAgg(\"p\", lambda p: -(p * numpy.log(p)).sum()),\n",
    "    )\n",
    "    .assign(fitness_n_effective=lambda x: numpy.exp(x[\"fitness_entropy\"]))\n",
    "    .rename(columns={\"mean_fitness\": \"mean fitness\", \"fitness_entropy\": \"fitness entropy\"})\n",
    ")\n",
    "\n",
    "assert not (set(dnds[\"gene\"].unique()) - set(aa_fitness[\"gene\"].unique()))\n",
    "\n",
    "aa_fitness"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c55ad63-8ea2-4200-b1ab-5377b02d1355",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-26T23:29:30.097371Z",
     "iopub.status.busy": "2023-05-26T23:29:30.096723Z",
     "iopub.status.idle": "2023-05-26T23:29:30.112190Z",
     "shell.execute_reply": "2023-05-26T23:29:30.111483Z",
     "shell.execute_reply.started": "2023-05-26T23:29:30.097327Z"
    },
    "tags": []
   },
   "source": [
    "Merge amino-acid fitness estimates aggregated at site level with dN/dS values:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "dedc5481-5693-47d5-9073-64ada95049e4",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-28T13:11:01.617183Z",
     "iopub.status.busy": "2023-05-28T13:11:01.616884Z",
     "iopub.status.idle": "2023-05-28T13:11:01.628300Z",
     "shell.execute_reply": "2023-05-28T13:11:01.627614Z",
     "shell.execute_reply.started": "2023-05-28T13:11:01.617161Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "df = aa_fitness.merge(dnds, on=[\"gene\", \"site\"], validate=\"one_to_one\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6d8953ed-cc18-4471-8075-22caad17d881",
   "metadata": {},
   "source": [
    "Get correlations for each gene, then plot:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4181333b-2f0a-417d-82a8-6c994371de13",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-28T13:11:01.631489Z",
     "iopub.status.busy": "2023-05-28T13:11:01.631220Z",
     "iopub.status.idle": "2023-05-28T13:11:01.665177Z",
     "shell.execute_reply": "2023-05-28T13:11:01.664519Z",
     "shell.execute_reply.started": "2023-05-28T13:11:01.631467Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>dnds_metric</th>\n",
       "      <th>fitness metric</th>\n",
       "      <th>Pearson correlation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>E</td>\n",
       "      <td>dN</td>\n",
       "      <td>mean fitness</td>\n",
       "      <td>0.436386</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>E</td>\n",
       "      <td>log dN/dS</td>\n",
       "      <td>mean fitness</td>\n",
       "      <td>0.227921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>M</td>\n",
       "      <td>dN</td>\n",
       "      <td>mean fitness</td>\n",
       "      <td>0.492178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>M</td>\n",
       "      <td>log dN/dS</td>\n",
       "      <td>mean fitness</td>\n",
       "      <td>0.318985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>N</td>\n",
       "      <td>dN</td>\n",
       "      <td>mean fitness</td>\n",
       "      <td>0.553152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>nsp7</td>\n",
       "      <td>log dN/dS</td>\n",
       "      <td>fitness entropy</td>\n",
       "      <td>0.320153</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>nsp8</td>\n",
       "      <td>dN</td>\n",
       "      <td>fitness entropy</td>\n",
       "      <td>0.117486</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>nsp8</td>\n",
       "      <td>log dN/dS</td>\n",
       "      <td>fitness entropy</td>\n",
       "      <td>0.272405</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>nsp9</td>\n",
       "      <td>dN</td>\n",
       "      <td>fitness entropy</td>\n",
       "      <td>0.165345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>nsp9</td>\n",
       "      <td>log dN/dS</td>\n",
       "      <td>fitness entropy</td>\n",
       "      <td>0.327407</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>92 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene dnds_metric   fitness metric  Pearson correlation\n",
       "0      E          dN     mean fitness             0.436386\n",
       "1      E   log dN/dS     mean fitness             0.227921\n",
       "2      M          dN     mean fitness             0.492178\n",
       "3      M   log dN/dS     mean fitness             0.318985\n",
       "4      N          dN     mean fitness             0.553152\n",
       "..   ...         ...              ...                  ...\n",
       "87  nsp7   log dN/dS  fitness entropy             0.320153\n",
       "88  nsp8          dN  fitness entropy             0.117486\n",
       "89  nsp8   log dN/dS  fitness entropy             0.272405\n",
       "90  nsp9          dN  fitness entropy             0.165345\n",
       "91  nsp9   log dN/dS  fitness entropy             0.327407\n",
       "\n",
       "[92 rows x 4 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dnds_metrics = [\"dN\", \"log dN/dS\"]\n",
    "fitness_metrics = [\"mean fitness\", \"fitness entropy\"]\n",
    "\n",
    "gene_corrs = (\n",
    "    df\n",
    "    .melt(\n",
    "        id_vars=[c for c in df.columns if c not in dnds_metrics],\n",
    "        value_vars=dnds_metrics,\n",
    "        var_name=\"dnds_metric\",\n",
    "        value_name=\"dnds_value\",\n",
    "    )\n",
    "    .groupby([\"gene\", \"dnds_metric\"])\n",
    "    [[\"dnds_value\"] + fitness_metrics]\n",
    "    .corr()\n",
    "    .reset_index()\n",
    "    .query(\"level_2 == 'dnds_value'\")\n",
    "    .melt(\n",
    "        id_vars=[\"gene\", \"dnds_metric\"],\n",
    "        value_vars=fitness_metrics,\n",
    "        var_name=\"fitness metric\",\n",
    "        value_name=\"Pearson correlation\",\n",
    "    )\n",
    ")\n",
    "\n",
    "gene_corrs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a877423-809f-4ec2-a0f0-c3970dc5ecb8",
   "metadata": {},
   "source": [
    "Now plot the gene correlations:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "fabfdd62-f31b-4a12-813d-62976ce1698d",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-28T13:11:01.668310Z",
     "iopub.status.busy": "2023-05-28T13:11:01.668029Z",
     "iopub.status.idle": "2023-05-28T13:11:01.743110Z",
     "shell.execute_reply": "2023-05-28T13:11:01.742548Z",
     "shell.execute_reply.started": "2023-05-28T13:11:01.668292Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-2a54a3d8a8eb45baaab6d1c77f1c09b5\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  var VEGA_DEBUG = (typeof VEGA_DEBUG == \"undefined\") ? {} : VEGA_DEBUG;\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-2a54a3d8a8eb45baaab6d1c77f1c09b5\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-2a54a3d8a8eb45baaab6d1c77f1c09b5\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm//vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm//vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm//vega-lite@4.17.0?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm//vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function maybeLoadScript(lib, version) {\n",
       "      var key = `${lib.replace(\"-\", \"\")}_version`;\n",
       "      return (VEGA_DEBUG[key] == version) ?\n",
       "        Promise.resolve(paths[lib]) :\n",
       "        new Promise(function(resolve, reject) {\n",
       "          var s = document.createElement('script');\n",
       "          document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "          s.async = true;\n",
       "          s.onload = () => {\n",
       "            VEGA_DEBUG[key] = version;\n",
       "            return resolve(paths[lib]);\n",
       "          };\n",
       "          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "          s.src = paths[lib];\n",
       "        });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else {\n",
       "      maybeLoadScript(\"vega\", \"5\")\n",
       "        .then(() => maybeLoadScript(\"vega-lite\", \"4.17.0\"))\n",
       "        .then(() => maybeLoadScript(\"vega-embed\", \"6\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 400, \"continuousHeight\": 300}}, \"data\": {\"name\": \"data-34ddeab76a8f6bf8de0080e7d59d1ffa\"}, \"mark\": {\"type\": \"point\", \"filled\": true, \"size\": 50}, \"encoding\": {\"color\": {\"field\": \"fitness metric\", \"type\": \"nominal\"}, \"facet\": {\"columns\": 1, \"field\": \"dnds_metric\", \"header\": {\"labelFontSize\": 13, \"labelFontStyle\": \"bold\", \"labelPadding\": 2}, \"sort\": \"descending\", \"title\": null, \"type\": \"nominal\"}, \"shape\": {\"field\": \"fitness metric\", \"type\": \"nominal\"}, \"tooltip\": [{\"field\": \"gene\", \"type\": \"nominal\"}, {\"field\": \"dnds_metric\", \"type\": \"nominal\"}, {\"field\": \"fitness metric\", \"type\": \"nominal\"}, {\"field\": \"Pearson correlation\", \"format\": \".2f\", \"type\": \"quantitative\"}], \"x\": {\"field\": \"gene\", \"scale\": {\"domain\": [\"N\", \"nsp12 (RdRp)\", \"S\", \"nsp16\", \"nsp5 (Mpro)\", \"nsp13\", \"nsp4\", \"nsp14\", \"nsp3\", \"nsp10\", \"nsp2\", \"nsp6\", \"nsp15\", \"M\", \"nsp9\", \"nsp7\", \"nsp8\", \"E\", \"ORF3a\", \"nsp1\", \"ORF6\", \"ORF7a\", \"ORF8\"]}, \"type\": \"nominal\"}, \"y\": {\"field\": \"Pearson correlation\", \"type\": \"quantitative\"}}, \"height\": 175, \"width\": {\"step\": 25}, \"$schema\": \"https://vega.github.io/schema/vega-lite/v4.17.0.json\", \"datasets\": {\"data-34ddeab76a8f6bf8de0080e7d59d1ffa\": [{\"gene\": \"E\", \"dnds_metric\": \"dN\", \"fitness metric\": \"mean fitness\", \"Pearson correlation\": 0.43638578212371293}, {\"gene\": \"E\", \"dnds_metric\": \"log dN/dS\", \"fitness metric\": \"mean fitness\", \"Pearson correlation\": 0.22792073857558046}, {\"gene\": \"M\", \"dnds_metric\": \"dN\", \"fitness metric\": \"mean fitness\", \"Pearson correlation\": 0.4921778676511673}, {\"gene\": \"M\", \"dnds_metric\": \"log dN/dS\", \"fitness metric\": \"mean fitness\", \"Pearson correlation\": 0.31898491742070056}, {\"gene\": \"N\", \"dnds_metric\": \"dN\", \"fitness metric\": \"mean fitness\", \"Pearson correlation\": 0.5531524814511978}, {\"gene\": \"N\", \"dnds_metric\": \"log dN/dS\", \"fitness metric\": \"mean fitness\", \"Pearson correlation\": 0.5789046579528291}, {\"gene\": \"ORF3a\", \"dnds_metric\": \"dN\", \"fitness metric\": \"mean fitness\", \"Pearson correlation\": 0.3564009011579627}, {\"gene\": \"ORF3a\", \"dnds_metric\": \"log dN/dS\", \"fitness metric\": \"mean fitness\", \"Pearson correlation\": 0.24378935343601232}, {\"gene\": \"ORF6\", \"dnds_metric\": \"dN\", \"fitness metric\": \"mean fitness\", \"Pearson correlation\": 0.29941542607250476}, {\"gene\": \"ORF6\", \"dnds_metric\": \"log dN/dS\", \"fitness metric\": \"mean fitness\", \"Pearson correlation\": -0.041829563693621986}, {\"gene\": \"ORF7a\", \"dnds_metric\": \"dN\", \"fitness metric\": \"mean fitness\", \"Pearson correlation\": 0.29511536232395985}, {\"gene\": \"ORF7a\", \"dnds_metric\": \"log dN/dS\", \"fitness metric\": \"mean fitness\", \"Pearson correlation\": 0.028967319057216313}, {\"gene\": \"ORF8\", \"dnds_metric\": \"dN\", \"fitness metric\": \"mean fitness\", \"Pearson correlation\": 0.330475543515449}, {\"gene\": \"ORF8\", \"dnds_metric\": \"log dN/dS\", \"fitness metric\": \"mean fitness\", \"Pearson correlation\": 0.09021262487704017}, {\"gene\": \"S\", \"dnds_metric\": \"dN\", \"fitness metric\": \"mean fitness\", \"Pearson correlation\": 0.40438035535603295}, {\"gene\": \"S\", \"dnds_metric\": \"log dN/dS\", \"fitness metric\": \"mean fitness\", \"Pearson correlation\": 0.5055065306691134}, {\"gene\": \"nsp1\", \"dnds_metric\": \"dN\", \"fitness metric\": \"mean fitness\", \"Pearson correlation\": 0.15548171794332952}, {\"gene\": \"nsp1\", \"dnds_metric\": \"log dN/dS\", \"fitness metric\": \"mean fitness\", \"Pearson correlation\": 0.12063752772651795}, {\"gene\": \"nsp10\", \"dnds_metric\": \"dN\", \"fitness metric\": \"mean fitness\", \"Pearson correlation\": 0.14411574942212194}, {\"gene\": \"nsp10\", \"dnds_metric\": \"log dN/dS\", \"fitness metric\": \"mean fitness\", \"Pearson correlation\": 0.40303770867581795}, {\"gene\": \"nsp12 (RdRp)\", \"dnds_metric\": \"dN\", \"fitness metric\": \"mean fitness\", \"Pearson correlation\": 0.41489167930015447}, {\"gene\": \"nsp12 (RdRp)\", \"dnds_metric\": \"log dN/dS\", \"fitness metric\": \"mean fitness\", \"Pearson correlation\": 0.47915864080659887}, {\"gene\": \"nsp13\", \"dnds_metric\": \"dN\", \"fitness metric\": \"mean fitness\", \"Pearson correlation\": 0.3462976363330683}, {\"gene\": \"nsp13\", \"dnds_metric\": \"log dN/dS\", \"fitness metric\": \"mean fitness\", \"Pearson correlation\": 0.4387403703748892}, {\"gene\": \"nsp14\", \"dnds_metric\": \"dN\", \"fitness metric\": \"mean fitness\", \"Pearson correlation\": 0.3336866585032698}, {\"gene\": \"nsp14\", \"dnds_metric\": \"log dN/dS\", \"fitness metric\": \"mean fitness\", \"Pearson correlation\": 0.4110920514199762}, {\"gene\": \"nsp15\", \"dnds_metric\": \"dN\", \"fitness metric\": \"mean fitness\", \"Pearson correlation\": 0.38822915741295855}, {\"gene\": \"nsp15\", \"dnds_metric\": \"log dN/dS\", \"fitness metric\": \"mean fitness\", \"Pearson correlation\": 0.34311944441447234}, {\"gene\": \"nsp16\", \"dnds_metric\": \"dN\", \"fitness metric\": \"mean fitness\", \"Pearson correlation\": 0.4210930688666749}, {\"gene\": \"nsp16\", \"dnds_metric\": \"log dN/dS\", \"fitness metric\": \"mean fitness\", \"Pearson correlation\": 0.4429015464794279}, {\"gene\": \"nsp2\", \"dnds_metric\": \"dN\", \"fitness metric\": \"mean fitness\", \"Pearson correlation\": 0.2867340927976522}, {\"gene\": \"nsp2\", \"dnds_metric\": \"log dN/dS\", \"fitness metric\": \"mean fitness\", \"Pearson correlation\": 0.4219286313753679}, {\"gene\": \"nsp3\", \"dnds_metric\": \"dN\", \"fitness metric\": \"mean fitness\", \"Pearson correlation\": 0.3319946838442439}, {\"gene\": \"nsp3\", \"dnds_metric\": \"log dN/dS\", \"fitness metric\": \"mean fitness\", \"Pearson correlation\": 0.4290134515999794}, {\"gene\": \"nsp4\", \"dnds_metric\": \"dN\", \"fitness metric\": \"mean fitness\", \"Pearson correlation\": 0.3799075066782561}, {\"gene\": \"nsp4\", \"dnds_metric\": \"log dN/dS\", \"fitness metric\": \"mean fitness\", \"Pearson correlation\": 0.4434817811386978}, {\"gene\": \"nsp5 (Mpro)\", \"dnds_metric\": \"dN\", \"fitness metric\": \"mean fitness\", \"Pearson correlation\": 0.35778263607980554}, {\"gene\": \"nsp5 (Mpro)\", \"dnds_metric\": \"log dN/dS\", \"fitness metric\": \"mean fitness\", \"Pearson correlation\": 0.4532650140402962}, {\"gene\": \"nsp6\", \"dnds_metric\": \"dN\", \"fitness metric\": \"mean fitness\", \"Pearson correlation\": 0.3618584508833267}, {\"gene\": \"nsp6\", \"dnds_metric\": \"log dN/dS\", \"fitness metric\": \"mean fitness\", \"Pearson correlation\": 0.39115383671369797}, {\"gene\": \"nsp7\", \"dnds_metric\": \"dN\", \"fitness metric\": \"mean fitness\", \"Pearson correlation\": 0.27186224945509857}, {\"gene\": \"nsp7\", \"dnds_metric\": \"log dN/dS\", \"fitness metric\": \"mean fitness\", \"Pearson correlation\": 0.30692225070467766}, {\"gene\": \"nsp8\", \"dnds_metric\": \"dN\", \"fitness metric\": \"mean fitness\", \"Pearson correlation\": 0.10093126215133863}, {\"gene\": \"nsp8\", \"dnds_metric\": \"log dN/dS\", \"fitness metric\": \"mean fitness\", \"Pearson correlation\": 0.28325163833719047}, {\"gene\": \"nsp9\", \"dnds_metric\": \"dN\", \"fitness metric\": \"mean fitness\", \"Pearson correlation\": 0.18985534514257182}, {\"gene\": \"nsp9\", \"dnds_metric\": \"log dN/dS\", \"fitness metric\": \"mean fitness\", \"Pearson correlation\": 0.31146401822191594}, {\"gene\": \"E\", \"dnds_metric\": \"dN\", \"fitness metric\": \"fitness entropy\", \"Pearson correlation\": 0.19388134028734658}, {\"gene\": \"E\", \"dnds_metric\": \"log dN/dS\", \"fitness metric\": \"fitness entropy\", \"Pearson correlation\": 0.2387618702283814}, {\"gene\": \"M\", \"dnds_metric\": \"dN\", \"fitness metric\": \"fitness entropy\", \"Pearson correlation\": 0.39411238335765986}, {\"gene\": \"M\", \"dnds_metric\": \"log dN/dS\", \"fitness metric\": \"fitness entropy\", \"Pearson correlation\": 0.3294226880521861}, {\"gene\": \"N\", \"dnds_metric\": \"dN\", \"fitness metric\": \"fitness entropy\", \"Pearson correlation\": 0.3775119588157962}, {\"gene\": \"N\", \"dnds_metric\": \"log dN/dS\", \"fitness metric\": \"fitness entropy\", \"Pearson correlation\": 0.49439803711306246}, {\"gene\": \"ORF3a\", \"dnds_metric\": \"dN\", \"fitness metric\": \"fitness entropy\", \"Pearson correlation\": 0.1209137303800936}, {\"gene\": \"ORF3a\", \"dnds_metric\": \"log dN/dS\", \"fitness metric\": \"fitness entropy\", \"Pearson correlation\": 0.17636711188517346}, {\"gene\": \"ORF6\", \"dnds_metric\": \"dN\", \"fitness metric\": \"fitness entropy\", \"Pearson correlation\": 0.26854880578766793}, {\"gene\": \"ORF6\", \"dnds_metric\": \"log dN/dS\", \"fitness metric\": \"fitness entropy\", \"Pearson correlation\": 0.10053238697662877}, {\"gene\": \"ORF7a\", \"dnds_metric\": \"dN\", \"fitness metric\": \"fitness entropy\", \"Pearson correlation\": 0.0725400296930051}, {\"gene\": \"ORF7a\", \"dnds_metric\": \"log dN/dS\", \"fitness metric\": \"fitness entropy\", \"Pearson correlation\": -0.04023423055376647}, {\"gene\": \"ORF8\", \"dnds_metric\": \"dN\", \"fitness metric\": \"fitness entropy\", \"Pearson correlation\": -0.26688403465461863}, {\"gene\": \"ORF8\", \"dnds_metric\": \"log dN/dS\", \"fitness metric\": \"fitness entropy\", \"Pearson correlation\": -0.22823479401991936}, {\"gene\": \"S\", \"dnds_metric\": \"dN\", \"fitness metric\": \"fitness entropy\", \"Pearson correlation\": 0.22655626408812118}, {\"gene\": \"S\", \"dnds_metric\": \"log dN/dS\", \"fitness metric\": \"fitness entropy\", \"Pearson correlation\": 0.38733016248705504}, {\"gene\": \"nsp1\", \"dnds_metric\": \"dN\", \"fitness metric\": \"fitness entropy\", \"Pearson correlation\": 0.119177895718367}, {\"gene\": \"nsp1\", \"dnds_metric\": \"log dN/dS\", \"fitness metric\": \"fitness entropy\", \"Pearson correlation\": 0.14072221001274532}, {\"gene\": \"nsp10\", \"dnds_metric\": \"dN\", \"fitness metric\": \"fitness entropy\", \"Pearson correlation\": 0.11103342836421677}, {\"gene\": \"nsp10\", \"dnds_metric\": \"log dN/dS\", \"fitness metric\": \"fitness entropy\", \"Pearson correlation\": 0.39334848628162256}, {\"gene\": \"nsp12 (RdRp)\", \"dnds_metric\": \"dN\", \"fitness metric\": \"fitness entropy\", \"Pearson correlation\": 0.3913985506109665}, {\"gene\": \"nsp12 (RdRp)\", \"dnds_metric\": \"log dN/dS\", \"fitness metric\": \"fitness entropy\", \"Pearson correlation\": 0.4839417259247163}, {\"gene\": \"nsp13\", \"dnds_metric\": \"dN\", \"fitness metric\": \"fitness entropy\", \"Pearson correlation\": 0.31021264889193123}, {\"gene\": \"nsp13\", \"dnds_metric\": \"log dN/dS\", \"fitness metric\": \"fitness entropy\", \"Pearson correlation\": 0.38698350607272014}, {\"gene\": \"nsp14\", \"dnds_metric\": \"dN\", \"fitness metric\": \"fitness entropy\", \"Pearson correlation\": 0.3152894770520561}, {\"gene\": \"nsp14\", \"dnds_metric\": \"log dN/dS\", \"fitness metric\": \"fitness entropy\", \"Pearson correlation\": 0.4082138856899536}, {\"gene\": \"nsp15\", \"dnds_metric\": \"dN\", \"fitness metric\": \"fitness entropy\", \"Pearson correlation\": 0.37963944754034606}, {\"gene\": \"nsp15\", \"dnds_metric\": \"log dN/dS\", \"fitness metric\": \"fitness entropy\", \"Pearson correlation\": 0.33153806819909765}, {\"gene\": \"nsp16\", \"dnds_metric\": \"dN\", \"fitness metric\": \"fitness entropy\", \"Pearson correlation\": 0.2795785015405517}, {\"gene\": \"nsp16\", \"dnds_metric\": \"log dN/dS\", \"fitness metric\": \"fitness entropy\", \"Pearson correlation\": 0.4062954601591842}, {\"gene\": \"nsp2\", \"dnds_metric\": \"dN\", \"fitness metric\": \"fitness entropy\", \"Pearson correlation\": 0.13966492524911492}, {\"gene\": \"nsp2\", \"dnds_metric\": \"log dN/dS\", \"fitness metric\": \"fitness entropy\", \"Pearson correlation\": 0.35253512850780566}, {\"gene\": \"nsp3\", \"dnds_metric\": \"dN\", \"fitness metric\": \"fitness entropy\", \"Pearson correlation\": 0.29734424869498227}, {\"gene\": \"nsp3\", \"dnds_metric\": \"log dN/dS\", \"fitness metric\": \"fitness entropy\", \"Pearson correlation\": 0.37389736388695827}, {\"gene\": \"nsp4\", \"dnds_metric\": \"dN\", \"fitness metric\": \"fitness entropy\", \"Pearson correlation\": 0.277439889211916}, {\"gene\": \"nsp4\", \"dnds_metric\": \"log dN/dS\", \"fitness metric\": \"fitness entropy\", \"Pearson correlation\": 0.38023257063524674}, {\"gene\": \"nsp5 (Mpro)\", \"dnds_metric\": \"dN\", \"fitness metric\": \"fitness entropy\", \"Pearson correlation\": 0.19054818969342863}, {\"gene\": \"nsp5 (Mpro)\", \"dnds_metric\": \"log dN/dS\", \"fitness metric\": \"fitness entropy\", \"Pearson correlation\": 0.3841569341447376}, {\"gene\": \"nsp6\", \"dnds_metric\": \"dN\", \"fitness metric\": \"fitness entropy\", \"Pearson correlation\": 0.23498584915258808}, {\"gene\": \"nsp6\", \"dnds_metric\": \"log dN/dS\", \"fitness metric\": \"fitness entropy\", \"Pearson correlation\": 0.3630080028553597}, {\"gene\": \"nsp7\", \"dnds_metric\": \"dN\", \"fitness metric\": \"fitness entropy\", \"Pearson correlation\": 0.20855310490537754}, {\"gene\": \"nsp7\", \"dnds_metric\": \"log dN/dS\", \"fitness metric\": \"fitness entropy\", \"Pearson correlation\": 0.3201534156222172}, {\"gene\": \"nsp8\", \"dnds_metric\": \"dN\", \"fitness metric\": \"fitness entropy\", \"Pearson correlation\": 0.11748574135654878}, {\"gene\": \"nsp8\", \"dnds_metric\": \"log dN/dS\", \"fitness metric\": \"fitness entropy\", \"Pearson correlation\": 0.27240512702740677}, {\"gene\": \"nsp9\", \"dnds_metric\": \"dN\", \"fitness metric\": \"fitness entropy\", \"Pearson correlation\": 0.1653451820522668}, {\"gene\": \"nsp9\", \"dnds_metric\": \"log dN/dS\", \"fitness metric\": \"fitness entropy\", \"Pearson correlation\": 0.32740738493158844}]}}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.Chart(...)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# sort by correlation\n",
    "gene_order = (\n",
    "    gene_corrs\n",
    "    .query(\"dnds_metric == 'log dN/dS'\")\n",
    "    .groupby(\"gene\")\n",
    "    .aggregate({\"Pearson correlation\": \"mean\"})\n",
    "    .sort_values(\"Pearson correlation\", ascending=False)\n",
    "    .index\n",
    "    .tolist()\n",
    ")\n",
    "\n",
    "gene_corr_chart = (\n",
    "    alt.Chart(gene_corrs)\n",
    "    .encode(\n",
    "        x=alt.X(\"gene\", scale=alt.Scale(domain=gene_order)),\n",
    "        y=alt.Y(\"Pearson correlation\"),\n",
    "        color=alt.Color(\"fitness metric\"),\n",
    "        shape=alt.Shape(\"fitness metric\"),\n",
    "        tooltip=[\n",
    "            alt.Tooltip(c, format=\".2f\") if gene_corrs[c].dtype == float else c\n",
    "            for c in gene_corrs.columns\n",
    "        ],\n",
    "        facet=alt.Facet(\n",
    "            \"dnds_metric\",\n",
    "            columns=1,\n",
    "            sort=\"descending\",\n",
    "            title=None,\n",
    "            header=alt.Header(labelFontStyle=\"bold\", labelFontSize=13, labelPadding=2),\n",
    "        ),\n",
    "    )\n",
    "    .mark_point(filled=True, size=50)\n",
    "    .properties(width=alt.Step(25), height=175)\n",
    ")\n",
    "\n",
    "gene_corr_chart"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f32b640f-f893-494a-8d28-3f5062a79133",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-28T13:11:01.746286Z",
     "iopub.status.busy": "2023-05-28T13:11:01.746088Z",
     "iopub.status.idle": "2023-05-28T13:11:01.790845Z",
     "shell.execute_reply": "2023-05-28T13:11:01.790120Z",
     "shell.execute_reply.started": "2023-05-28T13:11:01.746265Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\u001b[0;31mInit signature:\u001b[0m\n",
       "\u001b[0malt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mFacet\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\u001b[0m\n",
       "\u001b[0;34m\u001b[0m    \u001b[0mshorthand\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mUndefined\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
       "\u001b[0;34m\u001b[0m    \u001b[0maggregate\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mUndefined\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
       "\u001b[0;34m\u001b[0m    \u001b[0malign\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mUndefined\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
       "\u001b[0;34m\u001b[0m    \u001b[0mband\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mUndefined\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
       "\u001b[0;34m\u001b[0m    \u001b[0mbin\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mUndefined\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
       "\u001b[0;34m\u001b[0m    \u001b[0mbounds\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mUndefined\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
       "\u001b[0;34m\u001b[0m    \u001b[0mcenter\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mUndefined\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
       "\u001b[0;34m\u001b[0m    \u001b[0mcolumns\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mUndefined\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
       "\u001b[0;34m\u001b[0m    \u001b[0mfield\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mUndefined\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
       "\u001b[0;34m\u001b[0m    \u001b[0mheader\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mUndefined\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
       "\u001b[0;34m\u001b[0m    \u001b[0msort\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mUndefined\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
       "\u001b[0;34m\u001b[0m    \u001b[0mspacing\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mUndefined\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
       "\u001b[0;34m\u001b[0m    \u001b[0mtimeUnit\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mUndefined\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
       "\u001b[0;34m\u001b[0m    \u001b[0mtitle\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mUndefined\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
       "\u001b[0;34m\u001b[0m    \u001b[0mtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mUndefined\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
       "\u001b[0;34m\u001b[0m    \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
       "\u001b[0;34m\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
       "\u001b[0;31mDocstring:\u001b[0m     \n",
       "Facet schema wrapper\n",
       "\n",
       "Mapping(required=[shorthand])\n",
       "\n",
       "Attributes\n",
       "----------\n",
       "\n",
       "shorthand : string\n",
       "    shorthand for field, aggregate, and type\n",
       "aggregate : :class:`Aggregate`\n",
       "    Aggregation function for the field (e.g., ``\"mean\"``, ``\"sum\"``, ``\"median\"``,\n",
       "    ``\"min\"``, ``\"max\"``, ``\"count\"`` ).\n",
       "\n",
       "    **Default value:** ``undefined`` (None)\n",
       "\n",
       "    **See also:** `aggregate <https://vega.github.io/vega-lite/docs/aggregate.html>`__\n",
       "    documentation.\n",
       "align : anyOf(:class:`LayoutAlign`, :class:`RowColLayoutAlign`)\n",
       "    The alignment to apply to grid rows and columns. The supported string values are\n",
       "    ``\"all\"``, ``\"each\"``, and ``\"none\"``.\n",
       "\n",
       "\n",
       "    * For ``\"none\"``, a flow layout will be used, in which adjacent subviews are simply\n",
       "      placed one after the other. - For ``\"each\"``, subviews will be aligned into a\n",
       "      clean grid structure, but each row or column may be of variable size. - For\n",
       "      ``\"all\"``, subviews will be aligned and each row or column will be sized\n",
       "      identically based on the maximum observed size. String values for this property\n",
       "      will be applied to both grid rows and columns.\n",
       "\n",
       "    Alternatively, an object value of the form ``{\"row\": string, \"column\": string}`` can\n",
       "    be used to supply different alignments for rows and columns.\n",
       "\n",
       "    **Default value:** ``\"all\"``.\n",
       "band : float\n",
       "    For rect-based marks ( ``rect``, ``bar``, and ``image`` ), mark size relative to\n",
       "    bandwidth of `band scales\n",
       "    <https://vega.github.io/vega-lite/docs/scale.html#band>`__, bins or time units. If\n",
       "    set to ``1``, the mark size is set to the bandwidth, the bin interval, or the time\n",
       "    unit interval. If set to ``0.5``, the mark size is half of the bandwidth or the time\n",
       "    unit interval.\n",
       "\n",
       "    For other marks, relative position on a band of a stacked, binned, time unit or band\n",
       "    scale. If set to ``0``, the marks will be positioned at the beginning of the band.\n",
       "    If set to ``0.5``, the marks will be positioned in the middle of the band.\n",
       "bin : anyOf(boolean, :class:`BinParams`, None)\n",
       "    A flag for binning a ``quantitative`` field, `an object defining binning parameters\n",
       "    <https://vega.github.io/vega-lite/docs/bin.html#params>`__, or indicating that the\n",
       "    data for ``x`` or ``y`` channel are binned before they are imported into Vega-Lite (\n",
       "    ``\"binned\"`` ).\n",
       "\n",
       "\n",
       "    If ``true``, default `binning parameters\n",
       "    <https://vega.github.io/vega-lite/docs/bin.html>`__ will be applied.\n",
       "\n",
       "    If ``\"binned\"``, this indicates that the data for the ``x`` (or ``y`` ) channel are\n",
       "    already binned. You can map the bin-start field to ``x`` (or ``y`` ) and the bin-end\n",
       "    field to ``x2`` (or ``y2`` ). The scale and axis will be formatted similar to\n",
       "    binning in Vega-Lite.  To adjust the axis ticks based on the bin step, you can also\n",
       "    set the axis's `tickMinStep\n",
       "    <https://vega.github.io/vega-lite/docs/axis.html#ticks>`__ property.\n",
       "\n",
       "    **Default value:** ``false``\n",
       "\n",
       "    **See also:** `bin <https://vega.github.io/vega-lite/docs/bin.html>`__\n",
       "    documentation.\n",
       "bounds : enum('full', 'flush')\n",
       "    The bounds calculation method to use for determining the extent of a sub-plot. One\n",
       "    of ``full`` (the default) or ``flush``.\n",
       "\n",
       "\n",
       "    * If set to ``full``, the entire calculated bounds (including axes, title, and\n",
       "      legend) will be used. - If set to ``flush``, only the specified width and height\n",
       "      values for the sub-view will be used. The ``flush`` setting can be useful when\n",
       "      attempting to place sub-plots without axes or legends into a uniform grid\n",
       "      structure.\n",
       "\n",
       "    **Default value:** ``\"full\"``\n",
       "center : anyOf(boolean, :class:`RowColboolean`)\n",
       "    Boolean flag indicating if subviews should be centered relative to their respective\n",
       "    rows or columns.\n",
       "\n",
       "    An object value of the form ``{\"row\": boolean, \"column\": boolean}`` can be used to\n",
       "    supply different centering values for rows and columns.\n",
       "\n",
       "    **Default value:** ``false``\n",
       "columns : float\n",
       "    The number of columns to include in the view composition layout.\n",
       "\n",
       "    **Default value** : ``undefined`` -- An infinite number of columns (a single row)\n",
       "    will be assumed. This is equivalent to ``hconcat`` (for ``concat`` ) and to using\n",
       "    the ``column`` channel (for ``facet`` and ``repeat`` ).\n",
       "\n",
       "    **Note** :\n",
       "\n",
       "    1) This property is only for: - the general (wrappable) ``concat`` operator (not\n",
       "    ``hconcat`` / ``vconcat`` ) - the ``facet`` and ``repeat`` operator with one\n",
       "    field/repetition definition (without row/column nesting)\n",
       "\n",
       "    2) Setting the ``columns`` to ``1`` is equivalent to ``vconcat`` (for ``concat`` )\n",
       "    and to using the ``row`` channel (for ``facet`` and ``repeat`` ).\n",
       "field : :class:`Field`\n",
       "    **Required.** A string defining the name of the field from which to pull a data\n",
       "    value or an object defining iterated values from the `repeat\n",
       "    <https://vega.github.io/vega-lite/docs/repeat.html>`__ operator.\n",
       "\n",
       "    **See also:** `field <https://vega.github.io/vega-lite/docs/field.html>`__\n",
       "    documentation.\n",
       "\n",
       "    **Notes:** 1)  Dots ( ``.`` ) and brackets ( ``[`` and ``]`` ) can be used to access\n",
       "    nested objects (e.g., ``\"field\": \"foo.bar\"`` and ``\"field\": \"foo['bar']\"`` ). If\n",
       "    field names contain dots or brackets but are not nested, you can use ``\\`` to\n",
       "    escape dots and brackets (e.g., ``\"a\\.b\"`` and ``\"a\\[0\\]\"`` ). See more details\n",
       "    about escaping in the `field documentation\n",
       "    <https://vega.github.io/vega-lite/docs/field.html>`__. 2) ``field`` is not required\n",
       "    if ``aggregate`` is ``count``.\n",
       "header : :class:`Header`\n",
       "    An object defining properties of a facet's header.\n",
       "sort : anyOf(:class:`SortArray`, :class:`SortOrder`, :class:`EncodingSortField`, None)\n",
       "    Sort order for the encoded field.\n",
       "\n",
       "    For continuous fields (quantitative or temporal), ``sort`` can be either\n",
       "    ``\"ascending\"`` or ``\"descending\"``.\n",
       "\n",
       "    For discrete fields, ``sort`` can be one of the following: - ``\"ascending\"`` or\n",
       "    ``\"descending\"`` -- for sorting by the values' natural order in JavaScript. - `A\n",
       "    sort field definition\n",
       "    <https://vega.github.io/vega-lite/docs/sort.html#sort-field>`__ for sorting by\n",
       "    another field. - `An array specifying the field values in preferred order\n",
       "    <https://vega.github.io/vega-lite/docs/sort.html#sort-array>`__. In this case, the\n",
       "    sort order will obey the values in the array, followed by any unspecified values in\n",
       "    their original order. For discrete time field, values in the sort array can be\n",
       "    `date-time definition objects <types#datetime>`__. In addition, for time units\n",
       "    ``\"month\"`` and ``\"day\"``, the values can be the month or day names (case\n",
       "    insensitive) or their 3-letter initials (e.g., ``\"Mon\"``, ``\"Tue\"`` ). - ``null``\n",
       "    indicating no sort.\n",
       "\n",
       "    **Default value:** ``\"ascending\"``\n",
       "\n",
       "    **Note:** ``null`` is not supported for ``row`` and ``column``.\n",
       "spacing : anyOf(float, :class:`RowColnumber`)\n",
       "    The spacing in pixels between sub-views of the composition operator. An object of\n",
       "    the form ``{\"row\": number, \"column\": number}`` can be used to set different spacing\n",
       "    values for rows and columns.\n",
       "\n",
       "    **Default value** : Depends on ``\"spacing\"`` property of `the view composition\n",
       "    configuration <https://vega.github.io/vega-lite/docs/config.html#view-config>`__ (\n",
       "    ``20`` by default)\n",
       "timeUnit : anyOf(:class:`TimeUnit`, :class:`TimeUnitParams`)\n",
       "    Time unit (e.g., ``year``, ``yearmonth``, ``month``, ``hours`` ) for a temporal\n",
       "    field. or `a temporal field that gets casted as ordinal\n",
       "    <https://vega.github.io/vega-lite/docs/type.html#cast>`__.\n",
       "\n",
       "    **Default value:** ``undefined`` (None)\n",
       "\n",
       "    **See also:** `timeUnit <https://vega.github.io/vega-lite/docs/timeunit.html>`__\n",
       "    documentation.\n",
       "title : anyOf(:class:`Text`, None)\n",
       "    A title for the field. If ``null``, the title will be removed.\n",
       "\n",
       "    **Default value:**  derived from the field's name and transformation function (\n",
       "    ``aggregate``, ``bin`` and ``timeUnit`` ). If the field has an aggregate function,\n",
       "    the function is displayed as part of the title (e.g., ``\"Sum of Profit\"`` ). If the\n",
       "    field is binned or has a time unit applied, the applied function is shown in\n",
       "    parentheses (e.g., ``\"Profit (binned)\"``, ``\"Transaction Date (year-month)\"`` ).\n",
       "    Otherwise, the title is simply the field name.\n",
       "\n",
       "    **Notes** :\n",
       "\n",
       "    1) You can customize the default field title format by providing the `fieldTitle\n",
       "    <https://vega.github.io/vega-lite/docs/config.html#top-level-config>`__ property in\n",
       "    the `config <https://vega.github.io/vega-lite/docs/config.html>`__ or `fieldTitle\n",
       "    function via the compile function's options\n",
       "    <https://vega.github.io/vega-lite/docs/compile.html#field-title>`__.\n",
       "\n",
       "    2) If both field definition's ``title`` and axis, header, or legend ``title`` are\n",
       "    defined, axis/header/legend title will be used.\n",
       "type : :class:`StandardType`\n",
       "    The type of measurement ( ``\"quantitative\"``, ``\"temporal\"``, ``\"ordinal\"``, or\n",
       "    ``\"nominal\"`` ) for the encoded field or constant value ( ``datum`` ). It can also\n",
       "    be a ``\"geojson\"`` type for encoding `'geoshape'\n",
       "    <https://vega.github.io/vega-lite/docs/geoshape.html>`__.\n",
       "\n",
       "    Vega-Lite automatically infers data types in many cases as discussed below. However,\n",
       "    type is required for a field if: (1) the field is not nominal and the field encoding\n",
       "    has no specified ``aggregate`` (except ``argmin`` and ``argmax`` ), ``bin``, scale\n",
       "    type, custom ``sort`` order, nor ``timeUnit`` or (2) if you wish to use an ordinal\n",
       "    scale for a field with ``bin`` or ``timeUnit``.\n",
       "\n",
       "    **Default value:**\n",
       "\n",
       "    1) For a data ``field``, ``\"nominal\"`` is the default data type unless the field\n",
       "    encoding has ``aggregate``, ``channel``, ``bin``, scale type, ``sort``, or\n",
       "    ``timeUnit`` that satisfies the following criteria: - ``\"quantitative\"`` is the\n",
       "    default type if (1) the encoded field contains ``bin`` or ``aggregate`` except\n",
       "    ``\"argmin\"`` and ``\"argmax\"``, (2) the encoding channel is ``latitude`` or\n",
       "    ``longitude`` channel or (3) if the specified scale type is `a quantitative scale\n",
       "    <https://vega.github.io/vega-lite/docs/scale.html#type>`__. - ``\"temporal\"`` is the\n",
       "    default type if (1) the encoded field contains ``timeUnit`` or (2) the specified\n",
       "    scale type is a time or utc scale - ``ordinal\"\"`` is the default type if (1) the\n",
       "    encoded field contains a `custom sort order\n",
       "    <https://vega.github.io/vega-lite/docs/sort.html#specifying-custom-sort-order>`__,\n",
       "    (2) the specified scale type is an ordinal/point/band scale, or (3) the encoding\n",
       "    channel is ``order``.\n",
       "\n",
       "    2) For a constant value in data domain ( ``datum`` ): - ``\"quantitative\"`` if the\n",
       "    datum is a number - ``\"nominal\"`` if the datum is a string - ``\"temporal\"`` if the\n",
       "    datum is `a date time object\n",
       "    <https://vega.github.io/vega-lite/docs/datetime.html>`__\n",
       "\n",
       "    **Note:** - Data ``type`` describes the semantics of the data rather than the\n",
       "    primitive data types (number, string, etc.). The same primitive data type can have\n",
       "    different types of measurement. For example, numeric data can represent\n",
       "    quantitative, ordinal, or nominal data. - Data values for a temporal field can be\n",
       "    either a date-time string (e.g., ``\"2015-03-07 12:32:17\"``, ``\"17:01\"``,\n",
       "    ``\"2015-03-16\"``. ``\"2015\"`` ) or a timestamp number (e.g., ``1552199579097`` ). -\n",
       "    When using with `bin <https://vega.github.io/vega-lite/docs/bin.html>`__, the\n",
       "    ``type`` property can be either ``\"quantitative\"`` (for using a linear bin scale) or\n",
       "    `\"ordinal\" (for using an ordinal bin scale)\n",
       "    <https://vega.github.io/vega-lite/docs/type.html#cast-bin>`__. - When using with\n",
       "    `timeUnit <https://vega.github.io/vega-lite/docs/timeunit.html>`__, the ``type``\n",
       "    property can be either ``\"temporal\"`` (default, for using a temporal scale) or\n",
       "    `\"ordinal\" (for using an ordinal scale)\n",
       "    <https://vega.github.io/vega-lite/docs/type.html#cast-bin>`__. - When using with\n",
       "    `aggregate <https://vega.github.io/vega-lite/docs/aggregate.html>`__, the ``type``\n",
       "    property refers to the post-aggregation data type. For example, we can calculate\n",
       "    count ``distinct`` of a categorical field ``\"cat\"`` using ``{\"aggregate\":\n",
       "    \"distinct\", \"field\": \"cat\"}``. The ``\"type\"`` of the aggregate output is\n",
       "    ``\"quantitative\"``. - Secondary channels (e.g., ``x2``, ``y2``, ``xError``,\n",
       "    ``yError`` ) do not have ``type`` as they must have exactly the same type as their\n",
       "    primary channels (e.g., ``x``, ``y`` ).\n",
       "\n",
       "    **See also:** `type <https://vega.github.io/vega-lite/docs/type.html>`__\n",
       "    documentation.\n",
       "\u001b[0;31mFile:\u001b[0m           /fh/fast/bloom_j/software/miniconda3/envs/SARS2-mut-fitness/lib/python3.9/site-packages/altair/vegalite/v4/schema/channels.py\n",
       "\u001b[0;31mType:\u001b[0m           type\n",
       "\u001b[0;31mSubclasses:\u001b[0m     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "?alt.Facet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec608e48-d582-4e10-b0cd-390407c0dfdb",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
