{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2fe25e19-bdfa-47cc-a88e-1a6b8730bce9",
   "metadata": {},
   "source": [
    "# Analyze comparator studies versus fitness estimates and deep mutational scanning for spike"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7b91b627-e2dc-40e4-809e-a710fb027642",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-30T17:54:48.774739Z",
     "iopub.status.busy": "2023-05-30T17:54:48.774315Z",
     "iopub.status.idle": "2023-05-30T17:54:49.349230Z",
     "shell.execute_reply": "2023-05-30T17:54:49.347407Z",
     "shell.execute_reply.started": "2023-05-30T17:54:48.774710Z"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "######## snakemake preamble start (automatically inserted, do not edit) ########\n",
    "import sys; sys.path.extend(['/fh/fast/bloom_j/software/miniconda3/envs/SARS2-mut-fitness/lib/python3.9/site-packages', '/home/jbloom/.cache/snakemake/snakemake/source-cache/runtime-cache/tmp8hhv_h_m/file/fh/fast/bloom_j/computational_notebooks/jbloom/2023/SARS2-mut-fitness/notebooks', '/fh/fast/bloom_j/computational_notebooks/jbloom/2023/SARS2-mut-fitness/notebooks']); import pickle; snakemake = pickle.loads(b'\\x80\\x04\\x95f\\x1a\\x00\\x00\\x00\\x00\\x00\\x00\\x8c\\x10snakemake.script\\x94\\x8c\\tSnakemake\\x94\\x93\\x94)\\x81\\x94}\\x94(\\x8c\\x05input\\x94\\x8c\\x0csnakemake.io\\x94\\x8c\\nInputFiles\\x94\\x93\\x94)\\x81\\x94(\\x8c3results_public_2023-05-11/aa_fitness/aa_fitness.csv\\x94\\x8c/data/comparator_studies/rodriguez_rivas_dca.csv\\x94\\x8c)data/comparator_studies/maher_drivers.csv\\x94\\x8c,data/comparator_studies/thadani_learning.csv\\x94\\x8c?results_public_2023-05-11/dms/dadonaite_ba1_spike/processed.csv\\x94\\x8c5results_public_2023-05-11/dms/starr_rbd/processed.csv\\x94\\x8c8results_public_2023-05-11/dms/iketani_mpro/processed.csv\\x94\\x8c;results_public_2023-05-11/dms/flynn_mpro_2022/processed.csv\\x94\\x8c;results_public_2023-05-11/dms/flynn_mpro_2023/processed.csv\\x94e}\\x94(\\x8c\\x06_names\\x94}\\x94(\\x8c\\naa_fitness\\x94K\\x00N\\x86\\x94\\x8c\\x13rodriguez_rivas_dca\\x94K\\x01N\\x86\\x94\\x8c\\rmaher_drivers\\x94K\\x02N\\x86\\x94\\x8c\\x10thadani_learning\\x94K\\x03N\\x86\\x94\\x8c\\x13dadonaite_ba1_spike\\x94K\\x04N\\x86\\x94\\x8c\\tstarr_rbd\\x94K\\x05N\\x86\\x94\\x8c\\x0ciketani_mpro\\x94K\\x06N\\x86\\x94\\x8c\\x0fflynn_mpro_2022\\x94K\\x07N\\x86\\x94\\x8c\\x0fflynn_mpro_2023\\x94K\\x08N\\x86\\x94u\\x8c\\x12_allowed_overrides\\x94]\\x94(\\x8c\\x05index\\x94\\x8c\\x04sort\\x94eh*\\x8c\\tfunctools\\x94\\x8c\\x07partial\\x94\\x93\\x94h\\x06\\x8c\\x19Namedlist._used_attribute\\x94\\x93\\x94\\x85\\x94R\\x94(h0)}\\x94\\x8c\\x05_name\\x94h*sNt\\x94bh+h.h0\\x85\\x94R\\x94(h0)}\\x94h4h+sNt\\x94bh\\x16h\\nh\\x18h\\x0bh\\x1ah\\x0ch\\x1ch\\rh\\x1eh\\x0eh h\\x0fh\"h\\x10h$h\\x11h&h\\x12ub\\x8c\\x06output\\x94h\\x06\\x8c\\x0bOutputFiles\\x94\\x93\\x94)\\x81\\x94\\x8cAresults_public_2023-05-11/comparator_studies/comparator_corr.html\\x94a}\\x94(h\\x14}\\x94\\x8c\\tcorr_html\\x94K\\x00N\\x86\\x94sh(]\\x94(h*h+eh*h.h0\\x85\\x94R\\x94(h0)}\\x94h4h*sNt\\x94bh+h.h0\\x85\\x94R\\x94(h0)}\\x94h4h+sNt\\x94bhAh>ub\\x8c\\x06params\\x94h\\x06\\x8c\\x06Params\\x94\\x93\\x94)\\x81\\x94(K\\n}\\x94(h\\x1e}\\x94(\\x8c\\x05study\\x94\\x8c;https://www.biorxiv.org/content/10.1101/2022.10.13.512056v1\\x94\\x8c\\x0bdescription\\x94\\x8c\\x1dspike (Dadonaite et al, 2022)\\x94\\x8c\\x04gene\\x94\\x8c\\x01S\\x94\\x8c\\x03url\\x94\\x8c\\x8bhttps://raw.githubusercontent.com/dms-vep/SARS-CoV-2_Omicron_BA.1_spike_DMS_mAbs/main/results/muteffects_functional/muteffects_observed.csv\\x94\\x8c\\x0bfilter_cols\\x94}\\x94\\x8c\\ntimes_seen\\x94K\\x03suh }\\x94(\\x8c\\x05study\\x94\\x8cOhttps://journals.plos.org/plospathogens/article?id=10.1371/journal.ppat.1010951\\x94\\x8c\\x0bdescription\\x94\\x8c\\x17RBD (Starr et al, 2022)\\x94\\x8c\\x04gene\\x94hW\\x8c\\x03url\\x94\\x8c\\xadhttps://media.githubusercontent.com/media/jbloomlab/SARS-CoV-2-RBD_DMS_Omicron/f76ba2b2bec18ede9fa9da18c9ccc52389b1ba3a/results/final_variant_scores/final_variant_scores.csv\\x94uh\"}\\x94(\\x8c\\x05study\\x94\\x8cChttps://www.sciencedirect.com/science/article/pii/S1931312822004024\\x94\\x8c\\x0bdescription\\x94\\x8c\\x1aMpro (Iketani et al, 2022)\\x94\\x8c\\x04gene\\x94\\x8c\\x0bnsp5 (Mpro)\\x94\\x8c\\x03url\\x94\\x8c\\xachttps://raw.githubusercontent.com/alejandrochavezlab/3CL_protease_DMS/dc802ee3ea9d43005ca97a28af078767fd66777c/outputs/results/normalized_to_wt_and_stop/activity_scores.csv\\x94\\x8c\\x06wt_seq\\x94\\x8c\\x0cdata/Mpro.fa\\x94uh$}\\x94(\\x8c\\x05study\\x94\\x8c(https://elifesciences.org/articles/77433\\x94\\x8c\\x0bdescription\\x94\\x8c\\x18Mpro (Flynn et al, 2022)\\x94\\x8c\\x04gene\\x94\\x8c\\x0bnsp5 (Mpro)\\x94\\x8c\\x03url\\x94\\x8cKhttps://cdn.elifesciences.org/articles/77433/elife-77433-fig2-data1-v2.xlsx\\x94\\x8c\\x06wt_seq\\x94\\x8c\\x0cdata/Mpro.fa\\x94uh&}\\x94(\\x8c\\x05study\\x94\\x8c)https://doi.org/10.1101/2023.03.02.530652\\x94\\x8c\\x0bdescription\\x94\\x8c\\x18Mpro (Flynn et al, 2023)\\x94\\x8c\\x04gene\\x94\\x8c\\x0bnsp5 (Mpro)\\x94\\x8c\\x03url\\x94\\x8cahttps://www.biorxiv.org/content/biorxiv/early/2023/03/02/2023.03.02.530652/DC1/embed/media-1.xlsx\\x94uu}\\x94(h\\x18}\\x94(\\x8c\\x04data\\x94\\x8c_https://raw.githubusercontent.com/GiancarloCroce/DCA_SARS-CoV-2/main/data/data_dca_proteome.csv\\x94\\x8c\\x04name\\x94\\x8c\\x1cRodriguez-Rivas et al (2022)\\x94uh\\x1a}\\x94(\\x8c\\x04data\\x94\\x8c>https://www.science.org/doi/suppl/10.1126/scitranslmed.abk3445\\x94\\x8c\\x04name\\x94\\x8c\\x12Maher et al (2022)\\x94uh\\x1c}\\x94(\\x8c\\x04data\\x94\\x8c`https://www.biorxiv.org/content/biorxiv/early/2023/04/12/2022.07.21.501023/DC4/embed/media-4.zip\\x94\\x8c\\x04name\\x94\\x8c\\x14Thadani et al (2023)\\x94uue}\\x94(h\\x14}\\x94(\\x8c\\x12min_expected_count\\x94K\\x00N\\x86\\x94\\x8c\\x0cdms_datasets\\x94K\\x01N\\x86\\x94\\x8c\\x12comparator_studies\\x94K\\x02N\\x86\\x94uh(]\\x94(h*h+eh*h.h0\\x85\\x94R\\x94(h0)}\\x94h4h*sNt\\x94bh+h.h0\\x85\\x94R\\x94(h0)}\\x94h4h+sNt\\x94bh\\x96K\\nh\\x98hPh\\x9ah\\x84ub\\x8c\\twildcards\\x94h\\x06\\x8c\\tWildcards\\x94\\x93\\x94)\\x81\\x94\\x8c\\x11public_2023-05-11\\x94a}\\x94(h\\x14}\\x94\\x8c\\x03mat\\x94K\\x00N\\x86\\x94sh(]\\x94(h*h+eh*h.h0\\x85\\x94R\\x94(h0)}\\x94h4h*sNt\\x94bh+h.h0\\x85\\x94R\\x94(h0)}\\x94h4h+sNt\\x94b\\x8c\\x03mat\\x94h\\xa9ub\\x8c\\x07threads\\x94K\\x01\\x8c\\tresources\\x94h\\x06\\x8c\\tResources\\x94\\x93\\x94)\\x81\\x94(K\\x01K\\x01\\x8c\\x04/tmp\\x94e}\\x94(h\\x14}\\x94(\\x8c\\x06_cores\\x94K\\x00N\\x86\\x94\\x8c\\x06_nodes\\x94K\\x01N\\x86\\x94\\x8c\\x06tmpdir\\x94K\\x02N\\x86\\x94uh(]\\x94(h*h+eh*h.h0\\x85\\x94R\\x94(h0)}\\x94h4h*sNt\\x94bh+h.h0\\x85\\x94R\\x94(h0)}\\x94h4h+sNt\\x94bh\\xc0K\\x01h\\xc2K\\x01h\\xc4h\\xbdub\\x8c\\x03log\\x94h\\x06\\x8c\\x03Log\\x94\\x93\\x94)\\x81\\x94\\x8cMresults_public_2023-05-11/comparator_studies/analyze_comparator_studies.ipynb\\x94a}\\x94(h\\x14}\\x94\\x8c\\x08notebook\\x94K\\x00N\\x86\\x94sh(]\\x94(h*h+eh*h.h0\\x85\\x94R\\x94(h0)}\\x94h4h*sNt\\x94bh+h.h0\\x85\\x94R\\x94(h0)}\\x94h4h+sNt\\x94bh\\xd6h\\xd3ub\\x8c\\x06config\\x94}\\x94(\\x8c\\tmat_trees\\x94}\\x94(\\x8c\\x11public_2023-05-11\\x94\\x8c\\x83http://hgdownload.soe.ucsc.edu/goldenPath/wuhCor1/UShER_SARS-CoV-2/2023/05/11/public-2023-05-11.all.masked.nextclade.pangolin.pb.gz\\x94\\x8c\\x11public_2023-03-27\\x94\\x8c\\x83http://hgdownload.soe.ucsc.edu/goldenPath/wuhCor1/UShER_SARS-CoV-2/2023/03/27/public-2023-03-27.all.masked.nextclade.pangolin.pb.gz\\x94\\x8c\\x11public_2022-12-18\\x94\\x8c\\x83http://hgdownload.soe.ucsc.edu/goldenPath/wuhCor1/UShER_SARS-CoV-2/2022/12/18/public-2022-12-18.all.masked.nextclade.pangolin.pb.gz\\x94\\x8c\\x11gisaid_2023-03-29\\x94\\x8cQhttps://hgwdev.gi.ucsc.edu/~angie/2847b92/gisaidAndPublic.2023-03-29.masked.pb.gz\\x94u\\x8c\\x0bcurrent_mat\\x94\\x8c\\x11public_2023-05-11\\x94\\x8c\\tref_fasta\\x94\\x8cGhttp://hgdownload.soe.ucsc.edu/goldenPath/wuhCor1/bigZips/wuhCor1.fa.gz\\x94\\x8c\\x07ref_gtf\\x94\\x8cPhttp://hgdownload.soe.ucsc.edu/goldenPath/wuhCor1/bigZips/genes/ncbiGenes.gtf.gz\\x94\\x8c\\x0eadd_to_ref_gtf\\x94}\\x94\\x8c\\x05ORF9b\\x94]\\x94(M|nM\\xa1oes\\x8c\\x11min_clade_samples\\x94M\\x10\\'\\x8c\\x0esample_subsets\\x94}\\x94(\\x8c\\x03all\\x94\\x8c\\x01.\\x94\\x8c\\x03USA\\x94\\x8c\\x03USA\\x94\\x8c\\x07England\\x94\\x8c\\x07England\\x94u\\x8c\\x12clade_founder_json\\x94\\x8cyhttps://raw.githubusercontent.com/neherlab/SC2_variant_rates/7e738194a8c6592082f1caa9a6ca70cb68289790/data/clade_gts.json\\x94\\x8c\\x11clades_to_exclude\\x94]\\x94\\x8c\\x0321A\\x94a\\x8c\\x10max_nt_mutations\\x94K\\x04\\x8c\\x15max_reversions_to_ref\\x94K\\x01\\x8c\\x1fmax_reversions_to_clade_founder\\x94K\\x01\\x8c\\x1bexclude_ref_to_founder_muts\\x94\\x88\\x8c\\x12site_include_range\\x94}\\x94(\\x8c\\x05start\\x94M\\n\\x01\\x8c\\x03end\\x94M\\xeasu\\x8c\\x10sites_to_exclude\\x94]\\x94(K\\x99M}\\x04M\\x96\\x08MI\\x0cM\\xec\\rM\\xc2\\x0eM\\xd2\\x0fMo\\x18MV\\x1fMV\"MZ4M{6M\\x12YME_MF_MeaM\\xfd\\x15M\\xc3\\x1aM\\xa0\\x1cM\\xbfmM\\xb2re\\x8c\\rsite_mask_vcf\\x94\\x8cvhttps://raw.githubusercontent.com/W-L/ProblematicSites_SARS-CoV2/master/subset_vcf/problematic_sites_sarsCov2.mask.vcf\\x94\\x8c\\x12usher_masked_sites\\x94\\x8c\\x1cdata/usher_masked_sites.yaml\\x94\\x8c\\x1dsynonymous_spectra_min_counts\\x94M\\x88\\x13\\x8c\\x0eorf1ab_to_nsps\\x94}\\x94(\\x8c\\x04nsp1\\x94]\\x94(K\\x01K\\xb4e\\x8c\\x04nsp2\\x94]\\x94(K\\xb5M2\\x03e\\x8c\\x04nsp3\\x94]\\x94(M3\\x03M\\xcb\\ne\\x8c\\x04nsp4\\x94]\\x94(M\\xcc\\nM\\xbf\\x0ce\\x8c\\x0bnsp5 (Mpro)\\x94]\\x94(M\\xc0\\x0cM\\xf1\\re\\x8c\\x04nsp6\\x94]\\x94(M\\xf2\\rM\\x13\\x0fe\\x8c\\x04nsp7\\x94]\\x94(M\\x14\\x0fMf\\x0fe\\x8c\\x04nsp8\\x94]\\x94(Mg\\x0fM,\\x10e\\x8c\\x04nsp9\\x94]\\x94(M-\\x10M\\x9d\\x10e\\x8c\\x05nsp10\\x94]\\x94(M\\x9e\\x10M(\\x11e\\x8c\\x0cnsp12 (RdRp)\\x94]\\x94(M)\\x11M\\xcc\\x14e\\x8c\\x05nsp13\\x94]\\x94(M\\xcd\\x14M%\\x17e\\x8c\\x05nsp14\\x94]\\x94(M&\\x17M4\\x19e\\x8c\\x05nsp15\\x94]\\x94(M5\\x19M\\x8e\\x1ae\\x8c\\x05nsp16\\x94]\\x94(M\\x8f\\x1aM\\xb8\\x1beu\\x8c\\rgene_overlaps\\x94}\\x94(\\x8c\\x07exclude\\x94]\\x94(]\\x94(\\x8c\\x05ORF1a\\x94\\x8c\\x06ORF1ab\\x94e]\\x94(\\x8c\\x05ORF7a\\x94\\x8c\\x05ORF7b\\x94ee\\x8c\\x06retain\\x94]\\x94]\\x94(\\x8c\\x01N\\x94\\x8c\\x05ORF9b\\x94eau\\x8c\\x13fitness_pseudocount\\x94G?\\xe0\\x00\\x00\\x00\\x00\\x00\\x00\\x8c\\x12min_expected_count\\x94K\\n\\x8c\\x14clade_corr_min_count\\x94J \\xa1\\x07\\x00\\x8c\\x19aa_fitness_init_ref_clade\\x94\\x8c\\x0319A\\x94\\x8c!aa_fitness_heatmap_minimal_domain\\x94]\\x94(J\\xfa\\xff\\xff\\xffK\\x02e\\x8c\\x0eclade_synonyms\\x94}\\x94(\\x8c\\x0319A\\x94\\x8c\\x01B\\x94\\x8c\\x0320A\\x94\\x8c\\x03B.1\\x94\\x8c\\x0320B\\x94\\x8c\\x05B.1.1\\x94\\x8c\\x0320C\\x94\\x8c\\x07B.1.367\\x94\\x8c\\x0320E\\x94\\x8c\\x07B.1.177\\x94\\x8c\\x0320F\\x94\\x8c\\x03D.2\\x94\\x8c\\x0320G\\x94\\x8c\\x05B.1.2\\x94\\x8c\\x0320I\\x94\\x8c\\x05Alpha\\x94\\x8c\\x0320J\\x94\\x8c\\x05Gamma\\x94\\x8c\\x0321C\\x94\\x8c\\x07Epsilon\\x94\\x8c\\x0321F\\x94\\x8c\\x04Iota\\x94\\x8c\\x0321I\\x94\\x8c\\x05Delta\\x94\\x8c\\x0321J\\x94\\x8c\\x05Delta\\x94\\x8c\\x0321K\\x94\\x8c\\x0cOmicron BA.1\\x94\\x8c\\x0321L\\x94\\x8c\\x0cOmicron BA.2\\x94\\x8c\\x0322A\\x94\\x8c\\x0cOmicron BA.4\\x94\\x8c\\x0322B\\x94\\x8c\\x0cOmicron BA.5\\x94\\x8c\\x0322C\\x94\\x8c\\x11Omicron BA.2.12.1\\x94\\x8c\\x0322D\\x94\\x8c\\x0fOmicron BA.2.75\\x94\\x8c\\x0322E\\x94\\x8c\\x0cOmicron BQ.1\\x94u\\x8c\\x0cdms_datasets\\x94hP\\x8c\\rneher_fitness\\x94\\x8c\\x19data/Neher_aa_fitness.csv\\x94\\x8c\\x14clade_fixed_muts_ref\\x94\\x8c\\x0319A\\x94\\x8c\\x14terminal_pseudocount\\x94G?\\xe0\\x00\\x00\\x00\\x00\\x00\\x00\\x8c\\x19terminal_min_actual_count\\x94K\\x05\\x8c\\x04dnds\\x94\\x8c^https://raw.githubusercontent.com/spond/SARS-CoV-2-variation/master/windowed-sites-fel-all.csv\\x94\\x8c\\x12comparator_studies\\x94h\\x84\\x8c\\x08docs_url\\x94\\x8c-https://jbloomlab.github.io/SARS2-mut-fitness\\x94\\x8c\\x15docs_plot_annotations\\x94\\x8c\\x1fdata/docs_plot_annotations.yaml\\x94u\\x8c\\x04rule\\x94\\x8c\\x1aanalyze_comparator_studies\\x94\\x8c\\x0fbench_iteration\\x94N\\x8c\\tscriptdir\\x94\\x8cP/fh/fast/bloom_j/computational_notebooks/jbloom/2023/SARS2-mut-fitness/notebooks\\x94ub.'); from snakemake.logging import logger; logger.printshellcmds = False; import os; os.chdir(r'/fh/fast/bloom_j/computational_notebooks/jbloom/2023/SARS2-mut-fitness');\n",
    "######## snakemake preamble end #########\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "bf1d9c69-175a-4c98-a584-96666ccaf301",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-30T17:54:49.353758Z",
     "iopub.status.busy": "2023-05-30T17:54:49.353300Z",
     "iopub.status.idle": "2023-05-30T17:54:51.635048Z",
     "shell.execute_reply": "2023-05-30T17:54:51.633188Z",
     "shell.execute_reply.started": "2023-05-30T17:54:49.353728Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "import altair as alt\n",
    "\n",
    "import numpy\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "import scipy.stats.mstats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2793260f-6dd8-498a-b892-bb36a5330d3b",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-30T17:54:51.640550Z",
     "iopub.status.busy": "2023-05-30T17:54:51.639980Z",
     "iopub.status.idle": "2023-05-30T17:54:51.690843Z",
     "shell.execute_reply": "2023-05-30T17:54:51.688548Z",
     "shell.execute_reply.started": "2023-05-30T17:54:51.640524Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# get variables from `snakemake`\n",
    "aa_fitness_csv = snakemake.input.aa_fitness\n",
    "corr_html = snakemake.output.corr_html\n",
    "min_expected_count = snakemake.params.min_expected_count\n",
    "comparator_studies = snakemake.params.comparator_studies\n",
    "\n",
    "# get DMS datasets with added entry for data\n",
    "dms_datasets = snakemake.params.dms_datasets\n",
    "for dataset in dms_datasets:\n",
    "    dms_datasets[dataset][\"data\"] = pd.read_csv(getattr(snakemake.input, dataset))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4ea92fe-918f-4e4b-a735-910e80ea0143",
   "metadata": {},
   "source": [
    "Read data from comparator studies:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7c80fddf-8cab-4096-b2d4-ba455da7305e",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-30T17:54:51.701425Z",
     "iopub.status.busy": "2023-05-30T17:54:51.700978Z",
     "iopub.status.idle": "2023-05-30T17:54:52.834903Z",
     "shell.execute_reply": "2023-05-30T17:54:52.830098Z",
     "shell.execute_reply.started": "2023-05-30T17:54:51.701388Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading data from data/comparator_studies/rodriguez_rivas_dca.csv\n",
      "Reading data from data/comparator_studies/maher_drivers.csv\n",
      "Reading data from data/comparator_studies/thadani_learning.csv\n"
     ]
    }
   ],
   "source": [
    "for study, study_d in comparator_studies.items():\n",
    "    \n",
    "    datafile = getattr(snakemake.input, study)\n",
    "    print(f\"Reading data from {datafile}\")\n",
    "    \n",
    "    if study == \"rodriguez_rivas_dca\":\n",
    "        comparator_studies[study][\"df\"] = (\n",
    "            pd.read_csv(datafile)\n",
    "            .rename(\n",
    "                columns={\n",
    "                    \"position_protein\": \"site\",\n",
    "                    \"mutability_score(DCA)\": \"DCA mutability score\",\n",
    "                }\n",
    "            )\n",
    "            .query(\"protein == 'Spike'\")\n",
    "            .assign(gene=\"S\")\n",
    "            [[\"gene\", \"site\", \"DCA mutability score\"]]\n",
    "            .query(\"`DCA mutability score`.notnull()\")\n",
    "            .reset_index(drop=True)\n",
    "        )   \n",
    "        \n",
    "    elif study == \"maher_drivers\":\n",
    "        comparator_studies[study][\"df\"] = (\n",
    "            pd.read_csv(\"data/comparator_studies/maher_drivers.csv\")\n",
    "            .assign(\n",
    "                gene=lambda x: x[\"Unnamed: 0\"].str.split(\"_\").str[0],\n",
    "                mutation=lambda x: x[\"Unnamed: 0\"].str.split(\"_\", 1).str[1],\n",
    "            )\n",
    "            .query(\"gene == 'Spike'\")\n",
    "            .assign(gene=\"S\")\n",
    "            .query(\"mutation.str.fullmatch('[A-Z]\\d+[A-Z]')\")\n",
    "            .assign(\n",
    "                wildtype=lambda x: x[\"mutation\"].str[0],\n",
    "                site=lambda x: x[\"mutation\"].str[1: -1].astype(int),\n",
    "                aa=lambda x: x[\"mutation\"].str[-1],\n",
    "            )\n",
    "            .rename(columns={\"EpiScore_AllLineages\": \"EpiScore\"})\n",
    "            .groupby([\"gene\", \"site\", \"aa\"], as_index=False)\n",
    "            .aggregate({\"EpiScore\": \"mean\"})\n",
    "            .reset_index(drop=True)\n",
    "        )\n",
    "        \n",
    "    elif study == \"thadani_learning\":\n",
    "        comparator_studies[study][\"df\"] = (\n",
    "            pd.read_csv(\"data/comparator_studies/thadani_learning.csv\")\n",
    "            .rename(columns={\"i\": \"site\", \"mut\": \"aa\", \"eve\": \"EVE score\"})\n",
    "            .assign(gene=\"S\")\n",
    "            [[\"gene\", \"site\", \"aa\", \"EVE score\"]]\n",
    "    \n",
    "        )\n",
    "\n",
    "    else:\n",
    "        raise ValueError(f\"invalid {study=}\")\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a2d4bf0-b5dc-4fa9-959f-cb415e63d447",
   "metadata": {},
   "source": [
    "Read amino-acid fitnesses, and for each site calculate the mean fitness of mutations at each site."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "95b09966-e7f3-467e-9386-512d566bf719",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-30T17:54:52.839983Z",
     "iopub.status.busy": "2023-05-30T17:54:52.839548Z",
     "iopub.status.idle": "2023-05-30T17:54:53.088869Z",
     "shell.execute_reply": "2023-05-30T17:54:53.087610Z",
     "shell.execute_reply.started": "2023-05-30T17:54:52.839931Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fitnesses for amino acids:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>site</th>\n",
       "      <th>aa</th>\n",
       "      <th>fitness</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>61583</th>\n",
       "      <td>S</td>\n",
       "      <td>1</td>\n",
       "      <td>I</td>\n",
       "      <td>-6.92550</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61584</th>\n",
       "      <td>S</td>\n",
       "      <td>1</td>\n",
       "      <td>K</td>\n",
       "      <td>-3.37120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61585</th>\n",
       "      <td>S</td>\n",
       "      <td>1</td>\n",
       "      <td>L</td>\n",
       "      <td>-4.09100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61586</th>\n",
       "      <td>S</td>\n",
       "      <td>1</td>\n",
       "      <td>M</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61588</th>\n",
       "      <td>S</td>\n",
       "      <td>1</td>\n",
       "      <td>T</td>\n",
       "      <td>-4.89210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70889</th>\n",
       "      <td>S</td>\n",
       "      <td>1272</td>\n",
       "      <td>C</td>\n",
       "      <td>-3.38030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70891</th>\n",
       "      <td>S</td>\n",
       "      <td>1272</td>\n",
       "      <td>F</td>\n",
       "      <td>-2.63510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70892</th>\n",
       "      <td>S</td>\n",
       "      <td>1272</td>\n",
       "      <td>H</td>\n",
       "      <td>-0.42621</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70893</th>\n",
       "      <td>S</td>\n",
       "      <td>1272</td>\n",
       "      <td>N</td>\n",
       "      <td>-3.37120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70895</th>\n",
       "      <td>S</td>\n",
       "      <td>1272</td>\n",
       "      <td>Y</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7376 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene  site aa  fitness\n",
       "61583    S     1  I -6.92550\n",
       "61584    S     1  K -3.37120\n",
       "61585    S     1  L -4.09100\n",
       "61586    S     1  M  0.00000\n",
       "61588    S     1  T -4.89210\n",
       "...    ...   ... ..      ...\n",
       "70889    S  1272  C -3.38030\n",
       "70891    S  1272  F -2.63510\n",
       "70892    S  1272  H -0.42621\n",
       "70893    S  1272  N -3.37120\n",
       "70895    S  1272  Y  0.00000\n",
       "\n",
       "[7376 rows x 4 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Mean fitnesses for sites:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>site</th>\n",
       "      <th>fitness</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>S</td>\n",
       "      <td>1</td>\n",
       "      <td>-4.044917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>S</td>\n",
       "      <td>2</td>\n",
       "      <td>-1.128540</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>S</td>\n",
       "      <td>3</td>\n",
       "      <td>-1.626888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>S</td>\n",
       "      <td>4</td>\n",
       "      <td>-1.097812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>S</td>\n",
       "      <td>6</td>\n",
       "      <td>-0.626228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1253</th>\n",
       "      <td>S</td>\n",
       "      <td>1268</td>\n",
       "      <td>-1.606183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1254</th>\n",
       "      <td>S</td>\n",
       "      <td>1269</td>\n",
       "      <td>-2.951920</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1255</th>\n",
       "      <td>S</td>\n",
       "      <td>1270</td>\n",
       "      <td>-1.680775</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1256</th>\n",
       "      <td>S</td>\n",
       "      <td>1271</td>\n",
       "      <td>-2.104300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1257</th>\n",
       "      <td>S</td>\n",
       "      <td>1272</td>\n",
       "      <td>-1.962562</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1258 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene  site   fitness\n",
       "0       S     1 -4.044917\n",
       "1       S     2 -1.128540\n",
       "2       S     3 -1.626888\n",
       "3       S     4 -1.097812\n",
       "4       S     6 -0.626228\n",
       "...   ...   ...       ...\n",
       "1253    S  1268 -1.606183\n",
       "1254    S  1269 -2.951920\n",
       "1255    S  1270 -1.680775\n",
       "1256    S  1271 -2.104300\n",
       "1257    S  1272 -1.962562\n",
       "\n",
       "[1258 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "aa_fitness = (\n",
    "    pd.read_csv(aa_fitness_csv)\n",
    "    .query(\"aa != '*'\")\n",
    "    .query(\"expected_count >= @min_expected_count\")\n",
    "    .rename(columns={\"aa_site\": \"site\"})\n",
    "    .query(\"gene == 'S'\")\n",
    "    [[\"gene\", \"site\", \"aa\", \"fitness\"]]\n",
    ")\n",
    "\n",
    "print(\"Fitnesses for amino acids:\")\n",
    "display(aa_fitness)\n",
    "\n",
    "print(\"\\nMean fitnesses for sites:\")\n",
    "site_fitness = (\n",
    "    aa_fitness\n",
    "    .groupby([\"gene\", \"site\"], as_index=False)\n",
    "    .aggregate(fitness=pd.NamedAgg(\"fitness\", \"mean\"))\n",
    ")\n",
    "display(site_fitness)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad96e6ea-f2db-44fd-a834-e100df944e74",
   "metadata": {},
   "source": [
    "Now for each spike DMS dataset and also get the mean DMS measured value at each site:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4ae57bd8-0792-4186-ac47-e0eaf63282ed",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-30T17:54:53.092262Z",
     "iopub.status.busy": "2023-05-30T17:54:53.091986Z",
     "iopub.status.idle": "2023-05-30T17:54:53.212546Z",
     "shell.execute_reply": "2023-05-30T17:54:53.211520Z",
     "shell.execute_reply.started": "2023-05-30T17:54:53.092239Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Deep mutational scanning data:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DMS study</th>\n",
       "      <th>gene</th>\n",
       "      <th>site</th>\n",
       "      <th>aa</th>\n",
       "      <th>DMS effect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>spike (Dadonaite et al, 2022)</td>\n",
       "      <td>S</td>\n",
       "      <td>1</td>\n",
       "      <td>I</td>\n",
       "      <td>-2.2303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>spike (Dadonaite et al, 2022)</td>\n",
       "      <td>S</td>\n",
       "      <td>1</td>\n",
       "      <td>T</td>\n",
       "      <td>-2.4823</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>spike (Dadonaite et al, 2022)</td>\n",
       "      <td>S</td>\n",
       "      <td>2</td>\n",
       "      <td>L</td>\n",
       "      <td>0.2451</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>spike (Dadonaite et al, 2022)</td>\n",
       "      <td>S</td>\n",
       "      <td>2</td>\n",
       "      <td>S</td>\n",
       "      <td>0.3820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>spike (Dadonaite et al, 2022)</td>\n",
       "      <td>S</td>\n",
       "      <td>2</td>\n",
       "      <td>V</td>\n",
       "      <td>0.3639</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3724</th>\n",
       "      <td>RBD (Starr et al, 2022)</td>\n",
       "      <td>S</td>\n",
       "      <td>473</td>\n",
       "      <td>Y</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3743</th>\n",
       "      <td>RBD (Starr et al, 2022)</td>\n",
       "      <td>S</td>\n",
       "      <td>489</td>\n",
       "      <td>Y</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3762</th>\n",
       "      <td>RBD (Starr et al, 2022)</td>\n",
       "      <td>S</td>\n",
       "      <td>495</td>\n",
       "      <td>Y</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3781</th>\n",
       "      <td>RBD (Starr et al, 2022)</td>\n",
       "      <td>S</td>\n",
       "      <td>505</td>\n",
       "      <td>Y</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3800</th>\n",
       "      <td>RBD (Starr et al, 2022)</td>\n",
       "      <td>S</td>\n",
       "      <td>508</td>\n",
       "      <td>Y</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11275 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          DMS study gene  site aa  DMS effect\n",
       "0     spike (Dadonaite et al, 2022)    S     1  I     -2.2303\n",
       "1     spike (Dadonaite et al, 2022)    S     1  T     -2.4823\n",
       "4     spike (Dadonaite et al, 2022)    S     2  L      0.2451\n",
       "5     spike (Dadonaite et al, 2022)    S     2  S      0.3820\n",
       "6     spike (Dadonaite et al, 2022)    S     2  V      0.3639\n",
       "...                             ...  ...   ... ..         ...\n",
       "3724        RBD (Starr et al, 2022)    S   473  Y      0.0000\n",
       "3743        RBD (Starr et al, 2022)    S   489  Y      0.0000\n",
       "3762        RBD (Starr et al, 2022)    S   495  Y      0.0000\n",
       "3781        RBD (Starr et al, 2022)    S   505  Y      0.0000\n",
       "3800        RBD (Starr et al, 2022)    S   508  Y      0.0000\n",
       "\n",
       "[11275 rows x 5 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Deep mutational scanning data mean site values:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DMS study</th>\n",
       "      <th>gene</th>\n",
       "      <th>site</th>\n",
       "      <th>DMS effect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RBD (Starr et al, 2022)</td>\n",
       "      <td>S</td>\n",
       "      <td>331</td>\n",
       "      <td>-0.258277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RBD (Starr et al, 2022)</td>\n",
       "      <td>S</td>\n",
       "      <td>332</td>\n",
       "      <td>-0.011155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RBD (Starr et al, 2022)</td>\n",
       "      <td>S</td>\n",
       "      <td>333</td>\n",
       "      <td>-0.236654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RBD (Starr et al, 2022)</td>\n",
       "      <td>S</td>\n",
       "      <td>334</td>\n",
       "      <td>-0.218237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RBD (Starr et al, 2022)</td>\n",
       "      <td>S</td>\n",
       "      <td>335</td>\n",
       "      <td>-0.059493</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1435</th>\n",
       "      <td>spike (Dadonaite et al, 2022)</td>\n",
       "      <td>S</td>\n",
       "      <td>1248</td>\n",
       "      <td>-0.299075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1436</th>\n",
       "      <td>spike (Dadonaite et al, 2022)</td>\n",
       "      <td>S</td>\n",
       "      <td>1249</td>\n",
       "      <td>-0.014183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1437</th>\n",
       "      <td>spike (Dadonaite et al, 2022)</td>\n",
       "      <td>S</td>\n",
       "      <td>1250</td>\n",
       "      <td>-0.114000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1438</th>\n",
       "      <td>spike (Dadonaite et al, 2022)</td>\n",
       "      <td>S</td>\n",
       "      <td>1251</td>\n",
       "      <td>0.137300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1439</th>\n",
       "      <td>spike (Dadonaite et al, 2022)</td>\n",
       "      <td>S</td>\n",
       "      <td>1252</td>\n",
       "      <td>0.212347</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1440 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          DMS study gene  site  DMS effect\n",
       "0           RBD (Starr et al, 2022)    S   331   -0.258277\n",
       "1           RBD (Starr et al, 2022)    S   332   -0.011155\n",
       "2           RBD (Starr et al, 2022)    S   333   -0.236654\n",
       "3           RBD (Starr et al, 2022)    S   334   -0.218237\n",
       "4           RBD (Starr et al, 2022)    S   335   -0.059493\n",
       "...                             ...  ...   ...         ...\n",
       "1435  spike (Dadonaite et al, 2022)    S  1248   -0.299075\n",
       "1436  spike (Dadonaite et al, 2022)    S  1249   -0.014183\n",
       "1437  spike (Dadonaite et al, 2022)    S  1250   -0.114000\n",
       "1438  spike (Dadonaite et al, 2022)    S  1251    0.137300\n",
       "1439  spike (Dadonaite et al, 2022)    S  1252    0.212347\n",
       "\n",
       "[1440 rows x 4 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dms_dfs = []\n",
    "\n",
    "for dms_dataset, dms_d in dms_datasets.items():\n",
    "    \n",
    "    if dms_d[\"gene\"] != \"S\":\n",
    "        continue\n",
    "   \n",
    "    dms_df = (\n",
    "        dms_d[\"data\"]\n",
    "        .query(\"(wildtype != '*') and (mutant != '*')\")\n",
    "    )\n",
    "    \n",
    "    if dms_df[\"site\"].dtype != int:\n",
    "        dms_df = (\n",
    "            dms_df\n",
    "            .query(\"site.str.isnumeric()\")\n",
    "            .assign(site=lambda x: x[\"site\"].astype(int))\n",
    "        )\n",
    "    \n",
    "    if not len(dms_df.query(\"wildtype == mutant\")):\n",
    "        # add wildtypes with effects of zero\n",
    "        dms_df = pd.concat(\n",
    "            [\n",
    "                dms_df,\n",
    "                (\n",
    "                    dms_df\n",
    "                    [[\"site\", \"wildtype\"]]\n",
    "                    .drop_duplicates()\n",
    "                    .assign(mutant=lambda x: x[\"wildtype\"], effect=0)\n",
    "                ),\n",
    "            ]\n",
    "        )\n",
    "    \n",
    "    if \"filter_cols\" in dms_d:\n",
    "        for c, x in dms_d[\"filter_cols\"].items():\n",
    "            dms_df = dms_df[dms_df[c] >= x]\n",
    "            \n",
    "    # calculate site summary DMS and merge with dN/dS and fitness\n",
    "    dms_dfs.append(\n",
    "        dms_df.assign(gene=dms_d[\"gene\"], study=dms_d[\"description\"])\n",
    "    )\n",
    "    \n",
    "dms_df = pd.concat(dms_dfs)[[\"study\", \"gene\", \"site\", \"mutant\", \"effect\"]].rename(\n",
    "    columns={\"mutant\": \"aa\", \"study\": \"DMS study\", \"effect\": \"DMS effect\"}\n",
    ")\n",
    "\n",
    "print(\"Deep mutational scanning data:\")\n",
    "display(dms_df)\n",
    "\n",
    "dms_site_df = (\n",
    "    dms_df.groupby([\"DMS study\", \"gene\", \"site\"], as_index=False)\n",
    "    .aggregate({\"DMS effect\": \"mean\"})\n",
    ")\n",
    "\n",
    "print(f\"\\nDeep mutational scanning data mean site values:\")\n",
    "dms_site_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d2a8ce23-ea11-4c4f-aa01-24b99034ae1b",
   "metadata": {},
   "source": [
    "Now concatenate everything together into one data frame with comparator study effects, fitness effects, and DMS:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "42ffa9c0-23d7-4ce9-9f45-dc310c58acc2",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-30T18:00:19.499251Z",
     "iopub.status.busy": "2023-05-30T18:00:19.498771Z",
     "iopub.status.idle": "2023-05-30T18:00:19.665148Z",
     "shell.execute_reply": "2023-05-30T18:00:19.664189Z",
     "shell.execute_reply.started": "2023-05-30T18:00:19.499230Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>aa_site</th>\n",
       "      <th>comparator score</th>\n",
       "      <th>comparator_study</th>\n",
       "      <th>comparator_metric</th>\n",
       "      <th>DMS study</th>\n",
       "      <th>DMS effect</th>\n",
       "      <th>fitness</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>S</td>\n",
       "      <td>33</td>\n",
       "      <td>-0.629800</td>\n",
       "      <td>Rodriguez-Rivas et al (2022)</td>\n",
       "      <td>DCA mutability score</td>\n",
       "      <td>spike (Dadonaite et al, 2022)</td>\n",
       "      <td>-0.205417</td>\n",
       "      <td>-0.314275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>S</td>\n",
       "      <td>34</td>\n",
       "      <td>-1.305700</td>\n",
       "      <td>Rodriguez-Rivas et al (2022)</td>\n",
       "      <td>DCA mutability score</td>\n",
       "      <td>spike (Dadonaite et al, 2022)</td>\n",
       "      <td>-1.699922</td>\n",
       "      <td>-1.892543</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>S</td>\n",
       "      <td>35</td>\n",
       "      <td>-1.375700</td>\n",
       "      <td>Rodriguez-Rivas et al (2022)</td>\n",
       "      <td>DCA mutability score</td>\n",
       "      <td>spike (Dadonaite et al, 2022)</td>\n",
       "      <td>-1.350217</td>\n",
       "      <td>-1.495669</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>S</td>\n",
       "      <td>36</td>\n",
       "      <td>-1.138800</td>\n",
       "      <td>Rodriguez-Rivas et al (2022)</td>\n",
       "      <td>DCA mutability score</td>\n",
       "      <td>spike (Dadonaite et al, 2022)</td>\n",
       "      <td>-0.097117</td>\n",
       "      <td>-1.077818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>S</td>\n",
       "      <td>37</td>\n",
       "      <td>-1.422600</td>\n",
       "      <td>Rodriguez-Rivas et al (2022)</td>\n",
       "      <td>DCA mutability score</td>\n",
       "      <td>spike (Dadonaite et al, 2022)</td>\n",
       "      <td>-0.562950</td>\n",
       "      <td>-2.123040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11125</th>\n",
       "      <td>S</td>\n",
       "      <td>Y743</td>\n",
       "      <td>-21.962940</td>\n",
       "      <td>Thadani et al (2023)</td>\n",
       "      <td>EVE score</td>\n",
       "      <td>spike (Dadonaite et al, 2022)</td>\n",
       "      <td>-2.650200</td>\n",
       "      <td>-5.747800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11126</th>\n",
       "      <td>S</td>\n",
       "      <td>Y738</td>\n",
       "      <td>-21.489722</td>\n",
       "      <td>Thadani et al (2023)</td>\n",
       "      <td>EVE score</td>\n",
       "      <td>spike (Dadonaite et al, 2022)</td>\n",
       "      <td>-2.472500</td>\n",
       "      <td>-5.747800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11127</th>\n",
       "      <td>S</td>\n",
       "      <td>Y1032</td>\n",
       "      <td>-21.721045</td>\n",
       "      <td>Thadani et al (2023)</td>\n",
       "      <td>EVE score</td>\n",
       "      <td>spike (Dadonaite et al, 2022)</td>\n",
       "      <td>-2.967200</td>\n",
       "      <td>-4.649200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11128</th>\n",
       "      <td>S</td>\n",
       "      <td>C999</td>\n",
       "      <td>-22.100269</td>\n",
       "      <td>Thadani et al (2023)</td>\n",
       "      <td>EVE score</td>\n",
       "      <td>spike (Dadonaite et al, 2022)</td>\n",
       "      <td>-2.061900</td>\n",
       "      <td>-3.672200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11129</th>\n",
       "      <td>S</td>\n",
       "      <td>Y749</td>\n",
       "      <td>-22.495264</td>\n",
       "      <td>Thadani et al (2023)</td>\n",
       "      <td>EVE score</td>\n",
       "      <td>spike (Dadonaite et al, 2022)</td>\n",
       "      <td>-2.349100</td>\n",
       "      <td>-5.747800</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11130 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene aa_site  comparator score              comparator_study  \\\n",
       "0        S      33         -0.629800  Rodriguez-Rivas et al (2022)   \n",
       "1        S      34         -1.305700  Rodriguez-Rivas et al (2022)   \n",
       "2        S      35         -1.375700  Rodriguez-Rivas et al (2022)   \n",
       "3        S      36         -1.138800  Rodriguez-Rivas et al (2022)   \n",
       "4        S      37         -1.422600  Rodriguez-Rivas et al (2022)   \n",
       "...    ...     ...               ...                           ...   \n",
       "11125    S    Y743        -21.962940          Thadani et al (2023)   \n",
       "11126    S    Y738        -21.489722          Thadani et al (2023)   \n",
       "11127    S   Y1032        -21.721045          Thadani et al (2023)   \n",
       "11128    S    C999        -22.100269          Thadani et al (2023)   \n",
       "11129    S    Y749        -22.495264          Thadani et al (2023)   \n",
       "\n",
       "          comparator_metric                      DMS study  DMS effect  \\\n",
       "0      DCA mutability score  spike (Dadonaite et al, 2022)   -0.205417   \n",
       "1      DCA mutability score  spike (Dadonaite et al, 2022)   -1.699922   \n",
       "2      DCA mutability score  spike (Dadonaite et al, 2022)   -1.350217   \n",
       "3      DCA mutability score  spike (Dadonaite et al, 2022)   -0.097117   \n",
       "4      DCA mutability score  spike (Dadonaite et al, 2022)   -0.562950   \n",
       "...                     ...                            ...         ...   \n",
       "11125             EVE score  spike (Dadonaite et al, 2022)   -2.650200   \n",
       "11126             EVE score  spike (Dadonaite et al, 2022)   -2.472500   \n",
       "11127             EVE score  spike (Dadonaite et al, 2022)   -2.967200   \n",
       "11128             EVE score  spike (Dadonaite et al, 2022)   -2.061900   \n",
       "11129             EVE score  spike (Dadonaite et al, 2022)   -2.349100   \n",
       "\n",
       "        fitness  \n",
       "0     -0.314275  \n",
       "1     -1.892543  \n",
       "2     -1.495669  \n",
       "3     -1.077818  \n",
       "4     -2.123040  \n",
       "...         ...  \n",
       "11125 -5.747800  \n",
       "11126 -5.747800  \n",
       "11127 -4.649200  \n",
       "11128 -3.672200  \n",
       "11129 -5.747800  \n",
       "\n",
       "[11130 rows x 8 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_dfs = []\n",
    "\n",
    "for study, study_d in comparator_studies.items():\n",
    "    comparator_df = (\n",
    "        study_d[\"df\"]\n",
    "        .assign(\n",
    "            comparator_study=study_d[\"name\"],\n",
    "            comparator_metric=study_d[\"df\"].columns[-1],\n",
    "        )\n",
    "        .rename(columns={study_d[\"df\"].columns[-1]: \"comparator score\"})\n",
    "    )\n",
    "    \n",
    "    if \"aa\" in comparator_df.columns:\n",
    "        comparator_df = (\n",
    "            comparator_df\n",
    "            .merge(dms_df, validate=\"one_to_many\")\n",
    "            .merge(aa_fitness, validate=\"many_to_one\")\n",
    "            .assign(aa_site=lambda x: x[\"aa\"] + x[\"site\"].astype(str))\n",
    "            .drop(columns=[\"site\", \"aa\"])\n",
    "        )\n",
    "    else:\n",
    "        comparator_df = (\n",
    "            comparator_df\n",
    "            .merge(dms_site_df, validate=\"one_to_many\")\n",
    "            .merge(site_fitness, validate=\"many_to_one\")\n",
    "            .rename(columns={\"site\": \"aa_site\"})\n",
    "        )\n",
    "        \n",
    "    merged_dfs.append(comparator_df)\n",
    "    \n",
    "merged_df = pd.concat(merged_dfs, ignore_index=True)\n",
    "\n",
    "merged_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a5f7c15-69da-42a9-9c73-b6561e10f274",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
