{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "008b56e0-8e16-4e76-851c-edfdc35ca119",
   "metadata": {},
   "source": [
    "# Analyze fitness effects of mutations fixed in each clade\n",
    "\n",
    "Import Python modules:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6718e87a-8da0-49fa-a00a-6311caca0f83",
   "metadata": {},
   "outputs": [],
   "source": [
    "import altair as alt\n",
    "\n",
    "import Bio.Seq\n",
    "\n",
    "import numpy\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "import yaml\n",
    "\n",
    "_ = alt.data_transformers.disable_max_rows()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60c785d9-1e75-4959-928b-c92580c31c62",
   "metadata": {},
   "source": [
    "Now get variables from `snakemake`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f47affb5-0a48-475e-a522-7de36413e745",
   "metadata": {},
   "outputs": [],
   "source": [
    "if \"snakemake\" not in globals() and \"snakemake\" not in locals():\n",
    "    # variables set manually for interactive debugging\n",
    "    aa_fitness_csv = \"../results/aa_fitness/aa_fitness.csv\"\n",
    "    aamut_by_clade_csv = \"../results/aa_fitness/aamut_fitness_by_clade.csv\"\n",
    "    clade_founder_nts_csv = \"../results/clade_founder_nts/clade_founder_nts.csv\"\n",
    "    \n",
    "    with open(\"../config.yaml\") as f:\n",
    "        config = yaml.safe_load(f)\n",
    "    min_expected_count = config[\"min_expected_count\"]\n",
    "    ref = config[\"clade_fixed_muts_ref\"]\n",
    "    \n",
    "else:\n",
    "    # get variables from `snakemake` when running pipeline\n",
    "    aa_fitness_csv = snakemake.input.aafitness\n",
    "    aamut_by_clade_csv = snakemake.input.aamut_by_clade\n",
    "    clade_founder_nts_csv = snakemake.input.clade_founder_nts_csv\n",
    "    min_expected_count = snakemake.params.min_expected_count\n",
    "    ref = snakemake.params.ref"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a7a6ba8-34d3-4d77-835a-60e7ecf166d5",
   "metadata": {},
   "source": [
    "Read the input data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "80047a0c-872a-4a6d-9953-0c7e88e84fa0",
   "metadata": {},
   "outputs": [],
   "source": [
    "aa_fitness = pd.read_csv(aa_fitness_csv).rename(columns={\"aa_site\": \"site\"})\n",
    "aamut_by_clade = pd.read_csv(aamut_by_clade_csv).rename(columns={\"aa_site\": \"site\"})\n",
    "clade_founder_nts = pd.read_csv(clade_founder_nts_csv)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d6b082ca-4629-48d3-aabb-b89bb54ea30f",
   "metadata": {},
   "source": [
    "We only consider clades that had sufficient counts to estimate the effects of mutations:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c2e2c993-aef9-42f4-9716-6b530f8733fa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analyzing clades=['20A', '20B', '20C', '20E', '20G', '20I', '21C', '21I', '21J', '21K', '21L', '22A', '22B', '22C']\n"
     ]
    }
   ],
   "source": [
    "clades = aamut_by_clade[\"clade\"].unique().tolist()\n",
    "\n",
    "print(f\"Analyzing {clades=}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "296bb419-0dee-4044-bb61-04851754a75f",
   "metadata": {},
   "source": [
    "First we get the amino acids at each position in each gene in each clade founder and the reference clade.\n",
    "We ignore ORF1a and just look at ORF1ab:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "205adb3b-25c8-4713-8eb5-b2f46fd2e53c",
   "metadata": {},
   "outputs": [],
   "source": [
    "codon_to_aa = {\n",
    "    f\"{nt1}{nt2}{nt3}\": str(Bio.Seq.Seq(f\"{nt1}{nt2}{nt3}\").translate())\n",
    "    for nt1 in \"ACGT\" for nt2 in \"ACGT\" for nt3 in \"ACGT\"\n",
    "}\n",
    "\n",
    "assert ref in clade_founder_nts[\"clade\"].unique()\n",
    "\n",
    "clade_founder_aas = (\n",
    "    clade_founder_nts\n",
    "    [[\"clade\", \"gene\", \"codon\", \"codon_site\"]]\n",
    "    .drop_duplicates()\n",
    "    .assign(\n",
    "        gene=lambda x: x[\"gene\"].str.split(\";\"),\n",
    "        codon=lambda x: x[\"codon\"].str.split(\";\"),\n",
    "        site=lambda x: x[\"codon_site\"].str.split(\";\"),\n",
    "    )\n",
    "    .explode([\"gene\", \"codon\", \"site\"])\n",
    "    .query(\"gene != 'ORF1a'\")\n",
    "    .assign(\n",
    "        amino_acid=lambda x: x[\"codon\"].map(codon_to_aa),\n",
    "        site=lambda x: x[\"site\"].astype(int),\n",
    "    )\n",
    "    [[\"clade\", \"gene\", \"site\", \"codon\", \"amino_acid\"]]\n",
    "    .drop_duplicates()\n",
    ")\n",
    "\n",
    "ref_aas = (\n",
    "    clade_founder_aas\n",
    "    .query(\"clade == @ref\")\n",
    "    [[\"gene\", \"site\", \"codon\", \"amino_acid\"]]\n",
    "    .rename(columns={\"codon\": \"ref_codon\", \"amino_acid\": \"ref_amino_acid\"})\n",
    ") \n",
    "\n",
    "clade_founder_aas = clade_founder_aas.query(\"clade in @clades\").merge(\n",
    "    ref_aas,\n",
    "    on=[\"gene\", \"site\"],\n",
    "    validate=\"many_to_one\",\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae08c6fc-1436-4ddd-9643-84f51aaad9b7",
   "metadata": {},
   "source": [
    "Now get just the amino acid mutations in each clade founder, and summarize the number of such mutations:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "5ea40145-5b24-46b9-b74d-fa0d9e3fb078",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_amino_acid_muts</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>clade</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20A</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20B</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20C</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20E</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20G</th>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20I</th>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21C</th>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21I</th>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21J</th>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21K</th>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21L</th>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22A</th>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22B</th>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22C</th>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       n_amino_acid_muts\n",
       "clade                   \n",
       "20A                    2\n",
       "20B                    4\n",
       "20C                    4\n",
       "20E                    5\n",
       "20G                   11\n",
       "20I                   18\n",
       "21C                    9\n",
       "21I                   22\n",
       "21J                   26\n",
       "21K                   42\n",
       "21L                   48\n",
       "22A                   50\n",
       "22B                   48\n",
       "22C                   50"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clade_founder_muts = clade_founder_aas.query(\"amino_acid != ref_amino_acid\")\n",
    "\n",
    "(\n",
    "    clade_founder_muts\n",
    "    .groupby(\"clade\")\n",
    "    .aggregate(n_amino_acid_muts=pd.NamedAgg(\"site\", \"count\"))\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a98233ef-7f48-4d2d-8ba5-9aa97feb7633",
   "metadata": {},
   "source": [
    "Get the estimated fitness effect of each mutation from the overall fitness estimates:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d186ced7-8f6b-437b-8d0c-24ca65e4d320",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>clade</th>\n",
       "      <th>gene</th>\n",
       "      <th>site</th>\n",
       "      <th>codon</th>\n",
       "      <th>amino_acid</th>\n",
       "      <th>ref_codon</th>\n",
       "      <th>ref_amino_acid</th>\n",
       "      <th>amino_acid_fitness</th>\n",
       "      <th>ref_amino_acid_fitness</th>\n",
       "      <th>mutation</th>\n",
       "      <th>fitness_effect</th>\n",
       "      <th>expected_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>21L</td>\n",
       "      <td>ORF1ab</td>\n",
       "      <td>135</td>\n",
       "      <td>AGG</td>\n",
       "      <td>R</td>\n",
       "      <td>AGT</td>\n",
       "      <td>S</td>\n",
       "      <td>0.44707</td>\n",
       "      <td>0.0</td>\n",
       "      <td>S135R</td>\n",
       "      <td>0.44707</td>\n",
       "      <td>96.22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>22A</td>\n",
       "      <td>ORF1ab</td>\n",
       "      <td>135</td>\n",
       "      <td>AGG</td>\n",
       "      <td>R</td>\n",
       "      <td>AGT</td>\n",
       "      <td>S</td>\n",
       "      <td>0.44707</td>\n",
       "      <td>0.0</td>\n",
       "      <td>S135R</td>\n",
       "      <td>0.44707</td>\n",
       "      <td>96.22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>22B</td>\n",
       "      <td>ORF1ab</td>\n",
       "      <td>135</td>\n",
       "      <td>AGG</td>\n",
       "      <td>R</td>\n",
       "      <td>AGT</td>\n",
       "      <td>S</td>\n",
       "      <td>0.44707</td>\n",
       "      <td>0.0</td>\n",
       "      <td>S135R</td>\n",
       "      <td>0.44707</td>\n",
       "      <td>96.22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22C</td>\n",
       "      <td>ORF1ab</td>\n",
       "      <td>135</td>\n",
       "      <td>AGG</td>\n",
       "      <td>R</td>\n",
       "      <td>AGT</td>\n",
       "      <td>S</td>\n",
       "      <td>0.44707</td>\n",
       "      <td>0.0</td>\n",
       "      <td>S135R</td>\n",
       "      <td>0.44707</td>\n",
       "      <td>96.22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>20C</td>\n",
       "      <td>ORF1ab</td>\n",
       "      <td>265</td>\n",
       "      <td>ATC</td>\n",
       "      <td>I</td>\n",
       "      <td>ACC</td>\n",
       "      <td>T</td>\n",
       "      <td>0.81922</td>\n",
       "      <td>0.0</td>\n",
       "      <td>T265I</td>\n",
       "      <td>0.81922</td>\n",
       "      <td>496.45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>334</th>\n",
       "      <td>21L</td>\n",
       "      <td>N</td>\n",
       "      <td>413</td>\n",
       "      <td>CGT</td>\n",
       "      <td>R</td>\n",
       "      <td>AGT</td>\n",
       "      <td>S</td>\n",
       "      <td>0.92395</td>\n",
       "      <td>0.0</td>\n",
       "      <td>S413R</td>\n",
       "      <td>0.92395</td>\n",
       "      <td>167.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>335</th>\n",
       "      <td>22A</td>\n",
       "      <td>N</td>\n",
       "      <td>413</td>\n",
       "      <td>CGT</td>\n",
       "      <td>R</td>\n",
       "      <td>AGT</td>\n",
       "      <td>S</td>\n",
       "      <td>0.92395</td>\n",
       "      <td>0.0</td>\n",
       "      <td>S413R</td>\n",
       "      <td>0.92395</td>\n",
       "      <td>167.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>336</th>\n",
       "      <td>22B</td>\n",
       "      <td>N</td>\n",
       "      <td>413</td>\n",
       "      <td>CGT</td>\n",
       "      <td>R</td>\n",
       "      <td>AGT</td>\n",
       "      <td>S</td>\n",
       "      <td>0.92395</td>\n",
       "      <td>0.0</td>\n",
       "      <td>S413R</td>\n",
       "      <td>0.92395</td>\n",
       "      <td>167.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>337</th>\n",
       "      <td>22C</td>\n",
       "      <td>N</td>\n",
       "      <td>413</td>\n",
       "      <td>CGT</td>\n",
       "      <td>R</td>\n",
       "      <td>AGT</td>\n",
       "      <td>S</td>\n",
       "      <td>0.92395</td>\n",
       "      <td>0.0</td>\n",
       "      <td>S413R</td>\n",
       "      <td>0.92395</td>\n",
       "      <td>167.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>338</th>\n",
       "      <td>20E</td>\n",
       "      <td>ORF10</td>\n",
       "      <td>30</td>\n",
       "      <td>TTA</td>\n",
       "      <td>L</td>\n",
       "      <td>GTA</td>\n",
       "      <td>V</td>\n",
       "      <td>0.86409</td>\n",
       "      <td>0.0</td>\n",
       "      <td>V30L</td>\n",
       "      <td>0.86409</td>\n",
       "      <td>330.45</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>339 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    clade    gene  site codon amino_acid ref_codon ref_amino_acid  \\\n",
       "0     21L  ORF1ab   135   AGG          R       AGT              S   \n",
       "1     22A  ORF1ab   135   AGG          R       AGT              S   \n",
       "2     22B  ORF1ab   135   AGG          R       AGT              S   \n",
       "3     22C  ORF1ab   135   AGG          R       AGT              S   \n",
       "4     20C  ORF1ab   265   ATC          I       ACC              T   \n",
       "..    ...     ...   ...   ...        ...       ...            ...   \n",
       "334   21L       N   413   CGT          R       AGT              S   \n",
       "335   22A       N   413   CGT          R       AGT              S   \n",
       "336   22B       N   413   CGT          R       AGT              S   \n",
       "337   22C       N   413   CGT          R       AGT              S   \n",
       "338   20E   ORF10    30   TTA          L       GTA              V   \n",
       "\n",
       "     amino_acid_fitness  ref_amino_acid_fitness mutation  fitness_effect  \\\n",
       "0               0.44707                     0.0    S135R         0.44707   \n",
       "1               0.44707                     0.0    S135R         0.44707   \n",
       "2               0.44707                     0.0    S135R         0.44707   \n",
       "3               0.44707                     0.0    S135R         0.44707   \n",
       "4               0.81922                     0.0    T265I         0.81922   \n",
       "..                  ...                     ...      ...             ...   \n",
       "334             0.92395                     0.0    S413R         0.92395   \n",
       "335             0.92395                     0.0    S413R         0.92395   \n",
       "336             0.92395                     0.0    S413R         0.92395   \n",
       "337             0.92395                     0.0    S413R         0.92395   \n",
       "338             0.86409                     0.0     V30L         0.86409   \n",
       "\n",
       "     expected_count  \n",
       "0             96.22  \n",
       "1             96.22  \n",
       "2             96.22  \n",
       "3             96.22  \n",
       "4            496.45  \n",
       "..              ...  \n",
       "334          167.67  \n",
       "335          167.67  \n",
       "336          167.67  \n",
       "337          167.67  \n",
       "338          330.45  \n",
       "\n",
       "[339 rows x 12 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# overall clade estimates of fitness effects of mutations\n",
    "fitness_effects_tidy = (\n",
    "    clade_founder_muts\n",
    "    .merge(\n",
    "        aa_fitness[[\"gene\", \"site\", \"aa\", \"fitness\", \"expected_count\"]].rename(\n",
    "            columns={\"aa\": \"amino_acid\", \"fitness\": \"amino_acid_fitness\"},\n",
    "        ),\n",
    "        on=[\"site\", \"gene\", \"amino_acid\"],\n",
    "        how=\"left\",\n",
    "        validate=\"many_to_one\",\n",
    "    )\n",
    "    .merge(\n",
    "        aa_fitness[[\"gene\", \"site\", \"aa\", \"fitness\", \"expected_count\"]].rename(\n",
    "            columns={\"aa\": \"ref_amino_acid\", \"fitness\": \"ref_amino_acid_fitness\"},\n",
    "        ),\n",
    "        on=[\"site\", \"gene\", \"ref_amino_acid\"],\n",
    "        how=\"left\",\n",
    "        validate=\"many_to_one\",\n",
    "    )\n",
    "    .assign(\n",
    "        mutation=lambda x: x[\"ref_amino_acid\"] + x[\"site\"].astype(str) + x[\"amino_acid\"],\n",
    "        fitness_effect=lambda x: x[\"amino_acid_fitness\"] - x[\"ref_amino_acid_fitness\"],\n",
    "        expected_count=lambda x: numpy.minimum(x[\"expected_count_x\"], x[\"expected_count_y\"]),\n",
    "    )\n",
    "    .drop(columns=[\"expected_count_x\", \"expected_count_y\"])\n",
    ")\n",
    "\n",
    "fitness_effects_tidy"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b196b10-9786-4e5d-8476-0c861df50654",
   "metadata": {},
   "source": [
    "Now get this information in a wide-form data frame that just lists the clades with the mutation, and also those clades alongside their clade founder codons:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "75821e08-0192-4ad1-941b-788bb33d8aa7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>site</th>\n",
       "      <th>mutation</th>\n",
       "      <th>ref_codon</th>\n",
       "      <th>ref_amino_acid</th>\n",
       "      <th>fitness_effect</th>\n",
       "      <th>expected_count</th>\n",
       "      <th>n_clades_with_mutation</th>\n",
       "      <th>clades_with_mutation</th>\n",
       "      <th>clade_codons_with_mutation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>E</td>\n",
       "      <td>9</td>\n",
       "      <td>T9I</td>\n",
       "      <td>ACA</td>\n",
       "      <td>T</td>\n",
       "      <td>1.03620</td>\n",
       "      <td>369.730</td>\n",
       "      <td>5</td>\n",
       "      <td>21K; 21L; 22A; 22B; 22C</td>\n",
       "      <td>21K (ATA); 21L (ATA); 22A (ATA); 22B (ATA); 22...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>M</td>\n",
       "      <td>3</td>\n",
       "      <td>D3G</td>\n",
       "      <td>GAT</td>\n",
       "      <td>D</td>\n",
       "      <td>1.48940</td>\n",
       "      <td>134.820</td>\n",
       "      <td>1</td>\n",
       "      <td>21K</td>\n",
       "      <td>21K (GGT)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>M</td>\n",
       "      <td>3</td>\n",
       "      <td>D3N</td>\n",
       "      <td>GAT</td>\n",
       "      <td>D</td>\n",
       "      <td>0.33533</td>\n",
       "      <td>102.120</td>\n",
       "      <td>1</td>\n",
       "      <td>22B</td>\n",
       "      <td>22B (AAT)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>M</td>\n",
       "      <td>19</td>\n",
       "      <td>Q19E</td>\n",
       "      <td>CAA</td>\n",
       "      <td>Q</td>\n",
       "      <td>-1.00460</td>\n",
       "      <td>153.870</td>\n",
       "      <td>5</td>\n",
       "      <td>21K; 21L; 22A; 22B; 22C</td>\n",
       "      <td>21K (GAA); 21L (GAA); 22A (GAA); 22B (GAA); 22...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>M</td>\n",
       "      <td>63</td>\n",
       "      <td>A63T</td>\n",
       "      <td>GCT</td>\n",
       "      <td>A</td>\n",
       "      <td>-1.09570</td>\n",
       "      <td>268.520</td>\n",
       "      <td>5</td>\n",
       "      <td>21K; 21L; 22A; 22B; 22C</td>\n",
       "      <td>21K (ACT); 21L (ACT); 22A (ACT); 22B (ACT); 22...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>S</td>\n",
       "      <td>954</td>\n",
       "      <td>Q954H</td>\n",
       "      <td>CAA</td>\n",
       "      <td>Q</td>\n",
       "      <td>0.29488</td>\n",
       "      <td>292.690</td>\n",
       "      <td>5</td>\n",
       "      <td>21K; 21L; 22A; 22B; 22C</td>\n",
       "      <td>21K (CAT); 21L (CAT); 22A (CAT); 22B (CAT); 22...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>S</td>\n",
       "      <td>969</td>\n",
       "      <td>N969K</td>\n",
       "      <td>AAT</td>\n",
       "      <td>N</td>\n",
       "      <td>0.37826</td>\n",
       "      <td>88.056</td>\n",
       "      <td>5</td>\n",
       "      <td>21K; 21L; 22A; 22B; 22C</td>\n",
       "      <td>21K (AAA); 21L (AAA); 22A (AAA); 22B (AAA); 22...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>S</td>\n",
       "      <td>981</td>\n",
       "      <td>L981F</td>\n",
       "      <td>CTT</td>\n",
       "      <td>L</td>\n",
       "      <td>-1.93490</td>\n",
       "      <td>455.390</td>\n",
       "      <td>1</td>\n",
       "      <td>21K</td>\n",
       "      <td>21K (TTT)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>S</td>\n",
       "      <td>982</td>\n",
       "      <td>S982A</td>\n",
       "      <td>TCA</td>\n",
       "      <td>S</td>\n",
       "      <td>1.72860</td>\n",
       "      <td>67.676</td>\n",
       "      <td>1</td>\n",
       "      <td>20I</td>\n",
       "      <td>20I (GCA)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>S</td>\n",
       "      <td>1118</td>\n",
       "      <td>D1118H</td>\n",
       "      <td>GAC</td>\n",
       "      <td>D</td>\n",
       "      <td>0.33954</td>\n",
       "      <td>74.501</td>\n",
       "      <td>1</td>\n",
       "      <td>20I</td>\n",
       "      <td>20I (CAC)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>119 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene  site mutation ref_codon ref_amino_acid  fitness_effect  \\\n",
       "0      E     9      T9I       ACA              T         1.03620   \n",
       "1      M     3      D3G       GAT              D         1.48940   \n",
       "2      M     3      D3N       GAT              D         0.33533   \n",
       "3      M    19     Q19E       CAA              Q        -1.00460   \n",
       "4      M    63     A63T       GCT              A        -1.09570   \n",
       "..   ...   ...      ...       ...            ...             ...   \n",
       "114    S   954    Q954H       CAA              Q         0.29488   \n",
       "115    S   969    N969K       AAT              N         0.37826   \n",
       "116    S   981    L981F       CTT              L        -1.93490   \n",
       "117    S   982    S982A       TCA              S         1.72860   \n",
       "118    S  1118   D1118H       GAC              D         0.33954   \n",
       "\n",
       "     expected_count  n_clades_with_mutation     clades_with_mutation  \\\n",
       "0           369.730                       5  21K; 21L; 22A; 22B; 22C   \n",
       "1           134.820                       1                      21K   \n",
       "2           102.120                       1                      22B   \n",
       "3           153.870                       5  21K; 21L; 22A; 22B; 22C   \n",
       "4           268.520                       5  21K; 21L; 22A; 22B; 22C   \n",
       "..              ...                     ...                      ...   \n",
       "114         292.690                       5  21K; 21L; 22A; 22B; 22C   \n",
       "115          88.056                       5  21K; 21L; 22A; 22B; 22C   \n",
       "116         455.390                       1                      21K   \n",
       "117          67.676                       1                      20I   \n",
       "118          74.501                       1                      20I   \n",
       "\n",
       "                            clade_codons_with_mutation  \n",
       "0    21K (ATA); 21L (ATA); 22A (ATA); 22B (ATA); 22...  \n",
       "1                                            21K (GGT)  \n",
       "2                                            22B (AAT)  \n",
       "3    21K (GAA); 21L (GAA); 22A (GAA); 22B (GAA); 22...  \n",
       "4    21K (ACT); 21L (ACT); 22A (ACT); 22B (ACT); 22...  \n",
       "..                                                 ...  \n",
       "114  21K (CAT); 21L (CAT); 22A (CAT); 22B (CAT); 22...  \n",
       "115  21K (AAA); 21L (AAA); 22A (AAA); 22B (AAA); 22...  \n",
       "116                                          21K (TTT)  \n",
       "117                                          20I (GCA)  \n",
       "118                                          20I (CAC)  \n",
       "\n",
       "[119 rows x 10 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fitness_effects = (\n",
    "    fitness_effects_tidy\n",
    "    .assign(clade_codon=lambda x: x[\"clade\"] + \" (\" + x[\"codon\"] + \")\")\n",
    "    .groupby(\n",
    "        [\"gene\", \"site\", \"mutation\", \"ref_codon\", \"ref_amino_acid\", \"fitness_effect\", \"expected_count\"],\n",
    "        as_index=False,\n",
    "        dropna=False,\n",
    "    )\n",
    "    .aggregate(\n",
    "        n_clades_with_mutation=pd.NamedAgg(\"clade\", \"count\"),\n",
    "        clades_with_mutation=pd.NamedAgg(\"clade\", lambda s: \"; \".join(s)),\n",
    "        clade_codons_with_mutation=pd.NamedAgg(\"clade_codon\", lambda s: \"; \".join(s)),\n",
    "    )\n",
    ")\n",
    "\n",
    "assert len(fitness_effects) == len(fitness_effects.groupby([\"gene\", \"mutation\"]))\n",
    "\n",
    "fitness_effects"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "418c8648-27cf-43eb-9bc6-5c8b4fbf8b96",
   "metadata": {},
   "source": [
    "Get amino-acid mutation fitness estimates by clade, polarized so they are coming from the reference amino-acid identity at the site, and keeping track whether each estimate is for a forward or reversion mutation in that clade.\n",
    "\n",
    "First in tidy form:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "4dd7d753-522e-4e52-93e1-0cc851738efe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>clade</th>\n",
       "      <th>gene</th>\n",
       "      <th>site</th>\n",
       "      <th>mutation</th>\n",
       "      <th>fitness_effect</th>\n",
       "      <th>expected_count</th>\n",
       "      <th>mutation_polarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>20A</td>\n",
       "      <td>E</td>\n",
       "      <td>22</td>\n",
       "      <td>A22D</td>\n",
       "      <td>-0.77399</td>\n",
       "      <td>0.58420</td>\n",
       "      <td>forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>20A</td>\n",
       "      <td>E</td>\n",
       "      <td>22</td>\n",
       "      <td>A22G</td>\n",
       "      <td>-0.36931</td>\n",
       "      <td>0.22337</td>\n",
       "      <td>forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>20A</td>\n",
       "      <td>E</td>\n",
       "      <td>22</td>\n",
       "      <td>A22P</td>\n",
       "      <td>-0.68952</td>\n",
       "      <td>0.49638</td>\n",
       "      <td>forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>20A</td>\n",
       "      <td>E</td>\n",
       "      <td>22</td>\n",
       "      <td>A22S</td>\n",
       "      <td>-3.01370</td>\n",
       "      <td>9.68100</td>\n",
       "      <td>forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>20A</td>\n",
       "      <td>E</td>\n",
       "      <td>22</td>\n",
       "      <td>A22T</td>\n",
       "      <td>-0.85863</td>\n",
       "      <td>3.03990</td>\n",
       "      <td>forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>840553</th>\n",
       "      <td>22C</td>\n",
       "      <td>S</td>\n",
       "      <td>917</td>\n",
       "      <td>Y917D</td>\n",
       "      <td>-0.33554</td>\n",
       "      <td>0.19935</td>\n",
       "      <td>forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>840554</th>\n",
       "      <td>22C</td>\n",
       "      <td>S</td>\n",
       "      <td>917</td>\n",
       "      <td>Y917F</td>\n",
       "      <td>-0.61792</td>\n",
       "      <td>0.42753</td>\n",
       "      <td>forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>840555</th>\n",
       "      <td>22C</td>\n",
       "      <td>S</td>\n",
       "      <td>917</td>\n",
       "      <td>Y917H</td>\n",
       "      <td>-1.45610</td>\n",
       "      <td>1.64450</td>\n",
       "      <td>forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>840556</th>\n",
       "      <td>22C</td>\n",
       "      <td>S</td>\n",
       "      <td>917</td>\n",
       "      <td>Y917N</td>\n",
       "      <td>-0.51064</td>\n",
       "      <td>0.33318</td>\n",
       "      <td>forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>840557</th>\n",
       "      <td>22C</td>\n",
       "      <td>S</td>\n",
       "      <td>917</td>\n",
       "      <td>Y917S</td>\n",
       "      <td>-0.27209</td>\n",
       "      <td>0.15635</td>\n",
       "      <td>forward</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>839018 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       clade gene  site mutation  fitness_effect  expected_count  \\\n",
       "0        20A    E    22     A22D        -0.77399         0.58420   \n",
       "1        20A    E    22     A22G        -0.36931         0.22337   \n",
       "2        20A    E    22     A22P        -0.68952         0.49638   \n",
       "3        20A    E    22     A22S        -3.01370         9.68100   \n",
       "4        20A    E    22     A22T        -0.85863         3.03990   \n",
       "...      ...  ...   ...      ...             ...             ...   \n",
       "840553   22C    S   917    Y917D        -0.33554         0.19935   \n",
       "840554   22C    S   917    Y917F        -0.61792         0.42753   \n",
       "840555   22C    S   917    Y917H        -1.45610         1.64450   \n",
       "840556   22C    S   917    Y917N        -0.51064         0.33318   \n",
       "840557   22C    S   917    Y917S        -0.27209         0.15635   \n",
       "\n",
       "       mutation_polarity  \n",
       "0                forward  \n",
       "1                forward  \n",
       "2                forward  \n",
       "3                forward  \n",
       "4                forward  \n",
       "...                  ...  \n",
       "840553           forward  \n",
       "840554           forward  \n",
       "840555           forward  \n",
       "840556           forward  \n",
       "840557           forward  \n",
       "\n",
       "[839018 rows x 7 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fitness_effects_by_clade_tidy = (\n",
    "    aamut_by_clade\n",
    "    .query(\"mutant_aa != clade_founder_aa\")\n",
    "    .merge(ref_aas, on=[\"gene\", \"site\"], validate=\"many_to_one\")\n",
    "    .query(\"(clade_founder_aa == ref_amino_acid) or (mutant_aa == ref_amino_acid)\")\n",
    "    .assign(\n",
    "        mutation_polarity=lambda x: (x[\"clade_founder_aa\"] == x[\"ref_amino_acid\"]).map(\n",
    "            {True: \"forward\", False: \"reverse\"}\n",
    "        ),\n",
    "        mutation=lambda x: numpy.where(\n",
    "            x[\"mutation_polarity\"] == \"forward\",\n",
    "            x[\"clade_founder_aa\"] + x[\"site\"].astype(str) + x[\"mutant_aa\"],\n",
    "            x[\"mutant_aa\"] + x[\"site\"].astype(str) + x[\"clade_founder_aa\"],\n",
    "        ),\n",
    "        fitness_effect=lambda x: numpy.where(\n",
    "            x[\"mutation_polarity\"] == \"forward\",\n",
    "            x[\"delta_fitness\"],\n",
    "            -x[\"delta_fitness\"],\n",
    "        ),\n",
    "    )\n",
    "    [[\n",
    "        \"clade\",\n",
    "        \"gene\",\n",
    "        \"site\",\n",
    "        \"mutation\",\n",
    "        \"fitness_effect\",\n",
    "        \"expected_count\",\n",
    "        \"mutation_polarity\",\n",
    "    ]]\n",
    ")\n",
    "\n",
    "fitness_effects_by_clade_tidy"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "060043dd-85e0-4c7e-9b2c-6b2090922b30",
   "metadata": {},
   "source": [
    "Now make a wide version with just the mutations that occurred in the clade founders, grouping together the forward and reverse clade estimates.\n",
    "We calculate the **weighted average** (weighted by clade expected count) for each mutation polarity:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "9b5872ca-37a6-42c6-bdc5-20afaea35bdf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>mutation</th>\n",
       "      <th>mutation_polarity</th>\n",
       "      <th>expected_count</th>\n",
       "      <th>clade_effect</th>\n",
       "      <th>fitness_effect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>E</td>\n",
       "      <td>T9I</td>\n",
       "      <td>forward</td>\n",
       "      <td>296.41490</td>\n",
       "      <td>20A (1.00); 20B (1.32); 20C (1.27); 20E (0.94)...</td>\n",
       "      <td>1.019092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>M</td>\n",
       "      <td>A63T</td>\n",
       "      <td>forward</td>\n",
       "      <td>79.21500</td>\n",
       "      <td>20A (0.44); 20B (-0.10); 20C (-0.39); 20E (-1....</td>\n",
       "      <td>-1.378999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>M</td>\n",
       "      <td>D3G</td>\n",
       "      <td>forward</td>\n",
       "      <td>49.21322</td>\n",
       "      <td>20A (2.01); 20B (2.09); 20C (1.01); 20E (0.07)...</td>\n",
       "      <td>1.325150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>M</td>\n",
       "      <td>D3N</td>\n",
       "      <td>forward</td>\n",
       "      <td>102.11730</td>\n",
       "      <td>20A (-0.35); 20B (-0.10); 20C (-1.49); 20E (0....</td>\n",
       "      <td>-0.104354</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>M</td>\n",
       "      <td>I82T</td>\n",
       "      <td>forward</td>\n",
       "      <td>34.49289</td>\n",
       "      <td>20A (2.44); 20B (1.74); 20C (1.11); 20E (-0.90...</td>\n",
       "      <td>1.479248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>S</td>\n",
       "      <td>T716I</td>\n",
       "      <td>forward</td>\n",
       "      <td>468.98270</td>\n",
       "      <td>20A (0.87); 20B (1.50); 20C (1.05); 20E (0.22)...</td>\n",
       "      <td>0.205669</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>S</td>\n",
       "      <td>T95I</td>\n",
       "      <td>forward</td>\n",
       "      <td>234.78070</td>\n",
       "      <td>20A (1.95); 20B (1.82); 20C (2.08); 20E (1.80)...</td>\n",
       "      <td>1.228520</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>S</td>\n",
       "      <td>V213G</td>\n",
       "      <td>forward</td>\n",
       "      <td>4.73354</td>\n",
       "      <td>20A (-0.31); 20B (0.82); 20C (-0.19); 20E (-0....</td>\n",
       "      <td>-0.114797</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>S</td>\n",
       "      <td>W152C</td>\n",
       "      <td>forward</td>\n",
       "      <td>330.71780</td>\n",
       "      <td>20A (-0.66); 20B (-0.70); 20C (0.38); 20E (-1....</td>\n",
       "      <td>-2.054985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>S</td>\n",
       "      <td>Y505H</td>\n",
       "      <td>forward</td>\n",
       "      <td>32.84014</td>\n",
       "      <td>20A (-1.23); 20B (0.00); 20C (0.27); 20E (0.20...</td>\n",
       "      <td>-0.672241</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>127 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene mutation mutation_polarity  expected_count  \\\n",
       "0      E      T9I           forward       296.41490   \n",
       "1      M     A63T           forward        79.21500   \n",
       "2      M      D3G           forward        49.21322   \n",
       "3      M      D3N           forward       102.11730   \n",
       "4      M     I82T           forward        34.49289   \n",
       "..   ...      ...               ...             ...   \n",
       "122    S    T716I           forward       468.98270   \n",
       "123    S     T95I           forward       234.78070   \n",
       "124    S    V213G           forward         4.73354   \n",
       "125    S    W152C           forward       330.71780   \n",
       "126    S    Y505H           forward        32.84014   \n",
       "\n",
       "                                          clade_effect  fitness_effect  \n",
       "0    20A (1.00); 20B (1.32); 20C (1.27); 20E (0.94)...        1.019092  \n",
       "1    20A (0.44); 20B (-0.10); 20C (-0.39); 20E (-1....       -1.378999  \n",
       "2    20A (2.01); 20B (2.09); 20C (1.01); 20E (0.07)...        1.325150  \n",
       "3    20A (-0.35); 20B (-0.10); 20C (-1.49); 20E (0....       -0.104354  \n",
       "4    20A (2.44); 20B (1.74); 20C (1.11); 20E (-0.90...        1.479248  \n",
       "..                                                 ...             ...  \n",
       "122  20A (0.87); 20B (1.50); 20C (1.05); 20E (0.22)...        0.205669  \n",
       "123  20A (1.95); 20B (1.82); 20C (2.08); 20E (1.80)...        1.228520  \n",
       "124  20A (-0.31); 20B (0.82); 20C (-0.19); 20E (-0....       -0.114797  \n",
       "125  20A (-0.66); 20B (-0.70); 20C (0.38); 20E (-1....       -2.054985  \n",
       "126  20A (-1.23); 20B (0.00); 20C (0.27); 20E (0.20...       -0.672241  \n",
       "\n",
       "[127 rows x 6 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fitness_effects_by_clade = (\n",
    "    fitness_effects_by_clade_tidy\n",
    "    # use this merge as a way to retain just mutations that occurred\n",
    "    .merge(fitness_effects[[\"gene\", \"site\", \"mutation\"]])\n",
    "    # group forward and reverse polarities, calculate weighted average fitness effects\n",
    "    .assign(\n",
    "        clade_effect=(\n",
    "            lambda x: (\n",
    "                x[\"clade\"]\n",
    "                + \" (\" + x[\"fitness_effect\"].map(lambda f: \"{:.2f}\".format(f)) + \")\"\n",
    "            )\n",
    "        ),\n",
    "        weighted_effect=lambda x: x[\"fitness_effect\"] * x[\"expected_count\"],\n",
    "    )\n",
    "    .groupby([\"gene\", \"mutation\", \"mutation_polarity\"], as_index=False)\n",
    "    .aggregate(\n",
    "        expected_count=pd.NamedAgg(\"expected_count\", \"sum\"),\n",
    "        weighted_effect=pd.NamedAgg(\"weighted_effect\", \"sum\"),\n",
    "        clade_effect=pd.NamedAgg(\"clade_effect\", lambda s: \"; \".join(s)),\n",
    "    )\n",
    "    .assign(fitness_effect=lambda x: x[\"weighted_effect\"] / x[\"expected_count\"])\n",
    "    .drop(columns=\"weighted_effect\")\n",
    ")\n",
    "\n",
    "fitness_effects_by_clade"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5807164-4ada-4e0b-8ea7-abd9dc2dad4c",
   "metadata": {},
   "source": [
    "### Old stuff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "4d0908f3-8413-4926-b7b8-3e0200e9f1c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>site</th>\n",
       "      <th>mutation</th>\n",
       "      <th>fitness_effect_20A</th>\n",
       "      <th>fitness_effect_20B</th>\n",
       "      <th>fitness_effect_20C</th>\n",
       "      <th>fitness_effect_20E</th>\n",
       "      <th>fitness_effect_20G</th>\n",
       "      <th>fitness_effect_20I</th>\n",
       "      <th>fitness_effect_21C</th>\n",
       "      <th>...</th>\n",
       "      <th>mutation_polarity_20G</th>\n",
       "      <th>mutation_polarity_20I</th>\n",
       "      <th>mutation_polarity_21C</th>\n",
       "      <th>mutation_polarity_21I</th>\n",
       "      <th>mutation_polarity_21J</th>\n",
       "      <th>mutation_polarity_21K</th>\n",
       "      <th>mutation_polarity_21L</th>\n",
       "      <th>mutation_polarity_22A</th>\n",
       "      <th>mutation_polarity_22B</th>\n",
       "      <th>mutation_polarity_22C</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>E</td>\n",
       "      <td>1</td>\n",
       "      <td>M1I</td>\n",
       "      <td>-3.3118</td>\n",
       "      <td>-3.1171</td>\n",
       "      <td>-2.7134</td>\n",
       "      <td>-2.6489</td>\n",
       "      <td>-3.0222</td>\n",
       "      <td>-4.4216</td>\n",
       "      <td>-2.2467</td>\n",
       "      <td>...</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>E</td>\n",
       "      <td>1</td>\n",
       "      <td>M1K</td>\n",
       "      <td>-0.42215</td>\n",
       "      <td>-0.34909</td>\n",
       "      <td>-0.27311</td>\n",
       "      <td>-0.28987</td>\n",
       "      <td>-0.37696</td>\n",
       "      <td>-1.083</td>\n",
       "      <td>-0.19017</td>\n",
       "      <td>...</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>E</td>\n",
       "      <td>1</td>\n",
       "      <td>M1L</td>\n",
       "      <td>-0.83326</td>\n",
       "      <td>-0.6457</td>\n",
       "      <td>-0.48813</td>\n",
       "      <td>-0.53621</td>\n",
       "      <td>-0.70729</td>\n",
       "      <td>-1.6055</td>\n",
       "      <td>-0.38299</td>\n",
       "      <td>...</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>E</td>\n",
       "      <td>1</td>\n",
       "      <td>M1R</td>\n",
       "      <td>-0.31181</td>\n",
       "      <td>-0.28223</td>\n",
       "      <td>-0.19009</td>\n",
       "      <td>-0.19696</td>\n",
       "      <td>-0.27321</td>\n",
       "      <td>-0.86945</td>\n",
       "      <td>-0.14951</td>\n",
       "      <td>...</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>E</td>\n",
       "      <td>1</td>\n",
       "      <td>M1T</td>\n",
       "      <td>-1.226</td>\n",
       "      <td>-1.0952</td>\n",
       "      <td>-0.83236</td>\n",
       "      <td>-0.90089</td>\n",
       "      <td>-1.0618</td>\n",
       "      <td>-2.2221</td>\n",
       "      <td>-0.63201</td>\n",
       "      <td>...</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60108</th>\n",
       "      <td>S</td>\n",
       "      <td>1272</td>\n",
       "      <td>Y1272D</td>\n",
       "      <td>-0.31181</td>\n",
       "      <td>-0.28223</td>\n",
       "      <td>-0.19009</td>\n",
       "      <td>-0.19696</td>\n",
       "      <td>-0.27321</td>\n",
       "      <td>-0.86945</td>\n",
       "      <td>-0.14951</td>\n",
       "      <td>...</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60109</th>\n",
       "      <td>S</td>\n",
       "      <td>1272</td>\n",
       "      <td>Y1272F</td>\n",
       "      <td>-0.60746</td>\n",
       "      <td>-0.50397</td>\n",
       "      <td>-0.35716</td>\n",
       "      <td>-0.39648</td>\n",
       "      <td>-0.55973</td>\n",
       "      <td>-1.3429</td>\n",
       "      <td>-0.2889</td>\n",
       "      <td>...</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60110</th>\n",
       "      <td>S</td>\n",
       "      <td>1272</td>\n",
       "      <td>Y1272H</td>\n",
       "      <td>-0.1274</td>\n",
       "      <td>-1.0952</td>\n",
       "      <td>-0.83236</td>\n",
       "      <td>0.19772</td>\n",
       "      <td>-1.0618</td>\n",
       "      <td>0.34288</td>\n",
       "      <td>-0.63201</td>\n",
       "      <td>...</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60111</th>\n",
       "      <td>S</td>\n",
       "      <td>1272</td>\n",
       "      <td>Y1272N</td>\n",
       "      <td>-0.42215</td>\n",
       "      <td>-0.34909</td>\n",
       "      <td>-0.27311</td>\n",
       "      <td>-0.28987</td>\n",
       "      <td>-0.37696</td>\n",
       "      <td>-1.083</td>\n",
       "      <td>-0.19017</td>\n",
       "      <td>...</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60112</th>\n",
       "      <td>S</td>\n",
       "      <td>1272</td>\n",
       "      <td>Y1272S</td>\n",
       "      <td>-0.38188</td>\n",
       "      <td>-0.22477</td>\n",
       "      <td>-0.18232</td>\n",
       "      <td>-0.20126</td>\n",
       "      <td>-0.24552</td>\n",
       "      <td>-0.7656</td>\n",
       "      <td>-0.12373</td>\n",
       "      <td>...</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>60113 rows × 45 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene  site mutation fitness_effect_20A fitness_effect_20B  \\\n",
       "0        E     1      M1I            -3.3118            -3.1171   \n",
       "1        E     1      M1K           -0.42215           -0.34909   \n",
       "2        E     1      M1L           -0.83326            -0.6457   \n",
       "3        E     1      M1R           -0.31181           -0.28223   \n",
       "4        E     1      M1T             -1.226            -1.0952   \n",
       "...    ...   ...      ...                ...                ...   \n",
       "60108    S  1272   Y1272D           -0.31181           -0.28223   \n",
       "60109    S  1272   Y1272F           -0.60746           -0.50397   \n",
       "60110    S  1272   Y1272H            -0.1274            -1.0952   \n",
       "60111    S  1272   Y1272N           -0.42215           -0.34909   \n",
       "60112    S  1272   Y1272S           -0.38188           -0.22477   \n",
       "\n",
       "      fitness_effect_20C fitness_effect_20E fitness_effect_20G  \\\n",
       "0                -2.7134            -2.6489            -3.0222   \n",
       "1               -0.27311           -0.28987           -0.37696   \n",
       "2               -0.48813           -0.53621           -0.70729   \n",
       "3               -0.19009           -0.19696           -0.27321   \n",
       "4               -0.83236           -0.90089            -1.0618   \n",
       "...                  ...                ...                ...   \n",
       "60108           -0.19009           -0.19696           -0.27321   \n",
       "60109           -0.35716           -0.39648           -0.55973   \n",
       "60110           -0.83236            0.19772            -1.0618   \n",
       "60111           -0.27311           -0.28987           -0.37696   \n",
       "60112           -0.18232           -0.20126           -0.24552   \n",
       "\n",
       "      fitness_effect_20I fitness_effect_21C  ... mutation_polarity_20G  \\\n",
       "0                -4.4216            -2.2467  ...               forward   \n",
       "1                 -1.083           -0.19017  ...               forward   \n",
       "2                -1.6055           -0.38299  ...               forward   \n",
       "3               -0.86945           -0.14951  ...               forward   \n",
       "4                -2.2221           -0.63201  ...               forward   \n",
       "...                  ...                ...  ...                   ...   \n",
       "60108           -0.86945           -0.14951  ...               forward   \n",
       "60109            -1.3429            -0.2889  ...               forward   \n",
       "60110            0.34288           -0.63201  ...               forward   \n",
       "60111             -1.083           -0.19017  ...               forward   \n",
       "60112            -0.7656           -0.12373  ...               forward   \n",
       "\n",
       "      mutation_polarity_20I mutation_polarity_21C mutation_polarity_21I  \\\n",
       "0                   forward               forward               forward   \n",
       "1                   forward               forward               forward   \n",
       "2                   forward               forward               forward   \n",
       "3                   forward               forward               forward   \n",
       "4                   forward               forward               forward   \n",
       "...                     ...                   ...                   ...   \n",
       "60108               forward               forward               forward   \n",
       "60109               forward               forward               forward   \n",
       "60110               forward               forward               forward   \n",
       "60111               forward               forward               forward   \n",
       "60112               forward               forward               forward   \n",
       "\n",
       "      mutation_polarity_21J mutation_polarity_21K mutation_polarity_21L  \\\n",
       "0                   forward               forward               forward   \n",
       "1                   forward               forward               forward   \n",
       "2                   forward               forward               forward   \n",
       "3                   forward               forward               forward   \n",
       "4                   forward               forward               forward   \n",
       "...                     ...                   ...                   ...   \n",
       "60108               forward               forward               forward   \n",
       "60109               forward               forward               forward   \n",
       "60110               forward               forward               forward   \n",
       "60111               forward               forward               forward   \n",
       "60112               forward               forward               forward   \n",
       "\n",
       "      mutation_polarity_22A mutation_polarity_22B mutation_polarity_22C  \n",
       "0                   forward               forward               forward  \n",
       "1                   forward               forward               forward  \n",
       "2                   forward               forward               forward  \n",
       "3                   forward               forward               forward  \n",
       "4                   forward               forward               forward  \n",
       "...                     ...                   ...                   ...  \n",
       "60108               forward               forward               forward  \n",
       "60109               forward               forward               forward  \n",
       "60110               forward               forward               forward  \n",
       "60111               forward               forward               forward  \n",
       "60112               forward               forward               forward  \n",
       "\n",
       "[60113 rows x 45 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fitness_effects_by_clade = fitness_effects_by_clade_tidy.pivot(\n",
    "    index=[\"gene\", \"site\", \"mutation\"],\n",
    "    columns=\"clade\",\n",
    "    values=[\"fitness_effect\", \"expected_count\", \"mutation_polarity\"],\n",
    ")\n",
    "\n",
    "# https://towardsdatascience.com/how-to-flatten-multiindex-columns-and-rows-in-pandas-f5406c50e569\n",
    "fitness_effects_by_clade.columns = [\n",
    "    \"_\".join(col) for col in fitness_effects_by_clade.columns\n",
    "]\n",
    "\n",
    "fitness_effects_by_clade = fitness_effects_by_clade.reset_index()\n",
    "\n",
    "assert (\n",
    "    len(fitness_effects_by_clade)\n",
    "    == len(fitness_effects_by_clade.groupby([\"gene\", \"mutation\"]))\n",
    ")\n",
    "\n",
    "fitness_effects_by_clade"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b513a3dd-99e3-447b-8de8-b69a8971f8aa",
   "metadata": {},
   "source": [
    "Finally, create a data frame that has both the overall average and the per-clade estimates for the mutations:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "d6bf64c8-4a42-4ca6-b811-8d5ea66c0560",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>site</th>\n",
       "      <th>mutation</th>\n",
       "      <th>ref_codon</th>\n",
       "      <th>ref_amino_acid</th>\n",
       "      <th>fitness_effect</th>\n",
       "      <th>expected_count</th>\n",
       "      <th>n_clades_with_mutation</th>\n",
       "      <th>clades_with_mutation</th>\n",
       "      <th>clade_codons_with_mutation</th>\n",
       "      <th>...</th>\n",
       "      <th>mutation_polarity_20G</th>\n",
       "      <th>mutation_polarity_20I</th>\n",
       "      <th>mutation_polarity_21C</th>\n",
       "      <th>mutation_polarity_21I</th>\n",
       "      <th>mutation_polarity_21J</th>\n",
       "      <th>mutation_polarity_21K</th>\n",
       "      <th>mutation_polarity_21L</th>\n",
       "      <th>mutation_polarity_22A</th>\n",
       "      <th>mutation_polarity_22B</th>\n",
       "      <th>mutation_polarity_22C</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>E</td>\n",
       "      <td>9</td>\n",
       "      <td>T9I</td>\n",
       "      <td>ACA</td>\n",
       "      <td>T</td>\n",
       "      <td>1.03620</td>\n",
       "      <td>369.730</td>\n",
       "      <td>5</td>\n",
       "      <td>21K; 21L; 22A; 22B; 22C</td>\n",
       "      <td>21K (ATA); 21L (ATA); 22A (ATA); 22B (ATA); 22...</td>\n",
       "      <td>...</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>M</td>\n",
       "      <td>3</td>\n",
       "      <td>D3G</td>\n",
       "      <td>GAT</td>\n",
       "      <td>D</td>\n",
       "      <td>1.48940</td>\n",
       "      <td>134.820</td>\n",
       "      <td>1</td>\n",
       "      <td>21K</td>\n",
       "      <td>21K (GGT)</td>\n",
       "      <td>...</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>NaN</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>NaN</td>\n",
       "      <td>forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>M</td>\n",
       "      <td>3</td>\n",
       "      <td>D3N</td>\n",
       "      <td>GAT</td>\n",
       "      <td>D</td>\n",
       "      <td>0.33533</td>\n",
       "      <td>102.120</td>\n",
       "      <td>1</td>\n",
       "      <td>22B</td>\n",
       "      <td>22B (AAT)</td>\n",
       "      <td>...</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>NaN</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>NaN</td>\n",
       "      <td>forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>M</td>\n",
       "      <td>19</td>\n",
       "      <td>Q19E</td>\n",
       "      <td>CAA</td>\n",
       "      <td>Q</td>\n",
       "      <td>-1.00460</td>\n",
       "      <td>153.870</td>\n",
       "      <td>5</td>\n",
       "      <td>21K; 21L; 22A; 22B; 22C</td>\n",
       "      <td>21K (GAA); 21L (GAA); 22A (GAA); 22B (GAA); 22...</td>\n",
       "      <td>...</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>M</td>\n",
       "      <td>63</td>\n",
       "      <td>A63T</td>\n",
       "      <td>GCT</td>\n",
       "      <td>A</td>\n",
       "      <td>-1.09570</td>\n",
       "      <td>268.520</td>\n",
       "      <td>5</td>\n",
       "      <td>21K; 21L; 22A; 22B; 22C</td>\n",
       "      <td>21K (ACT); 21L (ACT); 22A (ACT); 22B (ACT); 22...</td>\n",
       "      <td>...</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>S</td>\n",
       "      <td>954</td>\n",
       "      <td>Q954H</td>\n",
       "      <td>CAA</td>\n",
       "      <td>Q</td>\n",
       "      <td>0.29488</td>\n",
       "      <td>292.690</td>\n",
       "      <td>5</td>\n",
       "      <td>21K; 21L; 22A; 22B; 22C</td>\n",
       "      <td>21K (CAT); 21L (CAT); 22A (CAT); 22B (CAT); 22...</td>\n",
       "      <td>...</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>reverse</td>\n",
       "      <td>reverse</td>\n",
       "      <td>reverse</td>\n",
       "      <td>reverse</td>\n",
       "      <td>reverse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>S</td>\n",
       "      <td>969</td>\n",
       "      <td>N969K</td>\n",
       "      <td>AAT</td>\n",
       "      <td>N</td>\n",
       "      <td>0.37826</td>\n",
       "      <td>88.056</td>\n",
       "      <td>5</td>\n",
       "      <td>21K; 21L; 22A; 22B; 22C</td>\n",
       "      <td>21K (AAA); 21L (AAA); 22A (AAA); 22B (AAA); 22...</td>\n",
       "      <td>...</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>reverse</td>\n",
       "      <td>reverse</td>\n",
       "      <td>reverse</td>\n",
       "      <td>reverse</td>\n",
       "      <td>reverse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>S</td>\n",
       "      <td>981</td>\n",
       "      <td>L981F</td>\n",
       "      <td>CTT</td>\n",
       "      <td>L</td>\n",
       "      <td>-1.93490</td>\n",
       "      <td>455.390</td>\n",
       "      <td>1</td>\n",
       "      <td>21K</td>\n",
       "      <td>21K (TTT)</td>\n",
       "      <td>...</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>reverse</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>S</td>\n",
       "      <td>982</td>\n",
       "      <td>S982A</td>\n",
       "      <td>TCA</td>\n",
       "      <td>S</td>\n",
       "      <td>1.72860</td>\n",
       "      <td>67.676</td>\n",
       "      <td>1</td>\n",
       "      <td>20I</td>\n",
       "      <td>20I (GCA)</td>\n",
       "      <td>...</td>\n",
       "      <td>forward</td>\n",
       "      <td>NaN</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>S</td>\n",
       "      <td>1118</td>\n",
       "      <td>D1118H</td>\n",
       "      <td>GAC</td>\n",
       "      <td>D</td>\n",
       "      <td>0.33954</td>\n",
       "      <td>74.501</td>\n",
       "      <td>1</td>\n",
       "      <td>20I</td>\n",
       "      <td>20I (CAC)</td>\n",
       "      <td>...</td>\n",
       "      <td>forward</td>\n",
       "      <td>NaN</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>119 rows × 52 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene  site mutation ref_codon ref_amino_acid  fitness_effect  \\\n",
       "0      E     9      T9I       ACA              T         1.03620   \n",
       "1      M     3      D3G       GAT              D         1.48940   \n",
       "2      M     3      D3N       GAT              D         0.33533   \n",
       "3      M    19     Q19E       CAA              Q        -1.00460   \n",
       "4      M    63     A63T       GCT              A        -1.09570   \n",
       "..   ...   ...      ...       ...            ...             ...   \n",
       "114    S   954    Q954H       CAA              Q         0.29488   \n",
       "115    S   969    N969K       AAT              N         0.37826   \n",
       "116    S   981    L981F       CTT              L        -1.93490   \n",
       "117    S   982    S982A       TCA              S         1.72860   \n",
       "118    S  1118   D1118H       GAC              D         0.33954   \n",
       "\n",
       "     expected_count  n_clades_with_mutation     clades_with_mutation  \\\n",
       "0           369.730                       5  21K; 21L; 22A; 22B; 22C   \n",
       "1           134.820                       1                      21K   \n",
       "2           102.120                       1                      22B   \n",
       "3           153.870                       5  21K; 21L; 22A; 22B; 22C   \n",
       "4           268.520                       5  21K; 21L; 22A; 22B; 22C   \n",
       "..              ...                     ...                      ...   \n",
       "114         292.690                       5  21K; 21L; 22A; 22B; 22C   \n",
       "115          88.056                       5  21K; 21L; 22A; 22B; 22C   \n",
       "116         455.390                       1                      21K   \n",
       "117          67.676                       1                      20I   \n",
       "118          74.501                       1                      20I   \n",
       "\n",
       "                            clade_codons_with_mutation  ...  \\\n",
       "0    21K (ATA); 21L (ATA); 22A (ATA); 22B (ATA); 22...  ...   \n",
       "1                                            21K (GGT)  ...   \n",
       "2                                            22B (AAT)  ...   \n",
       "3    21K (GAA); 21L (GAA); 22A (GAA); 22B (GAA); 22...  ...   \n",
       "4    21K (ACT); 21L (ACT); 22A (ACT); 22B (ACT); 22...  ...   \n",
       "..                                                 ...  ...   \n",
       "114  21K (CAT); 21L (CAT); 22A (CAT); 22B (CAT); 22...  ...   \n",
       "115  21K (AAA); 21L (AAA); 22A (AAA); 22B (AAA); 22...  ...   \n",
       "116                                          21K (TTT)  ...   \n",
       "117                                          20I (GCA)  ...   \n",
       "118                                          20I (CAC)  ...   \n",
       "\n",
       "    mutation_polarity_20G mutation_polarity_20I mutation_polarity_21C  \\\n",
       "0                 forward               forward               forward   \n",
       "1                 forward               forward               forward   \n",
       "2                 forward               forward               forward   \n",
       "3                 forward               forward               forward   \n",
       "4                 forward               forward               forward   \n",
       "..                    ...                   ...                   ...   \n",
       "114               forward               forward               forward   \n",
       "115               forward               forward               forward   \n",
       "116               forward               forward               forward   \n",
       "117               forward                   NaN               forward   \n",
       "118               forward                   NaN               forward   \n",
       "\n",
       "    mutation_polarity_21I mutation_polarity_21J mutation_polarity_21K  \\\n",
       "0                 forward               forward                   NaN   \n",
       "1                 forward               forward                   NaN   \n",
       "2                 forward               forward                   NaN   \n",
       "3                 forward               forward                   NaN   \n",
       "4                 forward               forward                   NaN   \n",
       "..                    ...                   ...                   ...   \n",
       "114               forward               forward               reverse   \n",
       "115               forward               forward               reverse   \n",
       "116               forward               forward               reverse   \n",
       "117               forward               forward               forward   \n",
       "118               forward               forward               forward   \n",
       "\n",
       "    mutation_polarity_21L mutation_polarity_22A mutation_polarity_22B  \\\n",
       "0                     NaN                   NaN                   NaN   \n",
       "1                 forward               forward                   NaN   \n",
       "2                 forward               forward                   NaN   \n",
       "3                     NaN                   NaN                   NaN   \n",
       "4                     NaN                   NaN                   NaN   \n",
       "..                    ...                   ...                   ...   \n",
       "114               reverse               reverse               reverse   \n",
       "115               reverse               reverse               reverse   \n",
       "116               forward               forward               forward   \n",
       "117               forward               forward               forward   \n",
       "118               forward               forward               forward   \n",
       "\n",
       "    mutation_polarity_22C  \n",
       "0                     NaN  \n",
       "1                 forward  \n",
       "2                 forward  \n",
       "3                     NaN  \n",
       "4                     NaN  \n",
       "..                    ...  \n",
       "114               reverse  \n",
       "115               reverse  \n",
       "116               forward  \n",
       "117               forward  \n",
       "118               forward  \n",
       "\n",
       "[119 rows x 52 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mut_effects = fitness_effects.merge(fitness_effects_by_clade, how=\"left\")\n",
    "\n",
    "mut_effects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d7259c0a-8200-4079-9840-aa4657fcab6c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>site</th>\n",
       "      <th>mutation</th>\n",
       "      <th>ref_codon</th>\n",
       "      <th>ref_amino_acid</th>\n",
       "      <th>fitness_effect</th>\n",
       "      <th>expected_count</th>\n",
       "      <th>n_clades_with_mutation</th>\n",
       "      <th>clades_with_mutation</th>\n",
       "      <th>clade_codons_with_mutation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>S</td>\n",
       "      <td>614</td>\n",
       "      <td>D614G</td>\n",
       "      <td>GAT</td>\n",
       "      <td>D</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>14</td>\n",
       "      <td>20A; 20B; 20C; 20E; 20G; 20I; 21C; 21I; 21J; 2...</td>\n",
       "      <td>20A (GGT); 20B (GGT); 20C (GGT); 20E (GGT); 20...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene  site mutation ref_codon ref_amino_acid  fitness_effect  \\\n",
       "103    S   614    D614G       GAT              D             NaN   \n",
       "\n",
       "     expected_count  n_clades_with_mutation  \\\n",
       "103             NaN                      14   \n",
       "\n",
       "                                  clades_with_mutation  \\\n",
       "103  20A; 20B; 20C; 20E; 20G; 20I; 21C; 21I; 21J; 2...   \n",
       "\n",
       "                            clade_codons_with_mutation  \n",
       "103  20A (GGT); 20B (GGT); 20C (GGT); 20E (GGT); 20...  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fitness_effects.query(\"site == 614\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd7efa62-6e52-4aa5-b611-d7abbe792e55",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
