{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "008b56e0-8e16-4e76-851c-edfdc35ca119",
   "metadata": {},
   "source": [
    "# Analyze fitness effects of mutations fixed in each clade\n",
    "\n",
    "Import Python modules:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6718e87a-8da0-49fa-a00a-6311caca0f83",
   "metadata": {},
   "outputs": [],
   "source": [
    "import altair as alt\n",
    "\n",
    "import Bio.Seq\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "import yaml\n",
    "\n",
    "_ = alt.data_transformers.disable_max_rows()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60c785d9-1e75-4959-928b-c92580c31c62",
   "metadata": {},
   "source": [
    "Now get variables from `snakemake`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f47affb5-0a48-475e-a522-7de36413e745",
   "metadata": {},
   "outputs": [],
   "source": [
    "if \"snakemake\" not in globals() and \"snakemake\" not in locals():\n",
    "    # variables set manually for interactive debugging\n",
    "    aa_fitness_csv = \"../results/aa_fitness/aa_fitness.csv\"\n",
    "    aamut_by_clade_csv = \"../results/aa_fitness/aamut_fitness_by_clade.csv\"\n",
    "    clade_founder_nts_csv = \"../results/clade_founder_nts/clade_founder_nts.csv\"\n",
    "    \n",
    "    with open(\"../config.yaml\") as f:\n",
    "        config = yaml.safe_load(f)\n",
    "    min_expected_count = config[\"min_expected_count\"]\n",
    "    ref = config[\"clade_fixed_muts_ref\"]\n",
    "    \n",
    "else:\n",
    "    # get variables from `snakemake` when running pipeline\n",
    "    aa_fitness_csv = snakemake.input.aafitness\n",
    "    aamut_by_clade_csv = snakemake.input.aamut_by_clade\n",
    "    clade_founder_nts_csv = snakemake.input.clade_founder_nts_csv\n",
    "    min_expected_count = snakemake.params.min_expected_count\n",
    "    ref = snakemake.params.ref"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a7a6ba8-34d3-4d77-835a-60e7ecf166d5",
   "metadata": {},
   "source": [
    "Read the input data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "80047a0c-872a-4a6d-9953-0c7e88e84fa0",
   "metadata": {},
   "outputs": [],
   "source": [
    "aa_fitness = pd.read_csv(aa_fitness_csv).rename(columns={\"aa_site\": \"site\"})\n",
    "aamut_by_clade = pd.read_csv(aamut_by_clade_csv)\n",
    "clade_founder_nts = pd.read_csv(clade_founder_nts_csv)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "296bb419-0dee-4044-bb61-04851754a75f",
   "metadata": {},
   "source": [
    "First we get the amino acids at each position in each gene in each clade founder and the reference clade.\n",
    "We ignore ORF1a and just look at ORF1ab:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "205adb3b-25c8-4713-8eb5-b2f46fd2e53c",
   "metadata": {},
   "outputs": [],
   "source": [
    "codon_to_aa = {\n",
    "    f\"{nt1}{nt2}{nt3}\": str(Bio.Seq.Seq(f\"{nt1}{nt2}{nt3}\").translate())\n",
    "    for nt1 in \"ACGT\" for nt2 in \"ACGT\" for nt3 in \"ACGT\"\n",
    "}\n",
    "\n",
    "assert ref in clade_founder_nts[\"clade\"].unique()\n",
    "\n",
    "clade_founder_aas = (\n",
    "    clade_founder_nts\n",
    "    [[\"clade\", \"gene\", \"codon\", \"codon_site\"]]\n",
    "    .drop_duplicates()\n",
    "    .assign(\n",
    "        gene=lambda x: x[\"gene\"].str.split(\";\"),\n",
    "        codon=lambda x: x[\"codon\"].str.split(\";\"),\n",
    "        site=lambda x: x[\"codon_site\"].str.split(\";\"),\n",
    "    )\n",
    "    .explode([\"gene\", \"codon\", \"site\"])\n",
    "    .query(\"gene != 'ORF1a'\")\n",
    "    .assign(\n",
    "        amino_acid=lambda x: x[\"codon\"].map(codon_to_aa),\n",
    "        site=lambda x: x[\"site\"].astype(int),\n",
    "    )\n",
    "    [[\"clade\", \"gene\", \"site\", \"codon\", \"amino_acid\"]]\n",
    "    .drop_duplicates()\n",
    ")\n",
    "\n",
    "clade_founder_aas = clade_founder_aas.merge(\n",
    "    (\n",
    "        clade_founder_aas\n",
    "        .query(\"clade == @ref\")\n",
    "        [[\"gene\", \"site\", \"codon\", \"amino_acid\"]]\n",
    "        .rename(columns={\"codon\": \"ref_codon\", \"amino_acid\": \"ref_amino_acid\"})\n",
    "    ),\n",
    "    on=[\"gene\", \"site\"],\n",
    "    validate=\"many_to_one\",\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae08c6fc-1436-4ddd-9643-84f51aaad9b7",
   "metadata": {},
   "source": [
    "Now get just the amino acid mutations in each clade founder, and summarize the number of such mutations:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5ea40145-5b24-46b9-b74d-fa0d9e3fb078",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_amino_acid_muts</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>clade</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20A</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20B</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20C</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20E</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20F</th>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20G</th>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20I</th>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20J</th>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21C</th>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21F</th>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21I</th>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21J</th>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21K</th>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21L</th>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22A</th>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22B</th>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22C</th>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22E</th>\n",
       "      <td>55</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       n_amino_acid_muts\n",
       "clade                   \n",
       "20A                    2\n",
       "20B                    4\n",
       "20C                    4\n",
       "20E                    5\n",
       "20F                    6\n",
       "20G                   11\n",
       "20I                   18\n",
       "20J                   21\n",
       "21C                    9\n",
       "21F                   11\n",
       "21I                   22\n",
       "21J                   26\n",
       "21K                   42\n",
       "21L                   48\n",
       "22A                   50\n",
       "22B                   48\n",
       "22C                   50\n",
       "22E                   55"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clade_founder_muts = clade_founder_aas.query(\"amino_acid != ref_amino_acid\")\n",
    "\n",
    "(\n",
    "    clade_founder_muts\n",
    "    .groupby(\"clade\")\n",
    "    .aggregate(n_amino_acid_muts=pd.NamedAgg(\"site\", \"count\"))\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a98233ef-7f48-4d2d-8ba5-9aa97feb7633",
   "metadata": {},
   "source": [
    "Get the estimated fitness effect of each mutation from the overall fitness estimates:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d186ced7-8f6b-437b-8d0c-24ca65e4d320",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>clade</th>\n",
       "      <th>gene</th>\n",
       "      <th>site</th>\n",
       "      <th>codon</th>\n",
       "      <th>amino_acid</th>\n",
       "      <th>ref_codon</th>\n",
       "      <th>ref_amino_acid</th>\n",
       "      <th>amino_acid_fitness</th>\n",
       "      <th>ref_amino_acid_fitness</th>\n",
       "      <th>fitness_effect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>21L</td>\n",
       "      <td>ORF1ab</td>\n",
       "      <td>135</td>\n",
       "      <td>AGG</td>\n",
       "      <td>R</td>\n",
       "      <td>AGT</td>\n",
       "      <td>S</td>\n",
       "      <td>0.44707</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.44707</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>22A</td>\n",
       "      <td>ORF1ab</td>\n",
       "      <td>135</td>\n",
       "      <td>AGG</td>\n",
       "      <td>R</td>\n",
       "      <td>AGT</td>\n",
       "      <td>S</td>\n",
       "      <td>0.44707</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.44707</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>22B</td>\n",
       "      <td>ORF1ab</td>\n",
       "      <td>135</td>\n",
       "      <td>AGG</td>\n",
       "      <td>R</td>\n",
       "      <td>AGT</td>\n",
       "      <td>S</td>\n",
       "      <td>0.44707</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.44707</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22C</td>\n",
       "      <td>ORF1ab</td>\n",
       "      <td>135</td>\n",
       "      <td>AGG</td>\n",
       "      <td>R</td>\n",
       "      <td>AGT</td>\n",
       "      <td>S</td>\n",
       "      <td>0.44707</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.44707</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>22E</td>\n",
       "      <td>ORF1ab</td>\n",
       "      <td>135</td>\n",
       "      <td>AGG</td>\n",
       "      <td>R</td>\n",
       "      <td>AGT</td>\n",
       "      <td>S</td>\n",
       "      <td>0.44707</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.44707</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>427</th>\n",
       "      <td>22A</td>\n",
       "      <td>N</td>\n",
       "      <td>413</td>\n",
       "      <td>CGT</td>\n",
       "      <td>R</td>\n",
       "      <td>AGT</td>\n",
       "      <td>S</td>\n",
       "      <td>0.92395</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.92395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>428</th>\n",
       "      <td>22B</td>\n",
       "      <td>N</td>\n",
       "      <td>413</td>\n",
       "      <td>CGT</td>\n",
       "      <td>R</td>\n",
       "      <td>AGT</td>\n",
       "      <td>S</td>\n",
       "      <td>0.92395</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.92395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>429</th>\n",
       "      <td>22C</td>\n",
       "      <td>N</td>\n",
       "      <td>413</td>\n",
       "      <td>CGT</td>\n",
       "      <td>R</td>\n",
       "      <td>AGT</td>\n",
       "      <td>S</td>\n",
       "      <td>0.92395</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.92395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>430</th>\n",
       "      <td>22E</td>\n",
       "      <td>N</td>\n",
       "      <td>413</td>\n",
       "      <td>CGT</td>\n",
       "      <td>R</td>\n",
       "      <td>AGT</td>\n",
       "      <td>S</td>\n",
       "      <td>0.92395</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.92395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>431</th>\n",
       "      <td>20E</td>\n",
       "      <td>ORF10</td>\n",
       "      <td>30</td>\n",
       "      <td>TTA</td>\n",
       "      <td>L</td>\n",
       "      <td>GTA</td>\n",
       "      <td>V</td>\n",
       "      <td>0.86409</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.86409</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>432 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    clade    gene  site codon amino_acid ref_codon ref_amino_acid  \\\n",
       "0     21L  ORF1ab   135   AGG          R       AGT              S   \n",
       "1     22A  ORF1ab   135   AGG          R       AGT              S   \n",
       "2     22B  ORF1ab   135   AGG          R       AGT              S   \n",
       "3     22C  ORF1ab   135   AGG          R       AGT              S   \n",
       "4     22E  ORF1ab   135   AGG          R       AGT              S   \n",
       "..    ...     ...   ...   ...        ...       ...            ...   \n",
       "427   22A       N   413   CGT          R       AGT              S   \n",
       "428   22B       N   413   CGT          R       AGT              S   \n",
       "429   22C       N   413   CGT          R       AGT              S   \n",
       "430   22E       N   413   CGT          R       AGT              S   \n",
       "431   20E   ORF10    30   TTA          L       GTA              V   \n",
       "\n",
       "     amino_acid_fitness  ref_amino_acid_fitness  fitness_effect  \n",
       "0               0.44707                     0.0         0.44707  \n",
       "1               0.44707                     0.0         0.44707  \n",
       "2               0.44707                     0.0         0.44707  \n",
       "3               0.44707                     0.0         0.44707  \n",
       "4               0.44707                     0.0         0.44707  \n",
       "..                  ...                     ...             ...  \n",
       "427             0.92395                     0.0         0.92395  \n",
       "428             0.92395                     0.0         0.92395  \n",
       "429             0.92395                     0.0         0.92395  \n",
       "430             0.92395                     0.0         0.92395  \n",
       "431             0.86409                     0.0         0.86409  \n",
       "\n",
       "[432 rows x 10 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# overall clade estimates of fitness effects of mutations\n",
    "fitness_effects = (\n",
    "    clade_founder_muts\n",
    "    .merge(\n",
    "        aa_fitness[[\"gene\", \"site\", \"aa\", \"fitness\"]].rename(\n",
    "            columns={\"aa\": \"amino_acid\", \"fitness\": \"amino_acid_fitness\"},\n",
    "        ),\n",
    "        on=[\"site\", \"gene\", \"amino_acid\"],\n",
    "        how=\"left\",\n",
    "        validate=\"many_to_one\",\n",
    "    )\n",
    "    .merge(\n",
    "        aa_fitness[[\"gene\", \"site\", \"aa\", \"fitness\"]].rename(\n",
    "            columns={\"aa\": \"ref_amino_acid\", \"fitness\": \"ref_amino_acid_fitness\"},\n",
    "        ),\n",
    "        on=[\"site\", \"gene\", \"ref_amino_acid\"],\n",
    "        how=\"left\",\n",
    "        validate=\"many_to_one\",\n",
    "    )\n",
    "    .assign(\n",
    "        fitness_effect=lambda x: x[\"amino_acid_fitness\"] - x[\"ref_amino_acid_fitness\"]\n",
    "    )\n",
    ")\n",
    "\n",
    "fitness_effects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4dd7d753-522e-4e52-93e1-0cc851738efe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>clade</th>\n",
       "      <th>gene</th>\n",
       "      <th>clade_founder_aa</th>\n",
       "      <th>mutant_aa</th>\n",
       "      <th>aa_site</th>\n",
       "      <th>aa_mutation</th>\n",
       "      <th>expected_count</th>\n",
       "      <th>actual_count</th>\n",
       "      <th>subset_of_ORF1ab</th>\n",
       "      <th>delta_fitness</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>20A</td>\n",
       "      <td>E</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>22</td>\n",
       "      <td>A22A</td>\n",
       "      <td>1.64940</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>0.15112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>20A</td>\n",
       "      <td>E</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>32</td>\n",
       "      <td>A32A</td>\n",
       "      <td>13.04700</td>\n",
       "      <td>17</td>\n",
       "      <td>False</td>\n",
       "      <td>0.25607</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>20A</td>\n",
       "      <td>E</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>36</td>\n",
       "      <td>A36A</td>\n",
       "      <td>13.21700</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>-2.21320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>20A</td>\n",
       "      <td>E</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>41</td>\n",
       "      <td>A41A</td>\n",
       "      <td>13.21700</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>-1.70240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>20A</td>\n",
       "      <td>E</td>\n",
       "      <td>A</td>\n",
       "      <td>D</td>\n",
       "      <td>22</td>\n",
       "      <td>A22D</td>\n",
       "      <td>0.58420</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.77399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1678588</th>\n",
       "      <td>22C</td>\n",
       "      <td>nsp1</td>\n",
       "      <td>Y</td>\n",
       "      <td>F</td>\n",
       "      <td>97</td>\n",
       "      <td>Y97F</td>\n",
       "      <td>0.42753</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>-0.61792</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1678589</th>\n",
       "      <td>22C</td>\n",
       "      <td>nsp1</td>\n",
       "      <td>Y</td>\n",
       "      <td>H</td>\n",
       "      <td>97</td>\n",
       "      <td>Y97H</td>\n",
       "      <td>1.64450</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>0.15338</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1678590</th>\n",
       "      <td>22C</td>\n",
       "      <td>nsp1</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>97</td>\n",
       "      <td>Y97N</td>\n",
       "      <td>0.33318</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>-0.51064</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1678591</th>\n",
       "      <td>22C</td>\n",
       "      <td>nsp1</td>\n",
       "      <td>Y</td>\n",
       "      <td>S</td>\n",
       "      <td>97</td>\n",
       "      <td>Y97S</td>\n",
       "      <td>0.15635</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>-0.27209</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1678592</th>\n",
       "      <td>22C</td>\n",
       "      <td>nsp1</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>97</td>\n",
       "      <td>Y97Y</td>\n",
       "      <td>14.01200</td>\n",
       "      <td>23</td>\n",
       "      <td>True</td>\n",
       "      <td>0.48202</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1678593 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        clade  gene clade_founder_aa mutant_aa  aa_site aa_mutation  \\\n",
       "0         20A     E                A         A       22        A22A   \n",
       "1         20A     E                A         A       32        A32A   \n",
       "2         20A     E                A         A       36        A36A   \n",
       "3         20A     E                A         A       41        A41A   \n",
       "4         20A     E                A         D       22        A22D   \n",
       "...       ...   ...              ...       ...      ...         ...   \n",
       "1678588   22C  nsp1                Y         F       97        Y97F   \n",
       "1678589   22C  nsp1                Y         H       97        Y97H   \n",
       "1678590   22C  nsp1                Y         N       97        Y97N   \n",
       "1678591   22C  nsp1                Y         S       97        Y97S   \n",
       "1678592   22C  nsp1                Y         Y       97        Y97Y   \n",
       "\n",
       "         expected_count  actual_count  subset_of_ORF1ab  delta_fitness  \n",
       "0               1.64940             2             False        0.15112  \n",
       "1              13.04700            17             False        0.25607  \n",
       "2              13.21700             1             False       -2.21320  \n",
       "3              13.21700             2             False       -1.70240  \n",
       "4               0.58420             0             False       -0.77399  \n",
       "...                 ...           ...               ...            ...  \n",
       "1678588         0.42753             0              True       -0.61792  \n",
       "1678589         1.64450             2              True        0.15338  \n",
       "1678590         0.33318             0              True       -0.51064  \n",
       "1678591         0.15635             0              True       -0.27209  \n",
       "1678592        14.01200            23              True        0.48202  \n",
       "\n",
       "[1678593 rows x 10 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aamut_by_clade"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "890f1776-7a9f-474b-9c37-d25a31a70634",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
