<h2>Interactive plots of estimated fitness effects of SARS-CoV-2 amino-acid mutations</h2>
<p>Fitness effects estimated from observed versus expected counts of mutations.</p>
<p>Analysis described in <a href="https://academic.oup.com/ve/article/9/2/vead055/7265011">Bloom and Neher (2023)</a>.</p>
<p>Interactive plots for public_2023-10-01 dataset:</p>
<ul>
<li>
<p>Heatmaps of effects of mutations</p>
<ul>
<li><a href="nsp1.html">nsp1 amino-acid fitnesses</a></li>
<li><a href="nsp2.html">nsp2 amino-acid fitnesses</a></li>
<li><a href="nsp3.html">nsp3 amino-acid fitnesses</a></li>
<li><a href="nsp4.html">nsp4 amino-acid fitnesses</a></li>
<li><a href="nsp5.html">nsp5 (Mpro) amino-acid fitnesses</a></li>
<li><a href="nsp6.html">nsp6 amino-acid fitnesses</a></li>
<li><a href="nsp7.html">nsp7 amino-acid fitnesses</a></li>
<li><a href="nsp8.html">nsp8 amino-acid fitnesses</a></li>
<li><a href="nsp9.html">nsp9 amino-acid fitnesses</a></li>
<li><a href="nsp10.html">nsp10 amino-acid fitnesses</a></li>
<li><a href="nsp12.html">nsp12 (RdRp) amino-acid fitnesses</a></li>
<li><a href="nsp13.html">nsp13 (helicase) amino-acid fitnesses</a></li>
<li><a href="nsp14.html">nsp14 (ExoN) amino-acid fitnesses</a></li>
<li><a href="nsp15.html">nsp15 amino-acid fitnesses</a></li>
<li><a href="nsp16.html">nsp16 amino-acid fitnesses</a></li>
<li><a href="S.html">S (spike) amino-acid fitnesses</a></li>
<li><a href="ORF3a.html">ORF3a amino-acid fitnesses</a></li>
<li><a href="E.html">E (envelope) amino-acid fitnesses</a></li>
<li><a href="M.html">M (membrane) amino-acid fitnesses</a></li>
<li><a href="ORF6.html">ORF6 amino-acid fitnesses</a></li>
<li><a href="ORF7a.html">ORF7a amino-acid fitnesses</a></li>
<li><a href="ORF7b.html">ORF7b amino-acid fitnesses</a></li>
<li><a href="ORF8.html">ORF8 amino-acid fitnesses</a></li>
<li><a href="ORF9b.html">ORF9b amino-acid fitnesses</a></li>
<li><a href="N.html">N (nucleocapsid) amino-acid fitnesses</a></li>
<li><a href="ORF10.html">ORF10 amino-acid fitnesses</a></li>
</ul>
</li>
<li>
<p>Expected counts per site</p>
<ul>
<li><a href="mut_rates.html">neutral mutation rates for each clade</a></li>
<li><a href="avg_counts.html">average per-site counts of each nucleotide mutation type</a></li>
</ul>
</li>
<li>
<p>Distribution of estimated mutation fitness effects</p>
<ul>
<li><a href="effects_histogram.html">distribution of fitness effects of mutations</a></li>
<li><a href="effects_dist.html">mutation fitness effects for each gene</a></li>
<li><a href="effects_dist_position_order.html">mutation fitness effects for each gene (ordered by genomic position)</a></li>
</ul>
</li>
<li>
<p>Correlations among effect estimates</p>
<ul>
<li><a href="clade_corr_chart.html">correlation of fitness estimates for different viral clades</a></li>
<li><a href="subset_corr_chart.html">correlation of fitness estimates for different countries</a></li>
<li><a href="spike_nonspike_corrs.html">average correlation of fitness estimates for spike and non-spike proteins</a></li>
<li><a href="spike_nonspike_corrs_by_divergence.html">clade correlations of fitness estimates versus protein divergence</a></li>
</ul>
</li>
<li>
<p>Correlations of estimates with deep mutational scanning</p>
<ul>
<li><a href="dms_S_corr.html">correlation of fitness estimates with spike deep mutational scanning</a></li>
<li><a href="dms_S_all_corr.html">correlation of fitness estimated with spike deep mutational scanning (all phenotypes)</a></li>
<li><a href="dms_nsp5_corr.html">correlation of fitness estimates with Mpro (nsp5) deep mutational scanning</a></li>
<li><a href="dms_nsp5_all_corr.html">correlation of fitness estimates with Mpro (nsp5) deep mutational scanning (all phenotypes)</a></li>
</ul>
</li>
<li>
<p>Effects of fixed mutations</p>
<ul>
<li><a href="clade_fixed_muts.html">fitness effects of mutations fixed in at least one clade</a></li>
<li><a href="clade_fixed_muts_hist.html">fitness effects of mutations fixed in a clade versus all mutations</a></li>
</ul>
</li>
<li>
<p>Effects of mutations versus number of descendants</p>
<ul>
<li><a href="fitness_vs_terminal.html">fitness effects versus number of descendant sequences</a></li>
</ul>
</li>
<li>
<p>Correlations among estimates from different sequence sets</p>
<ul>
<li><a href="mat_aa_fitness_correlations.html">correlation of fitness effects estimated from different sequence sets</a></li>
</ul>
</li>
<li>
<p>Comparisons to other methods of estimating mutation effects</p>
<ul>
<li><a href="dnds_corr.html">correlation of dN/dS with site-average fitness estimates and DMS data</a></li>
<li><a href="comparator_corr.html">correlation of mutation effect estimates from other studies with fitness estimates and spike DMS data</a></li>
</ul>
</li>
</ul>
<p>See <a href="https://academic.oup.com/ve/article/9/2/vead055/7265011">Bloom and Neher (2023)</a> for a paper describing the work.</p>
<p>See <a href="https://github.com/jbloomlab/SARS2-mut-fitness">https://github.com/jbloomlab/SARS2-mut-fitness</a> for full computer code and data.</p>
<p>See <a href="https://jbloomlab.github.io/SARS2-mut-fitness/">https://jbloomlab.github.io/SARS2-mut-fitness/</a> for links to all interactive plots.</p>
<p>Here are plots for other datasets:</p>
<ul>
<li><a href="public_2023-10-01/index.html">public_2023-10-01 dataset</a></li>
<li><a href="public_2023-06-21/index.html">public_2023-06-21 dataset</a></li>
<li><a href="public_2023-05-11/index.html">public_2023-05-11 dataset</a></li>
<li><a href="public_2023-03-27/index.html">public_2023-03-27 dataset</a></li>
<li><a href="public_2022-12-18/index.html">public_2022-12-18 dataset</a></li>
<li><a href="gisaid_2023-03-29/index.html">gisaid_2023-03-29 dataset</a></li>
<li><a href="public_2022-01-31/index.html">public_2022-01-31 dataset</a></li>
</ul>